From ff56cea2d9e233df3a8727bb53b139392eeb76f3 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 9 Apr 2021 16:17:52 +0200
Subject: [PATCH] Add isTicked method


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 9312f6f0..9f573ce3 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -1200,6 +1200,10 @@ public class Chunk {
         });
     }
 
+    private boolean isTicked() {
+        return this.p;
+    }
+
     public boolean isReady() {
         // Spigot Start
         /*
-- 
2.32.0.windows.2

