From 4824545abd384a27927cc2259aeb77b64c5818c6 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 28 Apr 2022 21:05:28 +0200
Subject: [PATCH] Cache bukkit version


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index ccf0673b..c2199b2a 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -8,22 +8,30 @@ import java.util.logging.Logger;
 import org.bukkit.Bukkit;
 
 public final class Versioning {
+
+    private static String bukkitVersion;
+
     public static String getBukkitVersion() {
-        String result = "Unknown-Version";
+        if (bukkitVersion != null) {
+            return bukkitVersion;
+        }
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/app.ashcon/sportpaper-api/pom.properties");
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/net.titanvale/titanium-api/pom.properties");
         Properties properties = new Properties();
 
         if (stream != null) {
             try {
                 properties.load(stream);
 
-                result = properties.getProperty("version");
+                bukkitVersion = properties.getProperty("version");
+
+                return bukkitVersion;
             } catch (IOException ex) {
                 Logger.getLogger(Versioning.class.getName()).log(Level.SEVERE, "Could not get Bukkit version!", ex);
             }
         }
 
-        return result;
+        return "Unknown-Version";
     }
+
 }
-- 
2.32.0.windows.2

