From 90d1fad464b522db554d4f07767869fe2671597e Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 28 Apr 2022 21:09:10 +0200
Subject: [PATCH] Rename to Titanium in Metrics


diff --git a/src/main/java/org/spigotmc/Metrics.java b/src/main/java/org/spigotmc/Metrics.java
index 558989ce..e1067201 100644
--- a/src/main/java/org/spigotmc/Metrics.java
+++ b/src/main/java/org/spigotmc/Metrics.java
@@ -344,7 +344,7 @@ public class Metrics {
      */
     private void postPlugin(final boolean isPing) throws IOException {
         // Server software specific section
-        String pluginName = "PaperSpigot"; // PaperSpigot - We need some usage data
+        String pluginName = "Titanium"; // PaperSpigot - We need some usage data // Titanium - Rename to Titanium
         boolean onlineMode = Bukkit.getServer().getOnlineMode(); // TRUE if online mode is enabled
         String pluginVersion = (Metrics.class.getPackage().getImplementationVersion() != null) ? Metrics.class.getPackage().getImplementationVersion() : "unknown";
         String serverVersion = Bukkit.getVersion();
-- 
2.32.0.windows.2

