From bb7345ab7ea6a8e9f79ca0124107a9449c3d65de Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 8 May 2022 23:12:36 +0200
Subject: [PATCH] Implement [FlamePaper-0024] Disable Unloaded Chunk Movement


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 3964d4e9c..2c69cd7ca 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -7,6 +7,7 @@ import java.util.UUID;
 import java.util.concurrent.Callable;
 
 // CraftBukkit start
+import net.titanium.config.TitaniumConfig;
 import org.apache.logging.log4j.LogManager;
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
@@ -461,6 +462,12 @@ public abstract class Entity implements ICommandListener {
 
     public void move(double d0, double d1, double d2) {
         if (this.loadChunks) loadChunks(); // PaperSpigot - Load chunks
+        // FlamePaper start - Disable Unloaded Chunk Movement
+        if (!((ChunkProviderServer) world.chunkProvider).isChunkLoaded((int) locX >> 4, (int) locZ >> 4) && TitaniumConfig.get().disableUnloadedChunksMovement) {
+            this.a(this.getBoundingBox().c(d0, d1, d2));
+            this.recalcPosition();
+        } else
+         // FlamePaper end - Disable Unloaded Chunk Movement
         if (this.noclip) {
             this.a(this.getBoundingBox().c(d0, d1, d2));
             this.recalcPosition();
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 3289f6af0..d7527e302 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -54,6 +54,9 @@ public class TitaniumConfig extends OkaeriConfig {
     @Comment("Smaller values (Our 50000 is probably enough, but smaller value would be even better) prevents most NBT related overflow exploits (for eg. Books)")
     public long nbtReadLimiter = 50000L;
 
+    // TODO comment this
+    public boolean disableUnloadedChunksMovement = false;
+
     @Comment("Configuration of book limits.")
     @Comment("Setting these values prevent players from using NBT overflow exploits.")
     @Comment("See also nbt-read-limiter, maybe you don't have to change these values.")
-- 
2.36.0.windows.1

