From ec14eaa3f1ef2dd2a1eb163eed2889c683cd6502 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 27 Apr 2022 00:10:03 +0200
Subject: [PATCH] This is Titanium

This is Titanium 2

diff --git a/pom.xml b/pom.xml
index 76533897..87992a63 100644
--- a/pom.xml
+++ b/pom.xml
@@ -4,12 +4,12 @@
     <modelVersion>4.0.0</modelVersion>
 
     <parent>
-        <groupId>app.ashcon</groupId>
-        <artifactId>sportpaper-parent</artifactId>
-        <version>1.8-SNAPSHOT</version>
+        <groupId>net.titanvale</groupId>
+        <artifactId>titanium-parent</artifactId>
+        <version>1.8.8-SNAPSHOT</version>
     </parent>
 
-    <artifactId>sportpaper-api</artifactId>
+    <artifactId>titanium-api</artifactId>
     <version>1.8.8-R0.1-SNAPSHOT</version>
     <packaging>jar</packaging>
 
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 45dce10e..3ea91bd5 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -185,8 +185,12 @@ public class VersionCommand extends BukkitCommand {
     private void obtainVersion() {
         String version = Bukkit.getVersion();
         if (version == null) version = "Custom";
+        // Titanium start
+        if (version.startsWith("git-Titanium-")) {
+            setVersionMessage("Unable to check for updates at this time. Visit https://github.com/TitanVale/Titanium/releases for newer versions"); // TODO: Obtain version list and warn of updates
+        // Titanium end
         // SportPaper start
-        if (version.startsWith("git-SportPaper-")) {
+        } else if (version.startsWith("git-SportPaper-")) {
             setVersionMessage("Unable to check for updates at this time. Visit https://github.com/Electroid/SportPaper/releases for newer versions"); // TODO: Obtain version list and warn of updates
         // SportPaper end
         // PaperSpigot start
-- 
2.32.0.windows.2

