From bc20766a257edaba732312fc9220a8513a77f3fd Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 29 Apr 2022 20:44:16 +0200
Subject: [PATCH] Add ChatColor#decolor


diff --git a/src/main/java/org/bukkit/ChatColor.java b/src/main/java/org/bukkit/ChatColor.java
index adbae51a..c0cfbeee 100644
--- a/src/main/java/org/bukkit/ChatColor.java
+++ b/src/main/java/org/bukkit/ChatColor.java
@@ -215,7 +215,7 @@ public enum ChatColor{
      * you need to dynamically convert colour codes from your custom format.
      */
     public static final char COLOR_CHAR = '\u00A7';
-    private static final Pattern STRIP_COLOR_PATTERN = Pattern.compile("(?i)" + String.valueOf(COLOR_CHAR) + "[0-9A-FK-OR]");
+    private static final Pattern STRIP_COLOR_PATTERN = Pattern.compile("(?i)" + COLOR_CHAR + "[0-9A-FK-OR]");
 
     private final int intCode;
     private final char code;
@@ -310,6 +310,16 @@ public enum ChatColor{
         return STRIP_COLOR_PATTERN.matcher(input).replaceAll("");
     }
 
+    /**
+     * Replace § in input with &
+     *
+     * @param input String to replace
+     * @return String with § replaced with &
+     */
+    public static String decolor(final String input) {
+        return STRIP_COLOR_PATTERN.matcher(input).replaceAll("&$1");
+    }
+
     /**
      * Translates a string using an alternate color code character into a
      * string that uses the internal ChatColor.COLOR_CODE color code
-- 
2.32.0.windows.2

