From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 5 Jun 2022 02:40:10 +0200
Subject: [PATCH] Add more informations to /tps command


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 7c2311b7b777ed2afd6f49abd1468ba912c7802c..71de45380cabef43bb407abde25a85695ba99582 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -4,6 +4,8 @@ import eu.okaeri.configs.OkaeriConfig;
 import java.awt.image.BufferedImage;
 import java.io.File;
 import java.io.Serializable;
+import java.time.Duration;
+import java.time.Instant;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
@@ -1054,6 +1056,20 @@ public interface Server extends PluginMessageRecipient {
             throw new UnsupportedOperationException("Not supported yet.");
         }
 
+        /**
+         * @return time when the server was started
+         */
+        public Instant getStartTime() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
+        /**
+         * @return time which elapsed since the server was started
+         */
+        public Duration getUptime() {
+            return Duration.between(getStartTime(), Instant.now());
+        }
+
     }
 
     Titanium titanium();
