From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 5 Jun 2022 02:40:10 +0200
Subject: [PATCH] Add more informations to /tps command


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 7699908797b2183536f28e58f2f7a7ef49d9f1de..150560f611f8447294033f28709ad1e19524d678 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -4,6 +4,8 @@ import eu.okaeri.configs.OkaeriConfig;
 import java.awt.image.BufferedImage;
 import java.io.File;
 import java.io.Serializable;
+import java.time.Duration;
+import java.time.Instant;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
@@ -1051,6 +1053,14 @@ public interface Server extends PluginMessageRecipient {
             throw new UnsupportedOperationException( "Not supported yet." );
         }
 
+        public Instant getStartTime() {
+            throw new UnsupportedOperationException( "Not supported yet." );
+        }
+
+        public Duration getUptime() {
+            return Duration.between( getStartTime(), Instant.now() );
+        }
+
     }
 
     Titanium titanium();
