From 6678210b19f17398d7be2d74a4dbc41c5df442d0 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 5 Jun 2022 02:40:10 +0200
Subject: [PATCH] Add more informations to /tps command


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 76999087..150560f6 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -4,6 +4,8 @@ import eu.okaeri.configs.OkaeriConfig;
 import java.awt.image.BufferedImage;
 import java.io.File;
 import java.io.Serializable;
+import java.time.Duration;
+import java.time.Instant;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
@@ -1051,6 +1053,14 @@ public interface Server extends PluginMessageRecipient {
             throw new UnsupportedOperationException( "Not supported yet." );
         }
 
+        public Instant getStartTime() {
+            throw new UnsupportedOperationException( "Not supported yet." );
+        }
+
+        public Duration getUptime() {
+            return Duration.between( getStartTime(), Instant.now() );
+        }
+
     }
 
     Titanium titanium();
-- 
2.36.0.windows.1

