From 31e5d47330002b8674872c4659c7602ff99f474e Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 5 Jun 2022 02:00:07 +0200
Subject: [PATCH] Add allias for /gamemode


diff --git a/src/main/java/org/bukkit/command/defaults/GameModeCommand.java b/src/main/java/org/bukkit/command/defaults/GameModeCommand.java
index f9675a05..9d541bd8 100644
--- a/src/main/java/org/bukkit/command/defaults/GameModeCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/GameModeCommand.java
@@ -1,6 +1,7 @@
 package org.bukkit.command.defaults;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 
 import org.apache.commons.lang.Validate;
@@ -23,6 +24,7 @@ public class GameModeCommand extends VanillaCommand {
         this.description = "Changes the player to a specific game mode";
         this.usageMessage = "/gamemode <mode> [player]";
         this.setPermission("bukkit.command.gamemode");
+        this.setAliases(Arrays.asList("gm"));
     }
 
     @Override
-- 
2.36.0.windows.1

