From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Thu, 30 Apr 2020 16:56:54 +0200
Subject: [PATCH] Implement [Paper-0206] Add Raw Byte ItemStack Serialization
 by Mariell Hoversholm


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 568c83e371b700a441267661d5b10404d6b7b6a8..ee9156314390fedd867f142d2a826116458aaadb 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -30,4 +30,10 @@ public interface UnsafeValues {
     Achievement getAchievementFromInternalName(String name);
 
     List<String> tabCompleteInternalStatisticOrAchievementName(String token, List<String> completions);
+
+    // Paper start
+    byte[] serializeItem(ItemStack item);
+
+    ItemStack deserializeItem(byte[] data);
+    // Paper end
 }
diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index d9cff5b0df6e43ef0eb23227c4a7d0f4a0140d71..b40477b261284da56a5e396cf6aaae0fb6b01691 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -559,6 +559,26 @@ public class ItemStack implements Cloneable, ConfigurationSerializable {
         return result;
     }
 
+    // Paper start
+    /**
+     * Deserializes this ItemStack from raw NBT bytes.
+     *
+     * @param bytes bytes representing an item in NBT
+     * @return Deserialized ItemStack.
+     */
+    public static ItemStack deserializeBytes(byte[] bytes) {
+        return org.bukkit.Bukkit.getUnsafe().deserializeItem(bytes);
+    }
+
+    /**
+     * Serializes this ItemStack to raw bytes in NBT.
+     * @return bytes representing this item in NBT.
+     */
+    public byte[] serializeAsBytes() {
+        return org.bukkit.Bukkit.getUnsafe().serializeItem(this);
+    }
+    // Paper end
+
     /**
      * Get a copy of this ItemStack's {@link ItemMeta}.
      *
