From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 16 Nov 2022 14:20:02 +0100
Subject: [PATCH] Backport Plugin#getSLF4JLogger API


diff --git a/build.gradle b/build.gradle
index 33974e4247d93a7775970328d40822337430d547..aa9baf0598862940737dba7b230ad1a27c39866d 100644
--- a/build.gradle
+++ b/build.gradle
@@ -19,6 +19,7 @@ dependencies {
     api 'eu.okaeri:okaeri-configs-serdes-commons:4.0.7'
     // Titanium end
     api 'org.apache.commons:commons-lang3:3.12.0' // Titanium - updated commons-lang3 to 3.12.0 // Titanium - Add commons-lang3 to API
+    api 'org.slf4j:slf4j-simple:1.8.0-beta4' // Titanium - backport Plugin#getSLF4JLogger from Paper
     api 'org.jetbrains:annotations:23.0.0' // Titanium - add jetbrains annotations - future is now 😎
     // Titanium libraries end
 
diff --git a/src/main/java/org/bukkit/plugin/Plugin.java b/src/main/java/org/bukkit/plugin/Plugin.java
index 7bdc809c6d99eb837c276d43ebae36cd8ddb07d6..3f85b51fdecf05a192b3b09e48a22d70623e241f 100644
--- a/src/main/java/org/bukkit/plugin/Plugin.java
+++ b/src/main/java/org/bukkit/plugin/Plugin.java
@@ -10,6 +10,7 @@ import org.bukkit.configuration.file.FileConfiguration;
 import org.bukkit.generator.ChunkGenerator;
 
 import com.avaje.ebean.EbeanServer;
+import org.jetbrains.annotations.NotNull;
 
 /**
  * Represents a Plugin
@@ -177,6 +178,13 @@ public interface Plugin extends TabExecutor {
      */
     public Logger getLogger();
 
+    // Paper start - Add SLF4J logger
+    @NotNull
+    default org.slf4j.Logger getSLF4JLogger() {
+        return org.slf4j.LoggerFactory.getLogger(getLogger().getName());
+    }
+    // Paper end
+
     /**
      * Returns the name of the plugin.
      * <p>
