From 9fcd6373bedf3c22fc615d0f6c038c847a794eb7 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sat, 18 Jun 2022 20:22:25 +0200
Subject: [PATCH] Fix compilation warnings


diff --git a/src/main/java/org/bukkit/ChatColor.java b/src/main/java/org/bukkit/ChatColor.java
index b92e337b..984db3b0 100644
--- a/src/main/java/org/bukkit/ChatColor.java
+++ b/src/main/java/org/bukkit/ChatColor.java
@@ -313,10 +313,10 @@ public enum ChatColor{
     }
 
     /**
-     * Replace § in input with &
+     * Replace § in input with {@literal &}
      *
      * @param input String to replace
-     * @return String with § replaced with &
+     * @return String with § replaced with {@literal &}
      */
     public static String decolor(String input) {
         if (input == null) {
diff --git a/src/main/java/org/bukkit/material/Door.java b/src/main/java/org/bukkit/material/Door.java
index e02d8e66..07a40a00 100644
--- a/src/main/java/org/bukkit/material/Door.java
+++ b/src/main/java/org/bukkit/material/Door.java
@@ -11,13 +11,13 @@ import org.bukkit.block.BlockFace;
  * work with modern doors. Some methods are undefined dependant on <code>isTopHalf()</code>
  * due to Minecraft's internal representation of doors.
  *
- * @see Material.WOODEN_DOOR
- * @see Material.IRON_DOOR_BLOCK
- * @see Material.SPRUCE_DOOR
- * @see Material.BIRCH_DOOR
- * @see Material.JUNGLE_DOOR
- * @see Material.ACACIA_DOOR
- * @see Material.DARK_OAK_DOOR
+ * @see Material#WOODEN_DOOR
+ * @see Material#IRON_DOOR_BLOCK
+ * @see Material#SPRUCE_DOOR
+ * @see Material#BIRCH_DOOR
+ * @see Material#JUNGLE_DOOR
+ * @see Material#ACACIA_DOOR
+ * @see Material#DARK_OAK_DOOR
  */
 public class Door extends MaterialData implements Directional, Openable {
 
@@ -51,18 +51,18 @@ public class Door extends MaterialData implements Directional, Openable {
      * @param type The type of material this door is made of. This must match the type of the block above.
      * @param face The direction the door is facing.
      *
-     * @see Material.WOODEN_DOOR
-     * @see Material.IRON_DOOR_BLOCK
-     * @see Material.SPRUCE_DOOR
-     * @see Material.BIRCH_DOOR
-     * @see Material.JUNGLE_DOOR
-     * @see Material.ACACIA_DOOR
-     * @see Material.DARK_OAK_DOOR
+     * @see Material#WOODEN_DOOR
+     * @see Material#IRON_DOOR_BLOCK
+     * @see Material#SPRUCE_DOOR
+     * @see Material#BIRCH_DOOR
+     * @see Material#JUNGLE_DOOR
+     * @see Material#ACACIA_DOOR
+     * @see Material#DARK_OAK_DOOR
      *
-     * @see BlockFace.WEST
-     * @see BlockFace.NORTH
-     * @see BlockFace.EAST
-     * @see BlockFace.SOUTH
+     * @see BlockFace#WEST
+     * @see BlockFace#NORTH
+     * @see BlockFace#EAST
+     * @see BlockFace#SOUTH
      */
     public Door(final Material type, BlockFace face) {
         this(type, face, false);
@@ -76,18 +76,18 @@ public class Door extends MaterialData implements Directional, Openable {
      * @param face The direction the door is facing.
      * @param isOpen Whether the door is currently opened.
      *
-     * @see Material.WOODEN_DOOR
-     * @see Material.IRON_DOOR_BLOCK
-     * @see Material.SPRUCE_DOOR
-     * @see Material.BIRCH_DOOR
-     * @see Material.JUNGLE_DOOR
-     * @see Material.ACACIA_DOOR
-     * @see Material.DARK_OAK_DOOR
+     * @see Material#WOODEN_DOOR
+     * @see Material#IRON_DOOR_BLOCK
+     * @see Material#SPRUCE_DOOR
+     * @see Material#BIRCH_DOOR
+     * @see Material#JUNGLE_DOOR
+     * @see Material#ACACIA_DOOR
+     * @see Material#DARK_OAK_DOOR
      *
-     * @see BlockFace.WEST
-     * @see BlockFace.NORTH
-     * @see BlockFace.EAST
-     * @see BlockFace.SOUTH
+     * @see BlockFace#WEST
+     * @see BlockFace#NORTH
+     * @see BlockFace#EAST
+     * @see BlockFace#SOUTH
      */
     public Door(final Material type, BlockFace face, boolean isOpen) {
         super(type);
@@ -102,13 +102,13 @@ public class Door extends MaterialData implements Directional, Openable {
      * @param type The type of material this door is made of. This must match the type of the block below.
      * @param isHingeRight True if the hinge is on the right hand side, false if the hinge is on the left hand side.
      *
-     * @see Material.WOODEN_DOOR
-     * @see Material.IRON_DOOR_BLOCK
-     * @see Material.SPRUCE_DOOR
-     * @see Material.BIRCH_DOOR
-     * @see Material.JUNGLE_DOOR
-     * @see Material.ACACIA_DOOR
-     * @see Material.DARK_OAK_DOOR
+     * @see Material#WOODEN_DOOR
+     * @see Material#IRON_DOOR_BLOCK
+     * @see Material#SPRUCE_DOOR
+     * @see Material#BIRCH_DOOR
+     * @see Material#JUNGLE_DOOR
+     * @see Material#ACACIA_DOOR
+     * @see Material#DARK_OAK_DOOR
      */
     public Door(final Material type, boolean isHingeRight) {
         super(type);
@@ -125,10 +125,10 @@ public class Door extends MaterialData implements Directional, Openable {
      *
      * @see TreeSpecies
      *
-     * @see BlockFace.WEST
-     * @see BlockFace.NORTH
-     * @see BlockFace.EAST
-     * @see BlockFace.SOUTH
+     * @see BlockFace#WEST
+     * @see BlockFace#NORTH
+     * @see BlockFace#EAST
+     * @see BlockFace#SOUTH
      */
     public Door(final TreeSpecies species, BlockFace face) {
         this(getWoodDoorOfSpecies(species), face, false);
@@ -144,10 +144,10 @@ public class Door extends MaterialData implements Directional, Openable {
      *
      * @see TreeSpecies
      *
-     * @see BlockFace.WEST
-     * @see BlockFace.NORTH
-     * @see BlockFace.EAST
-     * @see BlockFace.SOUTH
+     * @see BlockFace#WEST
+     * @see BlockFace#NORTH
+     * @see BlockFace#EAST
+     * @see BlockFace#SOUTH
      */
     public Door(final TreeSpecies species, BlockFace face, boolean isOpen) {
         this(getWoodDoorOfSpecies(species), face, isOpen);
@@ -191,12 +191,12 @@ public class Door extends MaterialData implements Directional, Openable {
      * @param species The species of wood door required.
      * @return The item type for the given species.
      *
-     * @see Material.WOODEN_DOOR
-     * @see Material.SPRUCE_DOOR
-     * @see Material.BIRCH_DOOR
-     * @see Material.JUNGLE_DOOR
-     * @see Material.ACACIA_DOOR
-     * @see Material.DARK_OAK_DOOR
+     * @see Material#WOODEN_DOOR
+     * @see Material#SPRUCE_DOOR
+     * @see Material#BIRCH_DOOR
+     * @see Material#JUNGLE_DOOR
+     * @see Material#ACACIA_DOOR
+     * @see Material#DARK_OAK_DOOR
      */
     public static Material getWoodDoorOfSpecies(TreeSpecies species) {
         switch (species) {
diff --git a/src/test/java/org/bukkit/metadata/FixedMetadataValueTest.java b/src/test/java/org/bukkit/metadata/FixedMetadataValueTest.java
index 5583b274..677a5517 100644
--- a/src/test/java/org/bukkit/metadata/FixedMetadataValueTest.java
+++ b/src/test/java/org/bukkit/metadata/FixedMetadataValueTest.java
@@ -13,15 +13,15 @@ public class FixedMetadataValueTest {
 
     @Test
     public void testBasic() {
-        subject = new FixedMetadataValue(plugin, new Integer(50));
+        subject = new FixedMetadataValue(plugin, 50);
         assertSame(plugin, subject.getOwningPlugin());
-        assertEquals(new Integer(50), subject.value());
+        assertEquals(50, subject.value());
     }
 
     @Test
     public void testNumberTypes() {
-        subject = new FixedMetadataValue(plugin, new Integer(5));
-        assertEquals(new Integer(5), subject.value());
+        subject = new FixedMetadataValue(plugin, 5);
+        assertEquals(5, subject.value());
         assertEquals(5, subject.asInt());
         assertEquals(true, subject.asBoolean());
         assertEquals(5, subject.asByte());
-- 
2.38.1.windows.1

