From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 22 Jun 2022 09:05:16 +0200
Subject: [PATCH] Backport Player#sendActionBar API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 61115f886a525a2ab8be983a1fc129dc26b9b6d3..07cfd8566bbf2997aa66313acb5263940f07879a 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2,6 +2,8 @@ package org.bukkit.entity;
 
 import java.net.InetSocketAddress;
 
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.TextComponent;
 import org.bukkit.Achievement;
 import org.bukkit.ChatColor;
 import org.bukkit.Effect;
@@ -429,6 +431,7 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     @Deprecated
     public void showTitle(net.md_5.bungee.api.chat.BaseComponent title, net.md_5.bungee.api.chat.BaseComponent subtitle, int fadeInTicks, int stayTicks, int fadeOutTicks);
 
+    // Titanium start
     /**
      * Show the title to the player, overriding any previously displayed title.
      *
@@ -444,6 +447,7 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     default void sendTitle(String title, String subtitle, int fadeInTicks, int stayTicks, int fadeOutTicks) {
         this.sendTitle(new Title(title, subtitle, fadeInTicks, stayTicks, fadeOutTicks));
     }
+    // Titanium end
 
     /**
      * Show the title to the player, overriding any previously displayed title.
@@ -455,6 +459,7 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     void sendTitle(Title title);
 
+    // Titanium start
     /**
      * Show the title to the player, overriding any previously displayed title.
      *
@@ -470,6 +475,7 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     default void updateTitle(String title, String subtitle, int fadeInTicks, int stayTicks, int fadeOutTicks) {
         this.updateTitle(new Title(title, subtitle, fadeInTicks, stayTicks, fadeOutTicks));
     }
+    // Titanium end
 
     /**
      * Show the title to the player, overriding any previously displayed title.
@@ -487,6 +493,30 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     public void hideTitle();
     // Paper end
 
+    // Titanium start
+    /**
+     * Sends an Action Bar message to the client.
+     *
+     * @param message message to display
+     */
+    default void sendActionBar(String... message) {
+        BaseComponent[] messageComponents = new BaseComponent[message.length];
+        int i = 0;
+        for (String messageLine : message) {
+            messageComponents[i] = new TextComponent(messageLine);
+            i++;
+        }
+        sendActionBar(messageComponents);
+    }
+
+    /**
+     * Sends an Action Bar message to the client.
+     *
+     * @param message message to display
+     */
+    void sendActionBar(BaseComponent... message);
+    // Titanium end
+
     /**
      * Forces an update of the player's entire inventory.
      *
