From 15b15f21cbf4bb8571c6752fb34a0626d3e11f8a Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 16 Jun 2022 23:41:55 +0200
Subject: [PATCH] Update okaeri-configs to 4.0.5


diff --git a/build.gradle b/build.gradle
index 4b1fa2eb..977b00c3 100644
--- a/build.gradle
+++ b/build.gradle
@@ -14,8 +14,8 @@ dependencies {
     // Minecraft libraries end
 
     // Titanium start - Use okaeri-configs for server configuration
-    api 'eu.okaeri:okaeri-configs-yaml-bukkit:4.0.1'
-    api 'eu.okaeri:okaeri-configs-serdes-commons:4.0.1'
+    api 'eu.okaeri:okaeri-configs-yaml-bukkit:4.0.5'
+    api 'eu.okaeri:okaeri-configs-serdes-commons:4.0.5'
     // Titanium end
 
     api 'org.jetbrains:annotations:16.0.2' // Titanium - add jetbrains annotations - future is now 😎
-- 
2.36.0.windows.1

