From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 6 Jul 2022 10:28:28 +0200
Subject: [PATCH] Backport Plugin#getSLF4JLogger API


diff --git a/build.gradle b/build.gradle
index 33b6a6ee5b54b43dffc218e799c4943dbff539dd..66247c7923c6eff9acc52234af74c118f617eeb3 100644
--- a/build.gradle
+++ b/build.gradle
@@ -13,10 +13,13 @@ dependencies {
     compileOnlyApi 'net.sf.trove4j:trove4j:3.0.3' // provided by CraftBukkit
     // Minecraft libraries end
 
+    // Titanium libraries start
     // Titanium start - Use okaeri-configs for server configuration
     api 'eu.okaeri:okaeri-configs-yaml-bukkit:4.0.6'
     api 'eu.okaeri:okaeri-configs-serdes-commons:4.0.6'
     // Titanium end
+    api 'org.slf4j:slf4j-simple:1.7.36' // Titanium - backport Plugin#getSLF4JLogger from Paper
+    // Titanium libraries end
 
     api 'org.jetbrains:annotations:16.0.2' // Titanium - add jetbrains annotations - future is now 😎
 
diff --git a/src/main/java/org/bukkit/plugin/Plugin.java b/src/main/java/org/bukkit/plugin/Plugin.java
index 7bdc809c6d99eb837c276d43ebae36cd8ddb07d6..3f85b51fdecf05a192b3b09e48a22d70623e241f 100644
--- a/src/main/java/org/bukkit/plugin/Plugin.java
+++ b/src/main/java/org/bukkit/plugin/Plugin.java
@@ -10,6 +10,7 @@ import org.bukkit.configuration.file.FileConfiguration;
 import org.bukkit.generator.ChunkGenerator;
 
 import com.avaje.ebean.EbeanServer;
+import org.jetbrains.annotations.NotNull;
 
 /**
  * Represents a Plugin
@@ -177,6 +178,13 @@ public interface Plugin extends TabExecutor {
      */
     public Logger getLogger();
 
+    // Paper start - Add SLF4J logger
+    @NotNull
+    default org.slf4j.Logger getSLF4JLogger() {
+        return org.slf4j.LoggerFactory.getLogger(getLogger().getName());
+    }
+    // Paper end
+
     /**
      * Returns the name of the plugin.
      * <p>
