From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 24 Jul 2022 18:28:13 +0200
Subject: [PATCH] Implement [Purpur-0046] Spark Profiler


diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 17d5aa2de8ee814382f126eb908502f72f282340..37b5c57c26e9534bd2d7ee7f41476252b252e5ce 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -158,6 +158,12 @@ public final class SimplePluginManager implements PluginManager {
                 server.getLogger().log(Level.SEVERE, "Could not load '" + file.getPath() + "' in folder '" + file.getParentFile().getPath() + "'", ex); // Paper
                 continue;
             }
+             // Purpur start
+            if (!Boolean.getBoolean("Purpur.IReallyDontWantSpark") && plugins.containsKey(description.getName()) && description.getName().equalsIgnoreCase("spark")) {
+                server.getLogger().log(Level.INFO, "Purpur: Using user-provided spark plugin instead of our own.");
+                continue;
+            }
+            // Purpur end
 
             File replacedFile = plugins.put(description.getName(), file);
             if (replacedFile != null) {
