From 64b70ace7470f5972068fe0c0a30dae5e0e7a5ff Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 22 Jul 2022 13:08:54 +0200
Subject: [PATCH] Fix Command#testPermissionSilent


diff --git a/src/main/java/org/bukkit/command/Command.java b/src/main/java/org/bukkit/command/Command.java
index bbf4cbd4..523b4679 100644
--- a/src/main/java/org/bukkit/command/Command.java
+++ b/src/main/java/org/bukkit/command/Command.java
@@ -265,7 +265,7 @@ public abstract class Command {
      * @return true if they can use it, otherwise false
      */
     public boolean testPermissionSilent(CommandSender target) {
-        return this.testPermission(target, this.permission); // Titanium-
+        return this.testPermissionSilent(target, this.permission); // Titanium
     }
 
     /**
-- 
2.36.0.windows.1

