From a7b2b4ecb761fccc3382b60903c9cd3f06e7a85c Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 28 Aug 2022 14:55:50 +0200
Subject: [PATCH] Add EntityType#fromEnumName and #findType


diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index ccc024bf..7c6ca475 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -13,6 +13,7 @@ import org.bukkit.entity.minecart.StorageMinecart;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.Location;
 import org.bukkit.World;
+import org.jetbrains.annotations.Nullable;
 
 public enum EntityType {
 
@@ -231,12 +232,43 @@ public enum EntityType {
         return typeId;
     }
 
+    // Titanium start
+    public static EntityType findType(String string) {
+        EntityType type = fromName(string);
+
+        if (type == null) {
+            type = fromEnumName(string);
+        }
+
+        if (type == null) {
+            int id;
+            try {
+                type = fromId(Integer.parseInt(string));
+            } catch (NumberFormatException ignored) {
+            }
+        }
+
+        return type;
+    }
+
+    @Nullable
+    public static EntityType fromEnumName(String name) {
+        for (EntityType type : values()) {
+            if (type.name().equalsIgnoreCase(name)) {
+                return type;
+            }
+        }
+        return null;
+    }
+    // Titanium end
+
     /**
      *
      * @param name the entity type's name
      * @return the matching entity type or null
      * @deprecated Magic value
      */
+    @Nullable
     @Deprecated
     public static EntityType fromName(String name) {
         if (name == null) {
@@ -251,6 +283,7 @@ public enum EntityType {
      * @return the matching entity type or null
      * @deprecated Magic value
      */
+    @Nullable
     @Deprecated
     public static EntityType fromId(int id) {
         if (id > Short.MAX_VALUE) {
-- 
2.36.0.windows.1

