From 253eec3e6a7fed72534817fef20786f85c7cb9cf Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 26 Oct 2022 21:17:20 +0200
Subject: [PATCH] Add LivingEntity#removePotionEffects


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 6ea3cbd8..d304391a 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -323,6 +323,13 @@ public interface LivingEntity extends Entity, Damageable, ProjectileSource {
      */
     public void removePotionEffect(PotionEffectType type);
 
+    // Titanium start
+    /**
+     * Removes every active potion effect
+     */
+    void removePotionEffects();
+    // Titanium end
+
     /**
      * Returns all currently active {@link PotionEffect}s on the living
      * entity.
-- 
2.38.1.windows.1

