From 8d3d01b771c45eb8c150803a19b49c8dc4fd7d8e Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 19 May 2022 08:00:35 +0200
Subject: [PATCH] Cache bukkit version


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 3a4b1424..5aa56804 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -8,22 +8,27 @@ import java.util.logging.Logger;
 import org.bukkit.Bukkit;
 
 public final class Versioning {
+
+    private static String bukkitVersion;
+
     public static String getBukkitVersion() {
-        String result = "Unknown-Version";
+        if (bukkitVersion != null) {
+            return bukkitVersion;
+        }
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/org.github.paperspigot/paperspigot-api/pom.properties");
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/net.titanvale/titanium-api/pom.properties");
         Properties properties = new Properties();
 
         if (stream != null) {
             try {
                 properties.load(stream);
-
-                result = properties.getProperty("version");
+                bukkitVersion = properties.getProperty("version");
+                return bukkitVersion;
             } catch (IOException ex) {
                 Logger.getLogger(Versioning.class.getName()).log(Level.SEVERE, "Could not get Bukkit version!", ex);
             }
         }
 
-        return result;
+        return "Unknown-Version";
     }
 }
-- 
2.36.0.windows.1

