From 6ea84abb4839fd34bcb937d48aa286224091621f Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 27 Apr 2022 22:21:30 +0200
Subject: [PATCH] Add reloadCount to CraftServer for compatibility


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 878ac62a..0723ee3b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -162,6 +162,8 @@ public final class CraftServer implements Server {
     private final UUID invalidUserUUID = UUID.nameUUIDFromBytes("InvalidUsername".getBytes(Charsets.UTF_8));
     private final List<CraftPlayer> playerView;
 
+    public final int reloadCount = 0; // Titanium - Add for compatibility since /reload is disabled
+
     private @Nullable Instant emptySince;
 
     private final class BooleanWrapper {
-- 
2.32.0.windows.2

