From 6966fa40b1a617dfe4387fcb2565857d5a84ca87 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Tue, 10 May 2022 21:30:17 +0200
Subject: [PATCH] Implement [Nacho-0019] Avoid lock every packet send


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index ba422aec..3cc2ed72 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -209,6 +209,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
     }
 
     private void m() {
+        if(this.i.isEmpty()) return; // [Nacho-0019] :: Avoid lock every packet send
         if (this.channel != null && this.channel.isOpen()) {
             this.j.readLock().lock();
 
-- 
2.36.0.windows.1

