From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 15 May 2022 17:44:45 +0200
Subject: [PATCH] Implement [Spigot-0138] Branchless NibbleArray by md5


diff --git a/src/main/java/net/minecraft/server/NibbleArray.java b/src/main/java/net/minecraft/server/NibbleArray.java
index 9306f97af94b6559dd28a3f942696ac4d3f6b746..69e2f0a9d445b71b79135bd9749e51b19517e98a 100644
--- a/src/main/java/net/minecraft/server/NibbleArray.java
+++ b/src/main/java/net/minecraft/server/NibbleArray.java
@@ -30,17 +30,22 @@ public class NibbleArray {
     public int a(int i) {
         int j = this.c(i);
 
-        return this.b(i) ? this.a[j] & 15 : this.a[j] >> 4 & 15;
+        return this.a[j] >> ((i & 1) << 2) & 15; // Spigot
     }
 
     public void a(int i, int j) {
         int k = this.c(i);
 
-        if (this.b(i)) {
-            this.a[k] = (byte) (this.a[k] & 240 | j & 15);
-        } else {
-            this.a[k] = (byte) (this.a[k] & 15 | (j & 15) << 4);
-        }
+//      if (this.b(i)) {
+//          this.a[k] = (byte) (this.a[k] & 240 | j & 15);
+//      } else {
+//          this.a[k] = (byte) (this.a[k] & 15 | (j & 15) << 4);
+//      }
+
+        // Spigot start
+        int shift = (i & 1) << 2;
+        this.a[k] = (byte) (this.a[k] & ~(15 << shift) | (j & 15) << shift);
+        // Spigot end
 
     }
 
