From 9a3a83a6f737d4d6247077f06a16d3af8748b04b Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 29 Apr 2022 20:43:50 +0200
Subject: [PATCH] Create some transformer


diff --git a/src/main/java/net/minecraft/titanium/config/BukkitConfig.java b/src/main/java/net/minecraft/titanium/config/BukkitConfig.java
index 51fabfb1..8500745c 100644
--- a/src/main/java/net/minecraft/titanium/config/BukkitConfig.java
+++ b/src/main/java/net/minecraft/titanium/config/BukkitConfig.java
@@ -9,6 +9,7 @@ import eu.okaeri.configs.annotation.Names;
 import org.bukkit.Warning;
 
 @Names(strategy = NameStrategy.HYPHEN_CASE, modifier = NameModifier.TO_LOWER_CASE)
+
 public class BukkitConfig extends OkaeriConfig{
 
     @Exclude
@@ -27,9 +28,6 @@ public class BukkitConfig extends OkaeriConfig{
         @Comment("On restart, plugins in this folder will be moved and updated.")
         public String updateFolder = "update";
 
-        // TODO write comment
-        public String permissionFile = "permissions.yml";
-
         @Comment("Enable the /timings command to track the execution time of events by plugins.")
         public boolean pluginProfiling = false;
 
@@ -41,7 +39,6 @@ public class BukkitConfig extends OkaeriConfig{
         public boolean queryPlugins = false;
 
         @Comment("Should a warning message be sent to console when a deprecated event is used by a plugin.")
-        // TODO Write serdes for this to use WarningState#value
         public Warning.WarningState deprecatedVerbose = Warning.WarningState.DEFAULT;
 
         @Comment("Message sent to clients before shutdown.")
diff --git a/src/main/java/net/minecraft/titanium/config/serdes/DecolorTransformer.java b/src/main/java/net/minecraft/titanium/config/serdes/DecolorTransformer.java
new file mode 100644
index 00000000..ef5cbe69
--- /dev/null
+++ b/src/main/java/net/minecraft/titanium/config/serdes/DecolorTransformer.java
@@ -0,0 +1,26 @@
+package net.minecraft.titanium.config.serdes;
+
+import eu.okaeri.configs.annotation.Exclude;
+import eu.okaeri.configs.schema.GenericsPair;
+import eu.okaeri.configs.serdes.ObjectTransformer;
+import eu.okaeri.configs.serdes.SerdesContext;
+import eu.okaeri.configs.serdes.SimpleObjectTransformer;
+import java.util.regex.Pattern;
+import org.bukkit.ChatColor;
+
+public class DecolorTransformer extends ObjectTransformer<String, String> {
+
+    @Exclude
+    private static final ObjectTransformer<String, String> TRANSFORMER = SimpleObjectTransformer.of(String.class, String.class, ChatColor::decolor);
+
+    @Override
+    public GenericsPair<String, String> getPair() {
+        return TRANSFORMER.getPair();
+    }
+
+    @Override
+    public String transform(String input, SerdesContext context) {
+        return TRANSFORMER.transform(input, context);
+    }
+
+}
diff --git a/src/main/java/net/minecraft/titanium/config/serdes/WarningStateTransformer.java b/src/main/java/net/minecraft/titanium/config/serdes/WarningStateTransformer.java
new file mode 100644
index 00000000..6f75bcca
--- /dev/null
+++ b/src/main/java/net/minecraft/titanium/config/serdes/WarningStateTransformer.java
@@ -0,0 +1,25 @@
+package net.minecraft.titanium.config.serdes;
+
+import eu.okaeri.configs.schema.GenericsPair;
+import eu.okaeri.configs.serdes.BidirectionalTransformer;
+import eu.okaeri.configs.serdes.SerdesContext;
+import org.bukkit.Warning;
+
+public class WarningStateTransformer extends BidirectionalTransformer<String, Warning.WarningState> {
+
+    @Override
+    public GenericsPair<String, Warning.WarningState> getPair() {
+        return this.genericsPair(String.class, Warning.WarningState.class);
+    }
+
+    @Override
+    public Warning.WarningState leftToRight(String data, SerdesContext serdesContext) {
+        return Warning.WarningState.value(data);
+    }
+
+    @Override
+    public String rightToLeft(Warning.WarningState data, SerdesContext serdesContext) {
+        return data.name().toLowerCase();
+    }
+
+}
-- 
2.32.0.windows.2

