From 123de62653252a65e04ef2e8d69e365d237f86a4 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 19 May 2022 19:22:53 +0200
Subject: [PATCH] Implement [Paper-0044] Use UserCache for player heads


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index 55bbad76b..80224ab28 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -11,6 +11,7 @@ import net.minecraft.server.EntityPlayer;
 import net.minecraft.server.MinecraftServer;
 // PaperSpigot end
 
+import net.minecraft.server.TileEntitySkull;
 import org.bukkit.Material;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
 import org.bukkit.craftbukkit.inventory.CraftMetaItem.SerializableMeta;
@@ -139,10 +140,13 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
         } else {
             // PaperSpigot start - Check usercache if the player is online
             EntityPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(name);
-            profile = player != null ? player.getProfile() : new GameProfile(null, name);
+            if (profile == null && player != null) profile = player.getProfile();
             // PaperSpigot end
         }
 
+        if (profile == null) profile = TileEntitySkull.skinCache.getIfPresent(name.toLowerCase(java.util.Locale.ROOT)); // Paper
+        if (profile == null) profile = new GameProfile(null, name);
+
         return true;
     }
 
-- 
2.38.1.windows.1

