From 274018107ab7a41aa91d71b26de2aa11dc582bb1 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sat, 30 Apr 2022 16:00:42 +0200
Subject: [PATCH] Get rid of illegal reflective access


diff --git a/src/main/java/net/titanium/config/PaperConfig.java b/src/main/java/net/titanium/config/PaperConfig.java
index 6a3de4c9..4c9ce387 100644
--- a/src/main/java/net/titanium/config/PaperConfig.java
+++ b/src/main/java/net/titanium/config/PaperConfig.java
@@ -122,21 +122,10 @@ public class PaperConfig extends TitaniumConfigSection {
 
         @Override
         public void loadProcessedProperties() {
-            Field maxStack;
-
             try {
-                maxStack = Material.class.getDeclaredField("maxStack");
+                Field maxStack = Material.class.getDeclaredField("maxStack");
                 maxStack.setAccessible(true);
 
-                Field modifiers = Field.class.getDeclaredField("modifiers");
-                modifiers.setAccessible(true);
-                modifiers.setInt(maxStack, maxStack.getModifiers() & ~Modifier.FINAL);
-            } catch (Exception e) {
-                e.printStackTrace();
-                return;
-            }
-
-            try {
                 if (lava) {
                     maxStack.set(Material.LAVA_BUCKET, Material.BUCKET.getMaxStackSize());
                     Items.LAVA_BUCKET.c(Material.BUCKET.getMaxStackSize());
@@ -151,8 +140,8 @@ public class PaperConfig extends TitaniumConfigSection {
                     maxStack.set(Material.MILK_BUCKET, Material.BUCKET.getMaxStackSize());
                     Items.MILK_BUCKET.c(Material.BUCKET.getMaxStackSize());
                 }
-            } catch (Exception e) {
-                e.printStackTrace();
+            } catch (Exception ex) {
+                ex.printStackTrace();
             }
         }
 
-- 
2.32.0.windows.2

