From 37ce456e9c11ee91c3a02da9411f5a25bfd70b5c Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 8 May 2022 23:43:33 +0200
Subject: [PATCH] Implement [FlamePaper-0030] Adaptive chunk GC


diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 9d6216e5b..6da6e5426 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -105,6 +105,9 @@ public class TitaniumConfig extends OkaeriConfig {
         // TODO comment this
         public boolean disableUnloadedChunksMovement = false;
 
+        // TODO comment this
+        public boolean adaptiveChunkGC = false;
+
         @Comment("Configuration of chunk I/O (loading) since chunk loading is asynchronous.")
         @Comment("If you have CPU with more threads changing these values can have good impact on performance.")
         public IO io = new IO();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index c85049909..ae983a5fe 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -1554,7 +1554,14 @@ public class CraftWorld implements World {
     public void processChunkGC() {
         chunkGCTickCount++;
 
-        if (chunkLoadCount >= server.chunkGCLoadThresh && server.chunkGCLoadThresh > 0) {
+        // FlamePaper start - Adaptative chunk GC
+        int playerCount = getPlayers().size();
+        int viewDistance = getHandle().getServer().getViewDistance();
+        int viewChunks = ((viewDistance * 2) + 1);
+        int chunkGCLoadThreshold = TitaniumConfig.get().chunk.adaptiveChunkGC ? (world.keepSpawnInMemory ? 256 : 0) + playerCount * (viewChunks * viewChunks) : server.chunkGCLoadThresh;
+
+        if (chunkLoadCount > chunkGCLoadThreshold && chunkGCLoadThreshold > 0) {
+            // FlamePaper end - Adaptative chunk GC
             chunkLoadCount = 0;
         } else if (chunkGCTickCount >= server.chunkGCPeriod && server.chunkGCPeriod > 0) {
             chunkGCTickCount = 0;
-- 
2.36.0.windows.1

