From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 19 May 2022 23:28:31 +0200
Subject: [PATCH] Implement [Paper-0344] [MC-111480] Start Entity ID's at 1


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 9580599f175c68002d82076664d7db7abfa8d7bf..d7ad9edcdef05de8d5d95acd8863cc9cc4f2dfd5 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -48,7 +48,7 @@ public abstract class Entity implements ICommandListener {
     // CraftBukikt end
 
     private static final AxisAlignedBB a = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.0D, 0.0D, 0.0D);
-    private static int entityCount;
+    private static int entityCount = 1; // Paper - [MC-111480] Start Entity ID's at 1
     private int id;
     public double j;
     public boolean k;
