From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 19 May 2022 23:17:47 +0200
Subject: [PATCH] Implement [Nacho-0023] Optimize EntityTracker for the chunk
 updater


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index b788b398fa2e2ca996331e9a24a714ba204a4daa..b89849a101a5f58d2b17135b8bf208d0f2ad66b8 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -188,6 +188,10 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.playerConnection.sendPacket(new PacketPlayOutCombatEvent(this.bs(), PacketPlayOutCombatEvent.EnumCombatEventType.END_COMBAT));
     }
 
+    private long chunkToLong(int chunkX, int chunkZ) {
+        return ((long) chunkX << 32L) + chunkZ - -2147483648L;
+    }
+
     public void t_() {
         // CraftBukkit start
         if (this.joining) {
@@ -230,7 +234,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             Iterator iterator1 = this.chunkCoordIntPairQueue.iterator();
             ArrayList<TileEntity> tileEntities = Lists.newArrayList();
 
-            Chunk chunk;
+            Chunk chunk = null;
 
             while (iterator1.hasNext() && chunkList.size() < this.world.spigotConfigTitanium.maxBulkChunks) { // Spigot
                 ChunkCoordIntPair chunkcoordintpair = (ChunkCoordIntPair) iterator1.next();
@@ -262,11 +266,33 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
                     this.a(tileentity);
                 }
 
+                // Nacho - if there are a lot of entities, we end up scanning the WHOLE list of entities multiple times
+                // Which isn't the best if we have 100 players doing that
+                // So instead of updating all entities by chunk, we update all entities at once with a hashset of chunks
+                // This means we don't have to pass over the list x chunks
+                // o(chunk * entityList) => o(entitylist)
+                /*
                 Iterator<Chunk> chunkIterator = chunkList.iterator();
                 while (chunkIterator.hasNext()) {
-                    chunk = chunkIterator.next();
+                    chunk = (Chunk) chunkIterator.next();
                     this.u().getTracker().a(this, chunk);
                 }
+                */
+                // Nacho - end
+
+                LongOpenHashSet chunkPosSet = new LongOpenHashSet(chunkList.size());
+                for (Chunk newChunk : chunkList) {
+                    chunkPosSet.add(this.chunkToLong(newChunk.locX, newChunk.locZ));
+                }
+
+                Iterator<EntityTrackerEntry> trackerEntryIterator = this.u().getTracker().getEntityTrackerEntries();
+                while (trackerEntryIterator.hasNext()) {
+                    EntityTrackerEntry entitytrackerentry = trackerEntryIterator.next();
+
+                    if (entitytrackerentry.tracker != this && chunkPosSet.contains(this.chunkToLong(entitytrackerentry.tracker.ae, entitytrackerentry.tracker.ag))) {
+                        entitytrackerentry.updatePlayer(this);
+                    }
+                }
             }
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityTracker.java b/src/main/java/net/minecraft/server/EntityTracker.java
index e24644c425edf6a9aed5081a8de66d73b6563717..4668a1fca0a7c7016264c56f54ac8464768751c4 100644
--- a/src/main/java/net/minecraft/server/EntityTracker.java
+++ b/src/main/java/net/minecraft/server/EntityTracker.java
@@ -239,6 +239,10 @@ public class EntityTracker {
 
     }
 
+    public Iterator<EntityTrackerEntry> getEntityTrackerEntries() {
+        return this.c.iterator();
+    }
+
     public void a(EntityPlayer entityplayer, Chunk chunk) {
         Iterator iterator = this.c.iterator();
 
