From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lew <skycubesv@gmail.com>
Date: Thu, 19 May 2022 23:13:15 +0200
Subject: [PATCH] Implement [InsanePaper-269] Cache Chunk Coordinations


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 6cc496b4850d0865e0818ddfeb9be99a789f0eab..7d0aa864441060c0ba755d79462ccb0e6fb8d27f 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -34,6 +34,7 @@ public class Chunk {
     public final int[] heightMap;
     public final int locX;
     public final int locZ;
+    private final ChunkCoordIntPair chunkCoords; // InsanePaper
     private boolean k;
     public final Map<BlockPosition, TileEntity> tileEntities;
     public final List<Entity>[] entitySlices; // Spigot
@@ -154,6 +155,7 @@ public class Chunk {
         if (!(this instanceof EmptyChunk)) {
             this.bukkitChunk = new org.bukkit.craftbukkit.CraftChunk(this);
         }
+        this.chunkCoords = new ChunkCoordIntPair(i, j); // InsanePaper
     }
 
     public org.bukkit.Chunk bukkitChunk;
@@ -1204,7 +1206,7 @@ public class Chunk {
     }
 
     public ChunkCoordIntPair j() {
-        return new ChunkCoordIntPair(this.locX, this.locZ);
+        return this.chunkCoords; // InsanePaper
     }
 
     public boolean c(int i, int j) {
