From dee9add04f6a60489d9f20b9e930a1c57135c0e6 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 11 May 2022 08:11:55 +0200
Subject: [PATCH] Implement [IonSpigot-0005] Fix Sand Unloading


diff --git a/src/main/java/net/minecraft/server/BlockFalling.java b/src/main/java/net/minecraft/server/BlockFalling.java
index a9b887b29..56da1990f 100644
--- a/src/main/java/net/minecraft/server/BlockFalling.java
+++ b/src/main/java/net/minecraft/server/BlockFalling.java
@@ -36,7 +36,7 @@ public class BlockFalling extends Block {
         if (canFall(world, blockposition.down()) && blockposition.getY() >= 0) {
             byte b0 = 32;
 
-            if (!BlockFalling.instaFall && world.areChunksLoadedBetween(blockposition.a(-b0, -b0, -b0), blockposition.a(b0, b0, b0))) {
+            if (world.ionConfigTitanium.sand.fixUnloading || !BlockFalling.instaFall && world.areChunksLoadedBetween(blockposition.a(-b0, -b0, -b0), blockposition.a(b0, b0, b0))) {
                 if (!world.isClientSide) {
                     // PaperSpigot start - Add FallingBlock source location API
                     org.bukkit.Location loc = new org.bukkit.Location(world.getWorld(), (double) ((float) blockposition.getX() + 0.5F), (double) blockposition.getY(), (double) ((float) blockposition.getZ() + 0.5F));
diff --git a/src/main/java/net/titanium/config/IonWorldConfig.java b/src/main/java/net/titanium/config/IonWorldConfig.java
index 968c53e60..d9118a626 100644
--- a/src/main/java/net/titanium/config/IonWorldConfig.java
+++ b/src/main/java/net/titanium/config/IonWorldConfig.java
@@ -7,6 +7,15 @@ import eu.okaeri.configs.annotation.Names;
 @Names(strategy = NameStrategy.HYPHEN_CASE, modifier = NameModifier.TO_LOWER_CASE)
 public class IonWorldConfig extends TitaniumConfigSection {
 
+    public Sand sand = new Sand();
+
+    public static class Sand extends TitaniumConfigSection {
+
+        //TODO comment this
+        public boolean fixUnloading = false;
+
+    }
+
     public Movement movement = new Movement();
 
     public static class Movement extends TitaniumConfigSection {
-- 
2.36.0.windows.1

