From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 22 May 2022 22:36:59 +0200
Subject: [PATCH] Implement [FlamePaper-0017] Pearl through blocks


diff --git a/src/main/java/net/minecraft/server/EntityEnderPearl.java b/src/main/java/net/minecraft/server/EntityEnderPearl.java
index 80f2c116629321d0496ebb09cf183417769207c4..2ed7f2790217b5596005b80a8ed5e6ff133bc413 100644
--- a/src/main/java/net/minecraft/server/EntityEnderPearl.java
+++ b/src/main/java/net/minecraft/server/EntityEnderPearl.java
@@ -1,6 +1,7 @@
 package net.minecraft.server;
 
 // CraftBukkit start
+import net.titanium.config.TitaniumConfig;
 import org.bukkit.Bukkit;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.player.PlayerTeleportEvent;
@@ -38,6 +39,31 @@ public class EntityEnderPearl extends EntityProjectile {
         }
         // PaperSpigot end
 
+        // FlamePaper start
+        BlockPosition blockPosition = movingobjectposition.a();
+
+        if (blockPosition != null) {
+            IBlockData blockData = world.getType(blockPosition);
+            Block block = blockData.getBlock();
+            TitaniumConfig.PearlPassthrough pearlPassthrough = TitaniumConfig.get().pearlPassthrough;
+            boolean collides =
+                    pearlPassthrough.strings && block == Blocks.TRIPWIRE
+                            || pearlPassthrough.cobwebs && block == Blocks.WEB
+                            || pearlPassthrough.fenceGates && block == Blocks.FENCE_GATE && blockData.get(BlockFenceGate.OPEN)
+                            || pearlPassthrough.slabs && (block == Blocks.STONE_SLAB || block == Blocks.WOODEN_SLAB || block == Blocks.STONE_SLAB2)
+                            || pearlPassthrough.stairs && (block == Blocks.OAK_STAIRS || block == Blocks.STONE_STAIRS || block == Blocks.BRICK_STAIRS || block == Blocks.STONE_BRICK_STAIRS || block == Blocks.NETHER_BRICK_STAIRS || block == Blocks.SANDSTONE_STAIRS || block == Blocks.SPRUCE_STAIRS || block == Blocks.BIRCH_STAIRS || block == Blocks.JUNGLE_STAIRS || block == Blocks.ACACIA_STAIRS || block == Blocks.DARK_OAK_STAIRS || block == Blocks.QUARTZ_STAIRS)
+                            || pearlPassthrough.chests && (block == Blocks.CHEST || block == Blocks.TRAPPED_CHEST)
+                            || pearlPassthrough.enderChests && block == Blocks.ENDER_CHEST
+                            || pearlPassthrough.enchantingTables && block == Blocks.ENCHANTING_TABLE
+                            || pearlPassthrough.anvils && block == Blocks.ANVIL
+                            || pearlPassthrough.endPortalFrames && block == Blocks.END_PORTAL_FRAME;
+
+            if (collides) {
+                return;
+            }
+        }
+        // FlamePaper end
+
         for (int i = 0; i < 32; ++i) {
             this.world.addParticle(EnumParticle.PORTAL, this.locX, this.locY + this.random.nextDouble() * 2.0D, this.locZ, this.random.nextGaussian(), 0.0D, this.random.nextGaussian(), new int[0]);
         }
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index e5954d898840396a0292229624522987f98e2edd..73fc33e5af90f64d1647873f410682db7b76a40b 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -50,6 +50,24 @@ public class TitaniumConfig extends TitaniumConfigSection {
     @Comment("Configuration from ion.yml")
     public IonConfig ion = new IonConfig();
 
+    @Comment("Blocks through which enderpearls will pass through (AKA Taliban Pearls)")
+    public PearlPassthrough pearlPassthrough = new PearlPassthrough();
+
+    public static class PearlPassthrough extends TitaniumConfigSection {
+
+        public boolean slabs = false;
+        public boolean stairs = false;
+        public boolean fenceGates = false;
+        public boolean chests = false;
+        public boolean enderChests = false;
+        public boolean enchantingTables = false;
+        public boolean anvils = false;
+        public boolean endPortalFrames = false;
+        public boolean strings = false;
+        public boolean cobwebs = false;
+
+    }
+
     public Chunk chunk = new Chunk();
 
     public static class Chunk extends TitaniumConfigSection {
