From 253095d8366713706c1ac8859b8002e5a867490c Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 11 May 2022 16:19:25 +0200
Subject: [PATCH] Implement [Paper-0072] Fix Furnace cook time bug when lagging
 by Aikar


diff --git a/src/main/java/net/minecraft/server/TileEntityFurnace.java b/src/main/java/net/minecraft/server/TileEntityFurnace.java
index ffc902c94..14fc635da 100644
--- a/src/main/java/net/minecraft/server/TileEntityFurnace.java
+++ b/src/main/java/net/minecraft/server/TileEntityFurnace.java
@@ -100,7 +100,7 @@ public class TileEntityFurnace extends TileEntityContainer implements IUpdatePla
 
         if (i == 0 && !flag) {
             this.cookTimeTotal = this.a(itemstack);
-            this.cookTime = 0;
+            this.cookTime -= this.cookTimeTotal; // Paper
             this.update();
         }
 
-- 
2.36.0.windows.1

