From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 22 May 2022 23:11:39 +0200
Subject: [PATCH] Implement [Nacho-0025] Disable random tickSpeed being
 modified and make it configurable


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index f042671c3cfbdb82be717b9b5e39ea0de0bf5e47..275efc0d064d34968ebc2aeadb45725ba1d2fd82 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -8,6 +8,7 @@ import net.titanium.config.PaperWorldConfig;
 import net.titanium.config.SpigotWorldConfig;
 import net.titanium.config.TacoSpigotWorldConfig;
 import net.titanium.config.TitaniumConfig;
+import net.titanium.config.TitaniumWorldConfig;
 import org.bukkit.Bukkit;
 import org.bukkit.block.BlockState;
 import org.bukkit.craftbukkit.CraftServer;
@@ -169,6 +170,7 @@ public abstract class World implements IBlockAccess {
     public final org.github.paperspigot.PaperSpigotWorldConfig paperSpigotConfig; // PaperSpigot
      */
 
+    public final TitaniumWorldConfig titaniumConfig;
     public final SpigotWorldConfig spigotConfigTitanium;
     public final PaperWorldConfig paperConfigTitanium;
     public final TacoSpigotWorldConfig tacoConfigTitanium;
@@ -201,6 +203,7 @@ public abstract class World implements IBlockAccess {
         this.paperSpigotConfig = new org.github.paperspigot.PaperSpigotWorldConfig( worlddata.getName() ); // PaperSpigot
          */
         TitaniumConfig serverConfig = TitaniumConfig.get();
+        this.titaniumConfig = serverConfig.worldSettings;
         this.spigotConfigTitanium = serverConfig.spigot.worldSettings;
         this.paperConfigTitanium = serverConfig.paper.worldSettings;
         this.tacoConfigTitanium = serverConfig.taco.worldSettings;
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 8fa99c50f679a275f00cd1e9a7fa99c8f91274b8..1b7b28f9897f04141bdbaa86180fc923dcb962cf 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -479,7 +479,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
                 this.methodProfiler.c("tickBlocks");
                 timings.chunkTicksBlocks.startTiming(); // Spigot
-                i1 = this.getGameRules().c("randomTickSpeed");
+                i1 = titaniumConfig.disabledGamerules.randomTickSpeed ? 3 : this.getGameRules().c("randomTickSpeed");
                 if (i1 > 0) {
                     ChunkSection[] achunksection = chunk.getSections();
                     int j1 = achunksection.length;
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 5d689bc03763a1627c17597b70d533899f0da714..56f81724ff8127d525558ed7d1dd5f8ef227e50f 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -106,6 +106,8 @@ public class TitaniumConfig extends TitaniumConfigSection {
     @Comment("Enable this option only if some of your plugins are not compatible with the new configs.")
     public boolean enableLegacyConfigFiles = false;
 
+    public TitaniumWorldConfig worldSettings = new TitaniumWorldConfig();
+
     public Commands commands = new Commands();
 
     public static class Commands extends TitaniumConfigSection {
diff --git a/src/main/java/net/titanium/config/TitaniumWorldConfig.java b/src/main/java/net/titanium/config/TitaniumWorldConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..026aeb4004e8a4113988c6294557238ecde97eec
--- /dev/null
+++ b/src/main/java/net/titanium/config/TitaniumWorldConfig.java
@@ -0,0 +1,24 @@
+package net.titanium.config;
+
+import eu.okaeri.configs.annotation.Comment;
+import eu.okaeri.configs.annotation.NameModifier;
+import eu.okaeri.configs.annotation.NameStrategy;
+import eu.okaeri.configs.annotation.Names;
+
+@Names(strategy = NameStrategy.HYPHEN_CASE, modifier = NameModifier.TO_LOWER_CASE)
+public class TitaniumWorldConfig extends TitaniumConfigSection {
+
+    public DisabledGamerules disabledGamerules = new DisabledGamerules();
+
+    public static class DisabledGamerules extends TitaniumConfigSection {
+
+        @Comment("Whether gamerule randomTickSpeed should be enabled or not.")
+        @Comment("Plants growth, leaf decay, fire spread etc. depends on this value.")
+        @Comment("Normally we have to cast String to Integer every tick which is costly.")
+        @Comment("If this option is enabled, this gamerule is disabled and we use default vanilla value which is 3.")
+        @Comment("If you know you won't use this gamerule you should enable this option.")
+        public boolean randomTickSpeed = false;
+
+    }
+
+}
