From 47e5f409e5eed27aef9b7cbc3aa216f7217d9888 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 22 May 2022 23:06:55 +0200
Subject: [PATCH] Implement [Taco-0025] Fix east/west cannoning


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 1b59d7e9..cb1cf885 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -583,16 +583,37 @@ public abstract class Entity implements ICommandListener {
             this.a(this.getBoundingBox().c(0.0D, d1, 0.0D));
             boolean flag1 = this.onGround || d7 != d1 && d7 < 0.0D;
 
-            if(axisScan) list = this.world.getCubes(this, this.getBoundingBox().a(d0, 0, 0)); // TacoSpigot - get x axis blocks
-
             AxisAlignedBB axisalignedbb2;
             Iterator iterator1;
 
+            if (this.world.tacoConfigTitanium.fixEastWestCannoning && Math.abs(d0) > Math.abs(d2)) { //TacoSpigot - fix east/west cannoning by calculating the z movement before x if the x velocity is greater
+                if (axisScan) {
+                    list = this.world.getCubes(this, this.getBoundingBox().a(0, 0, d2)); // TacoSpigot - get z axis blocks
+                }
+
+                for (iterator1 = list.iterator(); iterator1.hasNext(); d2 = axisalignedbb2.c(this.getBoundingBox(), d2)) {
+                    axisalignedbb2 = (AxisAlignedBB) iterator1.next();
+                }
+
+                this.a(this.getBoundingBox().c(0.0D, 0.0D, d2));
+
+                if (axisScan) {
+                    list = this.world.getCubes(this, this.getBoundingBox().a(d0, 0, 0)); // TacoSpigot - get x axis blocks
+                }
+
+                for (iterator1 = list.iterator(); iterator1.hasNext(); d0 = axisalignedbb2.a(this.getBoundingBox(), d0)) {
+                    axisalignedbb2 = (AxisAlignedBB) iterator1.next();
+                }
+
+                this.a(this.getBoundingBox().c(d0, 0.0D, 0.0D));
+
+            } else {
             for (iterator1 = list.iterator(); iterator1.hasNext(); d0 = axisalignedbb2.a(this.getBoundingBox(), d0)) {
                 axisalignedbb2 = (AxisAlignedBB) iterator1.next();
             }
 
             this.a(this.getBoundingBox().c(d0, 0.0D, 0.0D));
+            }
 
             if(axisScan) list = this.world.getCubes(this, this.getBoundingBox().a(0, 0, d2)); // TacoSpigot - get z axis blocks
 
diff --git a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
index 3a356dd3..c8eb2773 100644
--- a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
@@ -7,6 +7,9 @@ import eu.okaeri.configs.annotation.Names;
 @Names(strategy = NameStrategy.HYPHEN_CASE, modifier = NameModifier.TO_LOWER_CASE)
 public class TacoSpigotWorldConfig extends TitaniumConfigSection {
 
+    //TODO comment this (See TacoSpigot-0025)
+    public boolean fixEastWestCannoning = false;
+
     public Tnt tnt = new Tnt();
 
     public static class Tnt extends TitaniumConfigSection {
-- 
2.36.0.windows.1

