From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 22 May 2022 23:50:59 +0200
Subject: [PATCH] Implement [PandaSpigot-0008] Remove hacky stackable buckets
 code


diff --git a/src/main/java/net/titanium/config/PaperConfig.java b/src/main/java/net/titanium/config/PaperConfig.java
index 7cd09808bec3db8fc87c3e75e61b01b08ca86220..07115a5dc9c3b052b559b84c68846f8033fa246a 100644
--- a/src/main/java/net/titanium/config/PaperConfig.java
+++ b/src/main/java/net/titanium/config/PaperConfig.java
@@ -84,29 +84,24 @@ public class PaperConfig extends TitaniumConfigSection {
 
         @Override
         public void loadProcessedProperties() {
-            try {
-                Field maxStack = Material.class.getDeclaredField("maxStack");
-                maxStack.setAccessible(true);
-
-                if (lava) {
-                    maxStack.set(Material.LAVA_BUCKET, Material.BUCKET.getMaxStackSize());
-                    Items.LAVA_BUCKET.c(Material.BUCKET.getMaxStackSize());
-                }
-
-                if (water) {
-                    maxStack.set(Material.WATER_BUCKET, Material.BUCKET.getMaxStackSize());
-                    Items.WATER_BUCKET.c(Material.BUCKET.getMaxStackSize());
-                }
-
-                if (milk) {
-                    maxStack.set(Material.MILK_BUCKET, Material.BUCKET.getMaxStackSize());
-                    Items.MILK_BUCKET.c(Material.BUCKET.getMaxStackSize());
-                }
-            } catch (Exception ex) {
-                ex.printStackTrace();
+            // PandaSpigot start - Remove hacky stackable buckets code that used reflection
+            int size = Material.BUCKET.getMaxStackSize();
+            if (lava) {
+                Material.LAVA_BUCKET.setMaxStack(size);
+                Items.LAVA_BUCKET.c(size);
+            }
+
+            if (water) {
+                Material.WATER_BUCKET.setMaxStack(size);
+                Items.WATER_BUCKET.c(size);
             }
-        }
 
+            if (milk) {
+                Material.MILK_BUCKET.setMaxStack(size);
+                Items.MILK_BUCKET.c(size);
+            }
+            // PandaSpigot end
+        }
     }
 
     @Comment("Send a warning to console when players are moving at excessive speeds.")
