From e41cb907582413a1ad24d656fbdbcd3594ce6a18 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 23 May 2022 19:47:37 +0200
Subject: [PATCH] Fix LegacyConfigHelper


diff --git a/src/main/java/net/titanium/config/LegacyConfigHelper.java b/src/main/java/net/titanium/config/LegacyConfigHelper.java
index 4d197bb7..b4a4615d 100644
--- a/src/main/java/net/titanium/config/LegacyConfigHelper.java
+++ b/src/main/java/net/titanium/config/LegacyConfigHelper.java
@@ -5,16 +5,10 @@ import java.io.File;
 import java.io.IOException;
 import java.time.Duration;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
-import java.util.stream.Collectors;
-import net.minecraft.server.MinecraftServer;
-import net.minecraft.server.WorldServer;
 import org.bukkit.Bukkit;
-import org.bukkit.Warning;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
 
@@ -22,7 +16,7 @@ public class LegacyConfigHelper {
 
     private static final File LEGACY_CONFIG_DIRECTORY = new File("legacy_configs");
 
-    private final TitaniumConfig titaniumConfig;
+    private final TitaniumConfigSection titaniumConfigSection;
 
     private final String section;
     private final File file;
@@ -30,8 +24,8 @@ public class LegacyConfigHelper {
 
     private final int configVersion;
 
-    public LegacyConfigHelper(TitaniumConfig titaniumConfig, String section, int configVersion) {
-        this.titaniumConfig = titaniumConfig;
+    public LegacyConfigHelper(TitaniumConfigSection titaniumConfigSection, String section, int configVersion) {
+        this.titaniumConfigSection = titaniumConfigSection;
         this.section = section;
 
         if (!LEGACY_CONFIG_DIRECTORY.exists()) {
@@ -40,13 +34,15 @@ public class LegacyConfigHelper {
 
         this.file = new File(LEGACY_CONFIG_DIRECTORY, section + ".yml");
 
-        if (!this.file.exists()) {
-            try {
-                this.file.createNewFile();
-            } catch (IOException ex) {
-                Bukkit.getLogger().severe("Could not create legacy config file of " + section);
-                ex.printStackTrace();
-            }
+        if (file.exists() && file.canWrite()) {
+            this.file.delete();
+        }
+
+        try {
+            this.file.createNewFile();
+        } catch (IOException ex) {
+            Bukkit.getLogger().severe("Could not create legacy config file of " + section);
+            ex.printStackTrace();
         }
 
         try {
@@ -59,8 +55,8 @@ public class LegacyConfigHelper {
         this.configVersion = configVersion;
     }
 
-    public LegacyConfigHelper(TitaniumConfig titaniumConfig, String section) {
-        this(titaniumConfig, section, -1);
+    public LegacyConfigHelper(TitaniumConfigSection titaniumConfigSection, String section) {
+        this(titaniumConfigSection, section, -1);
     }
 
     public String getSection() {
@@ -76,7 +72,7 @@ public class LegacyConfigHelper {
     }
 
     public void saveSectionToFile() {
-        this.mapData(null, null).forEach(this.configuration::set);
+        this.mapData(this.titaniumConfigSection, null).forEach(this.configuration::set);
 
         if (this.configVersion != -1) {
             this.configuration.set("config-version", this.configVersion);
@@ -93,27 +89,22 @@ public class LegacyConfigHelper {
     private Map<String, Object> mapData(OkaeriConfig parentConfig, String parentSection) {
         Map<String, Object> dataMap = new HashMap<>();
 
-        if (parentConfig == null) {
-            Object raw = this.titaniumConfig.get(this.section);
-            if (!(raw instanceof OkaeriConfig)) {
-                return dataMap;
-            }
-            parentConfig = (OkaeriConfig) raw;
-        }
-
         for (String key : parentConfig.getConfigurer().getAllKeys()) {
             Object data = parentConfig.get(key);
 
-            if (key.equalsIgnoreCase("world-settings")) {
-                this.mapData((OkaeriConfig) data, key).forEach((keyData, valueData) -> dataMap.put("world-settings.default." + keyData, valueData));
-                continue;
-            }
+            String dataKey = parentSection == null ? key : parentSection + "." + key;
 
             if (data instanceof OkaeriConfig) {
-                dataMap.putAll(this.mapData((OkaeriConfig) data, key));
+                if (key.equalsIgnoreCase("world-settings")) {
+                    this.mapData((OkaeriConfig) data, null)
+                            .forEach((mappedKey, mappedValue) -> dataMap.put(dataKey + ".default." + mappedKey, mappedValue));
+                    continue;
+                }
+
+                dataMap.putAll(this.mapData((OkaeriConfig) data, dataKey));
                 continue;
             }
-            dataMap.put((parentSection == null ? "" : parentSection + ".") + key, this.convertType(data));
+            dataMap.put(dataKey, this.convertType(data));
         }
 
         return dataMap;
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 05ab2e8e..b7843b6b 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -153,27 +153,27 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
         if (enableLegacyConfigFiles) {
             if (this.bukkitConfigHelper == null) {
-                this.bukkitConfigHelper = new LegacyConfigHelper(this, "bukkit");
+                this.bukkitConfigHelper = new LegacyConfigHelper(this.bukkit, "bukkit");
             }
             this.bukkitConfigHelper.saveSectionToFile();
 
             if (this.spigotConfigHelper == null) {
-                this.spigotConfigHelper = new LegacyConfigHelper(this, "spigot", 8);
+                this.spigotConfigHelper = new LegacyConfigHelper(this.spigot, "spigot", 8);
             }
             this.spigotConfigHelper.saveSectionToFile();
 
             if (this.paperConfigHelper == null) {
-                this.paperConfigHelper = new LegacyConfigHelper(this, "paper", 9);
+                this.paperConfigHelper = new LegacyConfigHelper(this.paper, "paper", 9);
             }
             this.paperConfigHelper.saveSectionToFile();
 
             if (this.tacoSpigotConfigHelper == null) {
-                this.tacoSpigotConfigHelper = new LegacyConfigHelper(this, "taco", 1);
+                this.tacoSpigotConfigHelper = new LegacyConfigHelper(this.taco, "taco", 1);
             }
             this.tacoSpigotConfigHelper.saveSectionToFile();
 
             if (this.ionConfigHelper == null) {
-                this.ionConfigHelper = new LegacyConfigHelper(this, "ion", 1);
+                this.ionConfigHelper = new LegacyConfigHelper(this.ion, "ion", 1);
             }
             this.ionConfigHelper.saveSectionToFile();
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index afbb710e..94a94588 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -33,6 +33,7 @@ import javax.imageio.ImageIO;
 import net.minecraft.server.*;
 
 import net.titanium.config.BukkitConfig;
+import net.titanium.config.LegacyConfigHelper;
 import net.titanium.config.TitaniumConfig;
 import net.titanium.config.serdes.DecolorTransformer;
 import net.titanium.config.serdes.WarningStateTransformer;
@@ -1845,34 +1846,54 @@ public final class CraftServer implements Server {
         @Override
         public YamlConfiguration getBukkitConfig()
         {
-            return serverConfig.getBukkitConfigHelper().getConfiguration();
+            LegacyConfigHelper configHelper = serverConfig.getBukkitConfigHelper();
+             if (configHelper == null) {
+                 throw new IllegalStateException("Legacy config files are disable. Set `enable-legacy-config-files` in titanium.yml to true to use them.");
+             }
+            return configHelper.getConfiguration();
         }
 
         @Deprecated // Titanium
         @Override
         public YamlConfiguration getSpigotConfig()
         {
-            return serverConfig.getSpigotConfigHelper().getConfiguration();
+            LegacyConfigHelper configHelper = serverConfig.getSpigotConfigHelper();
+            if (configHelper == null) {
+                throw new IllegalStateException("Legacy config files are disable. Set `enable-legacy-config-files` in titanium.yml to true to use them.");
+            }
+            return configHelper.getConfiguration();
         }
 
         @Deprecated // Titanium
         @Override
         public YamlConfiguration getPaperSpigotConfig()
         {
-            return serverConfig.getPaperConfigHelper().getConfiguration();
+            LegacyConfigHelper configHelper = serverConfig.getPaperConfigHelper();
+            if (configHelper == null) {
+                throw new IllegalStateException("Legacy config files are disable. Set `enable-legacy-config-files` in titanium.yml to true to use them.");
+            }
+            return configHelper.getConfiguration();
         }
 
         @Deprecated // Titanium
         @Override
         public YamlConfiguration getTacoSpigotConfig() {
-            return serverConfig.getTacoSpigotConfigHelper().getConfiguration();
+            LegacyConfigHelper configHelper = serverConfig.getTacoSpigotConfigHelper();
+            if (configHelper == null) {
+                throw new IllegalStateException("Legacy config files are disable. Set `enable-legacy-config-files` in titanium.yml to true to use them.");
+            }
+            return configHelper.getConfiguration();
         }
 
         // IonSpigot start
         @Deprecated // Titanium
         @Override
         public YamlConfiguration getIonConfig() {
-            return serverConfig.getIonConfigHelper().getConfiguration();
+            LegacyConfigHelper configHelper = serverConfig.getIonConfigHelper();
+            if (configHelper == null) {
+                throw new IllegalStateException("Legacy config files are disable. Set `enable-legacy-config-files` in titanium.yml to true to use them.");
+            }
+            return configHelper.getConfiguration();
         }
         // IonSpigot end
 
-- 
2.36.0.windows.1

