From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Sat, 4 Apr 2020 17:00:20 -0700
Subject: [PATCH] Implement [Tuinity-0046] Consolidate flush calls for entity
 tracker packets by Spottedleaf


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 6707b5b92b447de613d4540695d21b46c45a043d..a6fa6928fab133f352bda8aa44d2cf9cb9d563bc 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -887,7 +887,26 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
                 this.methodProfiler.b();
                 this.methodProfiler.a("tracker");
                 worldserver.timings.tracker.startTiming(); // Spigot
-                worldserver.getTracker().updatePlayers();
+                // Tuinity start - controlled flush for entity tracker packets
+                List<NetworkManager> disabledFlushes = new ArrayList<>(worldserver.players.size());
+                for (EntityHuman player : worldserver.players) {
+                    if (!(player instanceof EntityPlayer)) continue; // skip non-player entities
+                    PlayerConnection connection = ((EntityPlayer) player).playerConnection;
+                    if (connection != null) {
+                        connection.networkManager.disableAutomaticFlush();
+                        disabledFlushes.add(connection.networkManager);
+                    }
+                }
+                try {
+                    // Tuinity end
+                    worldserver.getTracker().updatePlayers();
+                    // Tuinity start - controlled flush for entity tracker packets
+                } finally {
+                    for (NetworkManager networkManager : disabledFlushes) {
+                        networkManager.enableAutomaticFlush();
+                    }
+                }
+                // Tuinity end
                 worldserver.timings.tracker.stopTiming(); // Spigot
                 this.methodProfiler.b();
                 this.methodProfiler.b();
