From b8881018ca2178eb7a3e596a688d48ae19e40a2e Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 15 May 2022 18:46:50 +0200
Subject: [PATCH] Implement [Nacho-0025] Disable random tickSpeed being
 modified and make it configurable


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index ba6c9be9..eb67bfbb 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -10,6 +10,7 @@ import net.titanium.config.PaperWorldConfig;
 import net.titanium.config.SpigotWorldConfig;
 import net.titanium.config.TacoSpigotWorldConfig;
 import net.titanium.config.TitaniumConfig;
+import net.titanium.config.TitaniumWorldConfig;
 import org.bukkit.Bukkit;
 import org.bukkit.block.BlockState;
 import org.bukkit.craftbukkit.CraftServer;
@@ -179,6 +180,7 @@ public abstract class World implements IBlockAccess {
     public final PaperWorldConfig paperConfigTitanium;
     public final TacoSpigotWorldConfig tacoSpigotConfigTitanium;
     public final IonWorldConfig ionConfigTitanium;
+    public final TitaniumWorldConfig titaniumConfig;
     // Titanium end
 
     public final co.aikar.timings.WorldTimingsHandler timings; // Spigot
@@ -204,10 +206,12 @@ public abstract class World implements IBlockAccess {
     protected World(IDataManager idatamanager, WorldData worlddata, WorldProvider worldprovider, MethodProfiler methodprofiler, boolean flag, ChunkGenerator gen, org.bukkit.World.Environment env) {
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(); // Spigot
         this.paperSpigotConfig = new org.github.paperspigot.PaperSpigotWorldConfig(); // PaperSpigot
-        this.spigotConfigTitanium = TitaniumConfig.get().spigot.worldSettings;
-        this.paperConfigTitanium = TitaniumConfig.get().paper.worldSettings;
-        this.tacoSpigotConfigTitanium = TitaniumConfig.get().taco.worldSettings;
-        this.ionConfigTitanium = TitaniumConfig.get().ion.worldSettings;
+        TitaniumConfig titaniumConfig = TitaniumConfig.get();
+        this.spigotConfigTitanium = titaniumConfig.spigot.worldSettings;
+        this.paperConfigTitanium = titaniumConfig.paper.worldSettings;
+        this.tacoSpigotConfigTitanium = titaniumConfig.taco.worldSettings;
+        this.ionConfigTitanium = titaniumConfig.ion.worldSettings;
+        this.titaniumConfig = titaniumConfig.worldSettings;
         this.generator = gen;
         this.world = new CraftWorld((WorldServer) this, gen, env);
         this.ticksPerAnimalSpawns = this.getServer().getTicksPerAnimalSpawns(); // CraftBukkit
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index d67bb695..ad769d2b 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -480,7 +480,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
                 this.methodProfiler.c("tickBlocks");
                 timings.chunkTicksBlocks.startTiming(); // Spigot
-                i1 = this.getGameRules().c("randomTickSpeed");
+                i1 = titaniumConfig.disabledGamerules.randomTickSpeed ? 3 : this.getGameRules().c("randomTickSpeed");
                 if (i1 > 0) {
                     ChunkSection[] achunksection = chunk.getSections();
                     int j1 = achunksection.length;
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 8bf00dfa..c3b8e36d 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -130,6 +130,8 @@ public class TitaniumConfig extends OkaeriConfig {
 
     }
 
+    public TitaniumWorldConfig worldSettings = new TitaniumWorldConfig();
+
     public Commands commands = new Commands();
 
     public static class Commands extends TitaniumConfigSection {
diff --git a/src/main/java/net/titanium/config/TitaniumWorldConfig.java b/src/main/java/net/titanium/config/TitaniumWorldConfig.java
new file mode 100644
index 00000000..1b1bc138
--- /dev/null
+++ b/src/main/java/net/titanium/config/TitaniumWorldConfig.java
@@ -0,0 +1,19 @@
+package net.titanium.config;
+
+import eu.okaeri.configs.annotation.NameModifier;
+import eu.okaeri.configs.annotation.NameStrategy;
+import eu.okaeri.configs.annotation.Names;
+
+@Names(strategy = NameStrategy.HYPHEN_CASE, modifier = NameModifier.TO_LOWER_CASE)
+public class TitaniumWorldConfig extends TitaniumConfigSection {
+
+    public DisabledGamerules disabledGamerules = new DisabledGamerules();
+
+    public static class DisabledGamerules extends TitaniumConfigSection {
+
+        //TODO comment this
+        public boolean randomTickSpeed = false;
+
+    }
+
+}
-- 
2.36.0.windows.1

