From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hpfxd <me@hpfxd.com>
Date: Thu, 4 Nov 2021 09:16:42 -0400
Subject: [PATCH] Implement [PandaSpigot-0033] Configurable arrow trajectory by
 hpfxd


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 80c1e71e03e168821152f2e1b19ea19a9de02b50..1d3a3e39e1b631669ab219ef04b40ff42cf435c4 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -3,6 +3,7 @@ package net.minecraft.server;
 import java.util.List;
 
 // CraftBukkit start
+import net.titanium.config.TitaniumConfig;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.player.PlayerPickupItemEvent;
@@ -95,7 +96,7 @@ public class EntityArrow extends Entity implements IProjectile {
         this.motX = (double) (-MathHelper.sin(this.yaw / 180.0F * 3.1415927F) * MathHelper.cos(this.pitch / 180.0F * 3.1415927F));
         this.motZ = (double) (MathHelper.cos(this.yaw / 180.0F * 3.1415927F) * MathHelper.cos(this.pitch / 180.0F * 3.1415927F));
         this.motY = (double) (-MathHelper.sin(this.pitch / 180.0F * 3.1415927F));
-        this.shoot(this.motX, this.motY, this.motZ, f * 1.5F, 1.0F);
+        this.shoot(this.motX, this.motY, this.motZ, f * 1.5F, TitaniumConfig.get().paper.settings.includeRandomnessInArrowTrajectory ? 1.0F : 0); // PandaSpigot
     }
 
     protected void h() {
@@ -108,9 +109,11 @@ public class EntityArrow extends Entity implements IProjectile {
         d0 /= (double) f2;
         d1 /= (double) f2;
         d2 /= (double) f2;
+        if (f1 != 0) { // PandaSpigot - Configurable random arrow trajectory
         d0 += this.random.nextGaussian() * (double) (this.random.nextBoolean() ? -1 : 1) * 0.007499999832361937D * (double) f1;
         d1 += this.random.nextGaussian() * (double) (this.random.nextBoolean() ? -1 : 1) * 0.007499999832361937D * (double) f1;
         d2 += this.random.nextGaussian() * (double) (this.random.nextBoolean() ? -1 : 1) * 0.007499999832361937D * (double) f1;
+        } // PandaSpigot - closing bracket
         d0 *= (double) f;
         d1 *= (double) f;
         d2 *= (double) f;
diff --git a/src/main/java/net/titanium/config/PaperConfig.java b/src/main/java/net/titanium/config/PaperConfig.java
index 20db4ba1eb9523cc2186ef469f65456f1b391121..64208b6fcd2b3328fd06ef88bfaab51f0dde88a4 100644
--- a/src/main/java/net/titanium/config/PaperConfig.java
+++ b/src/main/java/net/titanium/config/PaperConfig.java
@@ -33,6 +33,9 @@ public class PaperConfig extends TitaniumConfigSection {
         @Comment("Whether to save player data to disk.")
         public boolean savePlayerData = true;
 
+        @Comment("Whether arrow projectiles should have a random factor (like in vanilla minecraft).")
+        public boolean includeRandomnessInArrowTrajectory = true;
+
         @Comment("Number of ticks between player data saves to disk.")
         public int playerAutoSaveRate = -1;
 
