From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 23 May 2022 22:20:52 +0200
Subject: [PATCH] Implement [Paper-0202] Shame on you Mojang moves chunk
 loading off https thread


diff --git a/src/main/java/net/minecraft/server/BlockBeacon.java b/src/main/java/net/minecraft/server/BlockBeacon.java
index a73d7b13a8dce788127c73ec1ae7a077d04777e2..b4289c43d0efe0b3e571f989bc2a69cfededfd46 100644
--- a/src/main/java/net/minecraft/server/BlockBeacon.java
+++ b/src/main/java/net/minecraft/server/BlockBeacon.java
@@ -62,8 +62,8 @@ public class BlockBeacon extends BlockContainer {
     }
 
     public static void f(final World world, final BlockPosition blockposition) {
-        HttpUtilities.a.submit(new Runnable() {
-            public void run() {
+        /*HttpUtilities.a.submit(new Runnable() {
+            public void run() {*/ // Paper
                 Chunk chunk = world.getChunkAtWorldCoords(blockposition);
 
                 for (int i = blockposition.getY() - 1; i >= 0; --i) {
@@ -76,8 +76,8 @@ public class BlockBeacon extends BlockContainer {
                     IBlockData iblockdata = world.getType(blockposition1);
 
                     if (iblockdata.getBlock() == Blocks.BEACON) {
-                        ((WorldServer) world).postToMainThread(new Runnable() {
-                            public void run() {
+                        /*((WorldServer) world).postToMainThread(new Runnable() {
+                            public void run() {*/ // Paper
                                 TileEntity tileentity = world.getTileEntity(blockposition);
 
                                 if (tileentity instanceof TileEntityBeacon) {
@@ -85,12 +85,12 @@ public class BlockBeacon extends BlockContainer {
                                     world.playBlockAction(blockposition, Blocks.BEACON, 1, 0);
                                 }
 
-                            }
-                        });
+                            /*}
+                        });*/ // Paper
                     }
                 }
 
-            }
-        });
+            /*}
+        });*/ // Paper
     }
 }
