From dbeb50ad7ff6bfdfd8c4dcd2707f196aafafc1c7 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 23 May 2022 22:12:19 +0200
Subject: [PATCH] Implement [Paper-0321] Cleanup allocated favicon ByteBuf


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index f1107aff8..fddd28c52 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -678,6 +678,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
 
         if (file.isFile()) {
             ByteBuf bytebuf = Unpooled.buffer();
+            ByteBuf bytebuf1 = null; // Paper - cleanup favicon bytebuf
 
             try {
                 BufferedImage bufferedimage = ImageIO.read(file);
@@ -685,13 +686,14 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
                 Validate.validState(bufferedimage.getWidth() == 64, "Must be 64 pixels wide", new Object[0]);
                 Validate.validState(bufferedimage.getHeight() == 64, "Must be 64 pixels high", new Object[0]);
                 ImageIO.write(bufferedimage, "PNG", new ByteBufOutputStream(bytebuf));
-                ByteBuf bytebuf1 = Base64.encode(bytebuf);
+                /*ByteBuf */ bytebuf1 = Base64.encode(bytebuf); // Paper - cleanup favicon bytebuf
 
                 serverping.setFavicon("data:image/png;base64," + bytebuf1.toString(Charsets.UTF_8));
             } catch (Exception exception) {
                 MinecraftServer.LOGGER.error("Couldn\'t load server icon", exception);
             } finally {
                 bytebuf.release();
+                if (bytebuf1 != null) bytebuf1.release(); // Paper - cleanup favicon bytebuf
             }
         }
 
-- 
2.36.0.windows.1

