From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 23 May 2022 22:08:01 +0200
Subject: [PATCH] Implement [Paper-0173] Optimize World.isLoaded(BlockPosition)


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 03c026c2fd2caa7509c3a033c4a6876916ec4e02..db69fd68bcc05a5325f5105671b36f0e5eda7d33 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -335,7 +335,7 @@ public abstract class World implements IBlockAccess {
     }
 
     public boolean isLoaded(BlockPosition blockposition) {
-        return this.a(blockposition, true);
+        return getChunkIfLoaded(blockposition.getX() >> 4, blockposition.getZ() >> 4) != null; // Paper
     }
 
     // Titanium start
