From aec04d73d0fb22763197c661e2349b479bba35dd Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 25 May 2022 21:08:49 +0200
Subject: [PATCH] Implement [SportPaper-0198] Optimize head rotation patch


diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 3fe42a9a7..8e9597417 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -413,7 +413,13 @@ public class EntityTrackerEntry {
                     // CraftBukkit start - Fix for nonsensical head yaw
                     if (this.tracker instanceof EntityLiving) { // [SportPaper-0027] Fix head rotation packet spam
                         this.i = MathHelper.d(this.tracker.getHeadRotation() * 256.0F / 360.0F);
-                        this.broadcast(new PacketPlayOutEntityHeadRotation(this.tracker, (byte) i));
+                        // SportPaper start
+                        /* This was originally introduced by CraftBukkit, though the implementation is wrong since it's broadcasting
+                           the packet again in a method that is already called for each player. This would create a very serious performance issue
+                           with high player and entity counts (each sendPacket call involves waking up the event loop and flushing the network stream).
+                           this.broadcast(new PacketPlayOutEntityHeadRotation(this.tracker, (byte) i));*/
+                        entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityHeadRotation(this.tracker, (byte) i));
+                        // CraftBukkit end
                     }
                     // CraftBukkit end
 
-- 
2.38.1.windows.1

