From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sat, 4 Jun 2022 12:39:26 +0200
Subject: [PATCH] Improve publishing project to maven repository


diff --git a/build.gradle b/build.gradle
index fea68dea8801c49c69b76b83801614ad94e95a6f..93aeca9e389538fa7fbc72fec70d8b722644faab 100644
--- a/build.gradle
+++ b/build.gradle
@@ -1,6 +1,6 @@
 import com.github.jengelman.gradle.plugins.shadow.transformers.Log4j2PluginsCacheFileTransformer
 
-apply plugin: 'com.github.johnrengelman.shadow'
+apply plugin: 'com.palantir.git-version'
 
 String minecraftVersion = "1_8_R3"
 
