From 8051823b83fd5dafef3946c75f13ee41ed243bc6 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 25 May 2022 22:17:14 +0200
Subject: [PATCH] Implement [Tuinity-????] Skip updating entity tracker without
 players


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 17a1629a9..c2bf4fc79 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -901,7 +901,9 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
                 }
                 try {
                     // PandaSpigot end
-                    worldserver.getTracker().updatePlayers();
+                    if (this.getPlayerList().getPlayerCount() != 0) { // Tuinity
+                        worldserver.getTracker().updatePlayers();
+                    }
                     // PandaSpigot start - controlled flush for entity tracker packets
                 } finally {
                     for (NetworkManager networkManager : disabledFlushes) {
-- 
2.36.0.windows.1

