From 4fcf53f41c806b61887f2405cb141fe47318c33b Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 5 Jun 2022 16:10:39 +0200
Subject: [PATCH] Remove unused fields in config


diff --git a/build.gradle b/build.gradle
index 93aeca9e3..02bdc89c8 100644
--- a/build.gradle
+++ b/build.gradle
@@ -26,11 +26,6 @@ dependencies {
 
     implementation 'com.eatthepath:fast-uuid:0.2.0' // Nacho - Use jchambers' FAST UUID methods
 
-    // Titanium start - Use okaeri-configs for server configuration
-    implementation 'eu.okaeri:okaeri-configs-yaml-bukkit:4.0.0-beta23'
-    implementation 'eu.okaeri:okaeri-configs-serdes-commons:4.0.0-beta23'
-    // Titanium end
-
     // PandaSpigot start - Use TerminalConsoleAppender
     implementation 'net.minecrell:terminalconsoleappender:1.3.0'
     implementation 'org.jline:jline-terminal-jansi:3.20.0'
diff --git a/src/main/java/net/titanium/config/PaperConfig.java b/src/main/java/net/titanium/config/PaperConfig.java
index 589f0103f..a0560a509 100644
--- a/src/main/java/net/titanium/config/PaperConfig.java
+++ b/src/main/java/net/titanium/config/PaperConfig.java
@@ -26,45 +26,15 @@ public class PaperConfig extends TitaniumConfigSection {
 
         @Comment("Whether interaction packets should be limited from clients.")
         public boolean limitPlayerInteractions = true;
-
-        @Comment("Size of the region file cache.")
-        public int regionFileCacheSize = 256;
-
         @Comment("Speed at which baby zombies can move.")
         public double babyZombieMovementSpeed = 0.5;
 
-        @Comment("Whether to save empty scoreboard teams.")
-        public boolean saveEmptyScoreboardTeams = false;
-
         @Comment("Whether to save player data to disk.")
         public boolean savePlayerData = true;
 
-        @Comment("Whether to actively tick worlds with no players.")
-        public boolean tickEmptyWorlds = true;
-
-        @Comment("Whether to abort server startup if at least one plugin fails to load.")
-        public boolean requireAllPlugins = false;
-
-        public SpamLimiter spamLimiter = new SpamLimiter();
-
-        public static class SpamLimiter extends TitaniumConfigSection {
-
-            @Comment("Value to increment for each tab complete event.")
-            public int tabSpamIncrement = 10;
-            @Comment("Maximum number of increment values before disconnecting the player.")
-            public int tabSpamLimit = 500;
-
-        }
-
-        @Comment("Duration to wait before suspending an server with no players online.")
-        public Duration emptyServerSuspend = Duration.ofMinutes(1);
-
         @Comment("Whether arrow projectiles should have a random factor (like in vanilla minecraft).")
         public boolean includeRandomnessInArrowTrajectory = true;
 
-        @Comment("Whether arrow projectiles should add a random amount of damage (like in vanilla minecraft).")
-        public boolean includeRandomnessInArrowDamage = true;
-
         @Comment("Number of ticks between player data saves to disk.")
         public int playerAutoSaveRate = -1;
 
@@ -95,9 +65,6 @@ public class PaperConfig extends TitaniumConfigSection {
 
     }
 
-    @Comment("Whether to fire a physics event for redstone activation.")
-    public boolean firePhysicsEventForRedstone = false;
-
     @Comment("Percentage improvement to player damage with certain potions.")
     public EffectModifiers effectModifiers = new EffectModifiers();
 
diff --git a/src/main/java/net/titanium/config/PaperWorldConfig.java b/src/main/java/net/titanium/config/PaperWorldConfig.java
index b68d46c34..9897a8ece 100644
--- a/src/main/java/net/titanium/config/PaperWorldConfig.java
+++ b/src/main/java/net/titanium/config/PaperWorldConfig.java
@@ -41,15 +41,6 @@ public class PaperWorldConfig extends TitaniumConfigSection {
     @Comment("Whether to disable the suffocation check before every teleportation event.")
     public boolean disableTeleportationSuffocationCheck = false;
 
-    @Comment("Whether to consider player time in chunks as a factor to unload chunks.")
-    public boolean useChunkInhabitedTimer = true;
-
-    @Comment("Maximum number of chunks to auto save per tick.")
-    public int maxAutoSaveChunksPerTick = 24;
-
-    @Comment("Whether to delete corrupt tile entities.")
-    public boolean removeCorruptTileEntities = false;
-
     @Comment("Y-level at which squids will spawn.")
     public SquidSpawnHeight squidSpawnHeight = new SquidSpawnHeight();
 
@@ -224,9 +215,6 @@ public class PaperWorldConfig extends TitaniumConfigSection {
     @Comment("Whether falling blocks should not break when colliding with signs.")
     public boolean fallingBlocksCollideWithSigns = false;
 
-    @Comment("Whether armor stands should be subject to collision checks.")
-    public boolean armorStandsDoCollisionEntityLookups = true;
-
     @Comment("Whether to not send cave and ambient sounds.")
     public boolean disableMoodSounds = false;
 
diff --git a/src/main/java/net/titanium/config/SpigotConfig.java b/src/main/java/net/titanium/config/SpigotConfig.java
index 06a5afc4c..899694cde 100644
--- a/src/main/java/net/titanium/config/SpigotConfig.java
+++ b/src/main/java/net/titanium/config/SpigotConfig.java
@@ -135,7 +135,6 @@ public class SpigotConfig extends TitaniumConfigSection {
         public String restart = "&cServer is restarting!";
         public String whitelist = "&cYou are not on the whitelist!";
         public String unknownCommand = "&7Unknown command. Type &c/help &7for help.";
-        public String internalError ="&cAn internal error occurred while attempting to perform this command";
         public String serverFull = "&cThe server is full!";
         public String outdatedClient = "&7Outdated client! Please use &c{0}";
         public String outdatedServer = "&7Outdated server! Server is on &c{0}";
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index fad595159..9173676ba 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -219,6 +219,7 @@ public final class CraftServer implements Server {
                 registry.register(new WarningStateTransformer());
             });
             it.withBindFile((File) console.options.valueOf("titanium-settings"));
+            it.withRemoveOrphans(false);
             it.saveDefaults();
             it.load(true);
         });
-- 
2.38.1.windows.1

