From bc9ed63ec44c3c8d3c1dbcf74466f156d6b99e64 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 5 Jun 2022 00:55:30 +0200
Subject: [PATCH] Revert using Jafama


diff --git a/src/main/java/net/minecraft/server/BiomeMesa.java b/src/main/java/net/minecraft/server/BiomeMesa.java
index 574c3cd97..e9ec299ff 100644
--- a/src/main/java/net/minecraft/server/BiomeMesa.java
+++ b/src/main/java/net/minecraft/server/BiomeMesa.java
@@ -2,7 +2,6 @@ package net.minecraft.server;
 
 import java.util.Arrays;
 import java.util.Random;
-import net.jafama.FastMath;
 
 public class BiomeMesa extends BiomeBase {
 
@@ -86,7 +85,7 @@ public class BiomeMesa extends BiomeBase {
         IBlockData iblockdata = Blocks.STAINED_HARDENED_CLAY.getBlockData();
         IBlockData iblockdata1 = this.al;
         int j1 = (int) (d0 / 3.0D + 3.0D + random.nextDouble() * 0.25D);
-        boolean flag = FastMath.cos(d0 / 3.0D * 3.141592653589793D) > 0.0D;
+        boolean flag = Math.cos(d0 / 3.0D * 3.141592653589793D) > 0.0D;
         int k1 = -1;
         boolean flag1 = false;
 
diff --git a/src/main/java/net/minecraft/server/ChunkProviderHell.java b/src/main/java/net/minecraft/server/ChunkProviderHell.java
index 4c9426528..51538b7c4 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderHell.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderHell.java
@@ -2,7 +2,6 @@ package net.minecraft.server;
 
 import java.util.List;
 import java.util.Random;
-import net.jafama.FastMath;
 
 public class ChunkProviderHell implements IChunkProvider {
 
@@ -238,7 +237,7 @@ public class ChunkProviderHell implements IChunkProvider {
         int l1;
 
         for (l1 = 0; l1 < i1; ++l1) {
-            adouble1[l1] = FastMath.cos((double) l1 * 3.141592653589793D * 6.0D / (double) i1) * 2.0D;
+            adouble1[l1] = Math.cos((double) l1 * 3.141592653589793D * 6.0D / (double) i1) * 2.0D;
             double d2 = (double) l1;
 
             if (l1 > i1 / 2) {
diff --git a/src/main/java/net/minecraft/server/EntityBoat.java b/src/main/java/net/minecraft/server/EntityBoat.java
index aeb1c5a88..590d84c92 100644
--- a/src/main/java/net/minecraft/server/EntityBoat.java
+++ b/src/main/java/net/minecraft/server/EntityBoat.java
@@ -3,7 +3,6 @@ package net.minecraft.server;
 import java.util.List;
 
 // CraftBukkit start
-import net.jafama.FastMath;
 import org.bukkit.Location;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.Vehicle;
@@ -189,8 +188,8 @@ public class EntityBoat extends Entity {
         int j;
 
         if (d3 > 0.2975D) {
-            d4 = FastMath.cos((double) this.yaw * 3.141592653589793D / 180.0D);
-            d5 = FastMath.sin((double) this.yaw * 3.141592653589793D / 180.0D);
+            d4 = Math.cos((double) this.yaw * 3.141592653589793D / 180.0D);
+            d5 = Math.sin((double) this.yaw * 3.141592653589793D / 180.0D);
 
             for (j = 0; (double) j < 1.0D + d3 * 60.0D; ++j) {
                 double d6 = (double) (this.random.nextFloat() * 2.0F - 1.0F);
@@ -256,8 +255,8 @@ public class EntityBoat extends Entity {
                 EntityLiving entityliving = (EntityLiving) this.passenger;
                 float f = this.passenger.yaw + -entityliving.aZ * 90.0F;
 
-                this.motX += -FastMath.sin((double) (f * 3.1415927F / 180.0F)) * this.b * (double) entityliving.ba * 0.05000000074505806D;
-                this.motZ += FastMath.cos((double) (f * 3.1415927F / 180.0F)) * this.b * (double) entityliving.ba * 0.05000000074505806D;
+                this.motX += -Math.sin((double) (f * 3.1415927F / 180.0F)) * this.b * (double) entityliving.ba * 0.05000000074505806D;
+                this.motZ += Math.cos((double) (f * 3.1415927F / 180.0F)) * this.b * (double) entityliving.ba * 0.05000000074505806D;
             }
             // CraftBukkit start - Support unoccupied deceleration
             else if (unoccupiedDeceleration >= 0) {
@@ -411,8 +410,8 @@ public class EntityBoat extends Entity {
 
     public void al() {
         if (this.passenger != null) {
-            double d0 = FastMath.cos((double) this.yaw * 3.141592653589793D / 180.0D) * 0.4D;
-            double d1 = FastMath.sin((double) this.yaw * 3.141592653589793D / 180.0D) * 0.4D;
+            double d0 = Math.cos((double) this.yaw * 3.141592653589793D / 180.0D) * 0.4D;
+            double d1 = Math.sin((double) this.yaw * 3.141592653589793D / 180.0D) * 0.4D;
 
             this.passenger.setPosition(this.locX + d0, this.locY + this.an() + this.passenger.am(), this.locZ + d1);
         }
diff --git a/src/main/java/net/minecraft/server/EntityEnderDragon.java b/src/main/java/net/minecraft/server/EntityEnderDragon.java
index 518543ffa..5bd0dcdd1 100644
--- a/src/main/java/net/minecraft/server/EntityEnderDragon.java
+++ b/src/main/java/net/minecraft/server/EntityEnderDragon.java
@@ -6,7 +6,7 @@ import java.util.Iterator;
 import java.util.List;
 
 // CraftBukkit start
-import net.jafama.FastMath;
+import net.titanium.util.MathUtil;
 import org.bukkit.block.BlockState;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.util.BlockStateListPopulator;
@@ -579,7 +579,7 @@ public class EntityEnderDragon extends EntityInsentient implements IComplex, IMo
                     double deltaX = this.locX - player.locX;
                     double deltaZ = this.locZ - player.locZ;
                     double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
-                    if ( world.spigotConfigTitanium.dragonDeathSoundRadius > 0 && distanceSquared > FastMath.pow2(world.spigotConfigTitanium.dragonDeathSoundRadius) ) continue; // Spigot
+                    if ( world.spigotConfigTitanium.dragonDeathSoundRadius > 0 && distanceSquared > MathUtil.pow2(world.spigotConfigTitanium.dragonDeathSoundRadius) ) continue; // Spigot
                     if (distanceSquared > viewDistance * viewDistance) {
                         double deltaLength = Math.sqrt(distanceSquared);
                         double relativeX = player.locX + (deltaX / deltaLength) * viewDistance;
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index fe479750c..ed4648325 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -9,7 +9,6 @@ import java.util.List;
 import java.util.UUID;
 
 // CraftBukkit start
-import net.jafama.FastMath;
 import net.titanium.config.TitaniumConfig;
 import org.bukkit.craftbukkit.entity.CraftHumanEntity;
 import org.bukkit.craftbukkit.entity.CraftItem;
@@ -591,9 +590,9 @@ public abstract class EntityHuman extends EntityLiving {
                 entityitem.motY = (double) (-MathHelper.sin(this.pitch / 180.0F * 3.1415927F) * f + 0.1F);
                 f1 = this.random.nextFloat() * 3.1415927F * 2.0F;
                 f = 0.02F * this.random.nextFloat();
-                entityitem.motX += FastMath.cos((double) f1) * (double) f;
+                entityitem.motX += Math.cos((double) f1) * (double) f;
                 entityitem.motY += (double) ((this.random.nextFloat() - this.random.nextFloat()) * 0.1F);
-                entityitem.motZ += FastMath.sin((double) f1) * (double) f;
+                entityitem.motZ += Math.sin((double) f1) * (double) f;
             }
 
             // CraftBukkit start - fire PlayerDropItemEvent
diff --git a/src/main/java/net/minecraft/server/EntityMinecartAbstract.java b/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
index 4fc65f795..0446e2be5 100644
--- a/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
+++ b/src/main/java/net/minecraft/server/EntityMinecartAbstract.java
@@ -5,7 +5,6 @@ import java.util.Iterator;
 import java.util.Map;
 
 // CraftBukkit start
-import net.jafama.FastMath;
 import org.bukkit.Location;
 import org.bukkit.entity.Vehicle;
 import org.bukkit.event.vehicle.VehicleDamageEvent;
@@ -441,8 +440,8 @@ public abstract class EntityMinecartAbstract extends Entity implements INamableT
         if (this.passenger instanceof EntityLiving) {
             d6 = (double) ((EntityLiving) this.passenger).ba;
             if (d6 > 0.0D) {
-                d7 = -FastMath.sin((double) (this.passenger.yaw * 3.1415927F / 180.0F));
-                d8 = FastMath.cos((double) (this.passenger.yaw * 3.1415927F / 180.0F));
+                d7 = -Math.sin((double) (this.passenger.yaw * 3.1415927F / 180.0F));
+                d8 = Math.cos((double) (this.passenger.yaw * 3.1415927F / 180.0F));
                 d9 = this.motX * this.motX + this.motZ * this.motZ;
                 if (d9 < 0.01D) {
                     this.motX += d7 * 0.1D;
diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index a8030338f..6ca8b5f07 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -1,6 +1,5 @@
 package net.minecraft.server;
 
-import net.jafama.FastMath;
 import org.bukkit.event.entity.ExplosionPrimeEvent; // CraftBukkit
 
 public class EntityTNTPrimed extends Entity {
@@ -30,9 +29,9 @@ public class EntityTNTPrimed extends Entity {
         this.setPosition(d0, d1, d2);
         float f = (float) (Math.random() * 3.1415927410125732D * 2.0D);
 
-        this.motX = (double) (-((float) FastMath.sin((double) f)) * 0.02F);
+        this.motX = (double) (-((float) Math.sin((double) f)) * 0.02F);
         this.motY = 0.20000000298023224D;
-        this.motZ = (double) (-((float) FastMath.cos((double) f)) * 0.02F);
+        this.motZ = (double) (-((float) Math.cos((double) f)) * 0.02F);
         this.fuseTicks = 80;
         this.lastX = d0;
         this.lastY = d1;
diff --git a/src/main/java/net/minecraft/server/EntityWither.java b/src/main/java/net/minecraft/server/EntityWither.java
index a351f71ab..b2cb97d78 100644
--- a/src/main/java/net/minecraft/server/EntityWither.java
+++ b/src/main/java/net/minecraft/server/EntityWither.java
@@ -6,7 +6,7 @@ import java.util.Iterator;
 import java.util.List;
 
 // CraftBukkit start
-import net.jafama.FastMath;
+import net.titanium.util.MathUtil;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.EntityRegainHealthEvent;
 import org.bukkit.event.entity.ExplosionPrimeEvent;
@@ -192,7 +192,7 @@ public class EntityWither extends EntityMonster implements IRangedEntity {
                     double deltaX = this.locX - player.locX;
                     double deltaZ = this.locZ - player.locZ;
                     double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
-                    if ( world.spigotConfigTitanium.witherSpawnSoundRadius > 0 && distanceSquared > FastMath.pow2(world.spigotConfigTitanium.witherSpawnSoundRadius)) continue; // Spigot
+                    if ( world.spigotConfigTitanium.witherSpawnSoundRadius > 0 && distanceSquared > MathUtil.pow2(world.spigotConfigTitanium.witherSpawnSoundRadius)) continue; // Spigot
                     if (distanceSquared > viewDistance * viewDistance) {
                         double deltaLength = Math.sqrt(distanceSquared);
                         double relativeX = player.locX + (deltaX / deltaLength) * viewDistance;
diff --git a/src/main/java/net/titanium/util/MathUtil.java b/src/main/java/net/titanium/util/MathUtil.java
new file mode 100644
index 000000000..f913847bb
--- /dev/null
+++ b/src/main/java/net/titanium/util/MathUtil.java
@@ -0,0 +1,12 @@
+package net.titanium.util;
+
+public final class MathUtil {
+
+    private MathUtil() {
+    }
+
+    public static float pow2(float value) {
+        return value * value;
+    }
+
+}
-- 
2.36.0.windows.1

