From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hpfxd <me@hpfxd.com>
Date: Sun, 8 May 2022 19:16:14 -0400
Subject: [PATCH] Implement [PandaSpigot-0049] Add setting for proxy online
 mode status by hpfxd


diff --git a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
index 4ad369ae4ddfeaa64b32d0f2768163c6c6b49b26..88f4d5eff025c75ef257af6fc6f74ee4acfd8616 100644
--- a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
+++ b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
@@ -68,7 +68,7 @@ public class NameReferencingFileConverter {
             }
         }), String.class);
 
-        if (minecraftserver.getOnlineMode() || TitaniumConfig.get().spigot.settings.bungeecord) { // Spigot: bungee = online mode, for now.
+        if (minecraftserver.getOnlineMode() || TitaniumConfig.get().isProxyOnlineMode()) { // Spigot: bungee = online mode, for now.
             minecraftserver.getGameProfileRepository().findProfilesByNames(astring, Agent.MINECRAFT, profilelookupcallback);
         } else {
             String[] astring1 = astring;
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index c39115e4909a66926417fe183e5974ea21844f54..2a4ab7e8d912fccddb3a53bf9b9a64e6b7638455 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -53,6 +53,14 @@ public class TitaniumConfig extends TitaniumConfigSection {
     @Comment("Configuration from ion.yml")
     public IonConfig ion = new IonConfig();
 
+    @Comment("Instructs the server how to handle player UUIDs and data when behind a proxy.")
+    @Comment("Set to match your proxy's online-mode setting.")
+    public boolean proxyOnlineMode = true;
+
+    public boolean isProxyOnlineMode() {
+        return Bukkit.getOnlineMode() || (spigot.settings.bungeecord && this.proxyOnlineMode);
+    }
+
     @Comment("Blocks through which enderpearls will pass through (AKA Taliban Pearls)")
     public PearlPassthrough pearlPassthrough = new PearlPassthrough();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 441a89b5bc0996ca422c5d84f9ea2bfe1d5f2828..3a1e8ad2305472581659fa97b50142f5a9843063 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1413,7 +1413,7 @@ public final class CraftServer implements Server {
             // Spigot Start
             GameProfile profile = null;
             // Only fetch an online UUID in online mode
-            if ( MinecraftServer.getServer().getOnlineMode() || serverConfig.spigot.settings.bungeecord )
+            if ( MinecraftServer.getServer().getOnlineMode() || serverConfig.isProxyOnlineMode())
             {
                 profile = MinecraftServer.getServer().getUserCache().getProfile( name );
             }
