From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Thu, 12 May 2022 21:32:01 -0300
Subject: [PATCH] Implement [PandaSpigot-0054] Disconnect for payload errors by
 uRyanxD


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index cc150f4af14874b7d63376181feeac5c46758183..1ec7ce85d1db6f457cd8a98da1519e795ebfc516 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2140,19 +2140,36 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
         }
         // CraftBukkit start
         else if (packetplayincustompayload.a().equals("REGISTER")) {
-            String channels = packetplayincustompayload.b().toString(com.google.common.base.Charsets.UTF_8);
-            for (String channel : channels.split("\0")) {
-                getPlayer().addChannel(channel);
+            // PandaSpigot start - Disconnect for payload errors
+            try {
+                String channels = packetplayincustompayload.b().toString(com.google.common.base.Charsets.UTF_8);
+                for (String channel : channels.split("\0")) {
+                    getPlayer().addChannel(channel);
+                }
+            } catch (Exception ex) {
+                PlayerConnection.c.error("Couldn\'t register custom payload", ex);
+                this.disconnect("Invalid payload REGISTER!");
             }
         } else if (packetplayincustompayload.a().equals("UNREGISTER")) {
-            String channels = packetplayincustompayload.b().toString(com.google.common.base.Charsets.UTF_8);
-            for (String channel : channels.split("\0")) {
-                getPlayer().removeChannel(channel);
+            try {
+                String channels = packetplayincustompayload.b().toString(com.google.common.base.Charsets.UTF_8);
+                for (String channel : channels.split("\0")) {
+                    getPlayer().removeChannel(channel);
+                }
+            } catch (Exception ex) {
+                PlayerConnection.c.error("Couldn\'t register custom payload", ex);
+                this.disconnect("Invalid payload UNREGISTER!");
             }
         } else {
-            byte[] data = new byte[packetplayincustompayload.b().readableBytes()];
-            packetplayincustompayload.b().readBytes(data);
-            server.getMessenger().dispatchIncomingMessage(player.getBukkitEntity(), packetplayincustompayload.a(), data);
+            try {
+                byte[] data = new byte[packetplayincustompayload.b().readableBytes()];
+                packetplayincustompayload.b().readBytes(data);
+                server.getMessenger().dispatchIncomingMessage(player.getBukkitEntity(), packetplayincustompayload.a(), data);
+            } catch (Exception ex) {
+                PlayerConnection.c.error("Couldn\'t register custom payload", ex);
+                this.disconnect("Invalid custom payload!");
+            }
+            // PandaSpigot end - Disconnect for payload errors
         }
         // CraftBukkit end
         // CraftBukkit start
