From 9d35be78788b5df0a2d0f7aebeda89d35d434116 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 5 Jun 2022 17:10:31 +0200
Subject: [PATCH] Add option to change internal error message


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 4e4610b24..d30402e4b 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1193,7 +1193,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                 return;
             }
         } catch (org.bukkit.command.CommandException ex) {
-            player.sendMessage(org.bukkit.ChatColor.RED + "An internal error occurred while attempting to perform this command");
+            player.sendMessage(server.serverConfig.spigot.messages.internalErrorCommand);
             java.util.logging.Logger.getLogger(PlayerConnection.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
             SpigotTimings.playerCommandTimer.stopTiming(); // Spigot
             return;
diff --git a/src/main/java/net/titanium/config/SpigotConfig.java b/src/main/java/net/titanium/config/SpigotConfig.java
index 899694cde..22d1073d4 100644
--- a/src/main/java/net/titanium/config/SpigotConfig.java
+++ b/src/main/java/net/titanium/config/SpigotConfig.java
@@ -134,11 +134,14 @@ public class SpigotConfig extends TitaniumConfigSection {
 
         public String restart = "&cServer is restarting!";
         public String whitelist = "&cYou are not on the whitelist!";
-        public String unknownCommand = "&7Unknown command. Type &c/help &7for help.";
         public String serverFull = "&cThe server is full!";
         public String outdatedClient = "&7Outdated client! Please use &c{0}";
         public String outdatedServer = "&7Outdated server! Server is on &c{0}";
 
+        public String unknownCommand = "&7Unknown command. Type &c/help &7for help.";
+        public String internalErrorCommand = "&cAn internal error occurred while attempting to perform this command";
+        public String internalErrorTabComplete = "&cAn internal error occurred while attempting to tab-complete this command";
+
     }
 
     public Commands commands = new Commands();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9d121127a..e0e1a91dd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1756,7 +1756,7 @@ public final class CraftServer implements Server {
                 completions = getCommandMap().tabComplete(player, message.substring(1), new Location(player.getWorld(), blockPosition.getX(), blockPosition.getY(), blockPosition.getZ()));
             }
         } catch (CommandException ex) {
-            player.sendMessage(ChatColor.RED + "An internal error occurred while attempting to tab-complete this command");
+            player.sendMessage(serverConfig.spigot.messages.internalErrorTabComplete);
             getLogger().log(Level.SEVERE, "Exception when " + player.getName() + " attempted to tab complete " + message, ex);
         }
 
-- 
2.36.0.windows.1

