From 3981ed22097977f6b2eae51a3c4b45b5ff3fcdc5 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 10 Jun 2022 23:08:26 +0200
Subject: [PATCH] Add missing comments in config


diff --git a/src/main/java/net/titanium/config/PaperConfig.java b/src/main/java/net/titanium/config/PaperConfig.java
index a0560a509..9d55dcbeb 100644
--- a/src/main/java/net/titanium/config/PaperConfig.java
+++ b/src/main/java/net/titanium/config/PaperConfig.java
@@ -39,7 +39,7 @@ public class PaperConfig extends TitaniumConfigSection {
         public int playerAutoSaveRate = -1;
 
         @Comment("Maximum number of player data to save per tick.")
-        @CustomKey("max-players-auto-save-per-tic")
+        @CustomKey("max-players-auto-save-per-tick")
         public int maxPlayersAutoSavePerTick_ = 10;
         @Exclude
         public int maxPlayersAutoSavePerTick = 10;
diff --git a/src/main/java/net/titanium/config/SpigotConfig.java b/src/main/java/net/titanium/config/SpigotConfig.java
index fbac6bd5b..19ae5ed01 100644
--- a/src/main/java/net/titanium/config/SpigotConfig.java
+++ b/src/main/java/net/titanium/config/SpigotConfig.java
@@ -189,7 +189,7 @@ public class SpigotConfig extends TitaniumConfigSection {
                 .add("tellraw")
                 .build();
 
-        //TODO comment
+        @Comment("Which commands should be excluded from spamming disconnect (if player use commands to often servers disconnect him).")
         public Set<String> spamExclusions = ImmutableSet.<String>builder()
                 .add("/skill")
                 .build();
diff --git a/src/main/java/net/titanium/config/SpigotWorldConfig.java b/src/main/java/net/titanium/config/SpigotWorldConfig.java
index ef4fd3047..47478b7ca 100644
--- a/src/main/java/net/titanium/config/SpigotWorldConfig.java
+++ b/src/main/java/net/titanium/config/SpigotWorldConfig.java
@@ -20,7 +20,6 @@ public class SpigotWorldConfig extends TitaniumConfigSection {
 
     public AntiXray antiXray = new AntiXray();
 
-    //TODO Create xray instance
     public static class AntiXray extends TitaniumConfigSection {
 
         @Comment("Whether to enable the anti-xray module.")
diff --git a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
index c8eb27730..1ab14d47a 100644
--- a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
@@ -1,5 +1,6 @@
 package net.titanium.config;
 
+import eu.okaeri.configs.annotation.Comment;
 import eu.okaeri.configs.annotation.NameModifier;
 import eu.okaeri.configs.annotation.NameStrategy;
 import eu.okaeri.configs.annotation.Names;
@@ -7,17 +8,19 @@ import eu.okaeri.configs.annotation.Names;
 @Names(strategy = NameStrategy.HYPHEN_CASE, modifier = NameModifier.TO_LOWER_CASE)
 public class TacoSpigotWorldConfig extends TitaniumConfigSection {
 
-    //TODO comment this (See TacoSpigot-0025)
+    @Comment("Cannoning on east and west normally is not calculated properly.")
+    @Comment("Enabling this option resolve this issue, but it's not vanilla behavior.")
     public boolean fixEastWestCannoning = false;
 
     public Tnt tnt = new Tnt();
 
     public static class Tnt extends TitaniumConfigSection {
 
-        //TODO comment this (See TacoSpigot-0024)
+        @Comment("Whether optimize movement if tnt are traveling on long distance.")
+        @Comment("This option may not reflect vanilla behavior.")
         public boolean optimizeMovement = false;
 
-        //TODO comment this (See TacoSpigot-0024)
+        @Comment("Whether calculation of explosions in liquids should be processed.")
         public boolean optimizeLiquidExplosions = true;
 
     }
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 25c914253..e607618c8 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -32,12 +32,12 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
     // For tests to work properly
     @Exclude
-    private static TitaniumConfig DEFAULT_INSTANCE = new TitaniumConfig();
+    private static final TitaniumConfig DEFAULT_INSTANCE = new TitaniumConfig();
     @Exclude
     private static TitaniumConfig INSTANCE;
 
     @Exclude
-    private static Map<String, Pair<String, Command>> COMMANDS = new HashMap<>();
+    private static final Map<String, Pair<String, Command>> COMMANDS = new HashMap<>();
 
     @Comment("Configuration from bukkit.yml")
     public BukkitConfig bukkit = new BukkitConfig();
@@ -80,7 +80,9 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
     }
 
-    //TODO comment this
+    @Comment("How often player statistics (ESC -> Statistics) should be ticked.")
+    @Comment("Default vanilla value is every 1 tick, but every 1 second (20 ticks) is enough in most cases.")
+    @Comment("Making this value even bigger could be good idea and have good impact on performance.")
     public int playerTimeStatisticsInterval = 20;
 
     @Comment("The max NBT size in bytes. The default value is 2097152 which is ~2Mb - it's insane big.")
@@ -118,7 +120,9 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
     public static class Chunk extends TitaniumConfigSection {
 
-        //TODO comment this
+        @Comment("Normally we use bukkit.chunk-gc.load-threshold - it's define how many chunks should be loaded before running GC on them.")
+        @Comment("This option make that value more flexible and depends it on view distance and player count (playerCount * viewDistance^2).")
+        @Comment("We don't want to have a lots chunks loaded if we have empty server, but having to few on full server is not good idea too (especially if some chunks would be unloaded and almost again loaded).")
         public boolean adaptiveChunkGC = false;
 
         @Comment("Enabling this option prevents entities from moving in unloaded chunk (and for eg. falling down in \"slow-motion\")")
@@ -148,7 +152,8 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
         public static class Async extends TitaniumConfigSection {
 
-            //TODO comment this
+            @Comment("Whether some calculation of tnt explosions should be calculated asynchronously.")
+            @Comment("Enabling this could be good idea since it can have good impact on performance.")
             public boolean enabled = false;
 
         }
diff --git a/src/main/java/net/titanium/config/TitaniumWorldConfig.java b/src/main/java/net/titanium/config/TitaniumWorldConfig.java
index 1b1bc1384..026aeb400 100644
--- a/src/main/java/net/titanium/config/TitaniumWorldConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumWorldConfig.java
@@ -1,5 +1,6 @@
 package net.titanium.config;
 
+import eu.okaeri.configs.annotation.Comment;
 import eu.okaeri.configs.annotation.NameModifier;
 import eu.okaeri.configs.annotation.NameStrategy;
 import eu.okaeri.configs.annotation.Names;
@@ -11,7 +12,11 @@ public class TitaniumWorldConfig extends TitaniumConfigSection {
 
     public static class DisabledGamerules extends TitaniumConfigSection {
 
-        //TODO comment this
+        @Comment("Whether gamerule randomTickSpeed should be enabled or not.")
+        @Comment("Plants growth, leaf decay, fire spread etc. depends on this value.")
+        @Comment("Normally we have to cast String to Integer every tick which is costly.")
+        @Comment("If this option is enabled, this gamerule is disabled and we use default vanilla value which is 3.")
+        @Comment("If you know you won't use this gamerule you should enable this option.")
         public boolean randomTickSpeed = false;
 
     }
-- 
2.38.1.windows.1

