From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 31 Mar 2016 19:17:58 -0400
Subject: [PATCH] Implement [Paper-0101] Do not load chunks for light check by
 Aikar


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 5a506772ff1801456ab0708988eabca6d953dcfd..18df311cafd2ca777a07292a69a75f3259d907bc 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -686,6 +686,7 @@ public abstract class World implements IBlockAccess {
             if (blockposition.getY() >= 256) {
                 blockposition = new BlockPosition(blockposition.getX(), 255, blockposition.getZ());
             }
+            if (!this.isLoaded(blockposition)) return 0; // Paper
 
             return this.getChunkAtWorldCoords(blockposition).a(blockposition, 0);
         }
