From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LinsaFTW <25271111+linsaftw@users.noreply.github.com>
Date: Thu, 24 Mar 2022 03:57:32 -0300
Subject: [PATCH] Implement [FlamePaper-0033] Dont check occluding hoppers by
 LinsaFTW


diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
index 4a8a3b84be2c565a5f358401693e549ee579a851..e83e932163cd07f24e3bfd9f783729787671b226 100644
--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
+++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
@@ -572,7 +572,7 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
 
         net.minecraft.server.Chunk chunk = world.getChunkAtWorldCoords(blockposition);
 
-        if (object == null && chunk.getItemCount(blockposition) > 0) {
+        if (object == null && chunk.getItemCount(blockposition) > 0 && !org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(block).isOccluding()) {
             List list = world.a((Entity) null, new AxisAlignedBB(d0 - 0.5D, d1 - 0.5D, d2 - 0.5D, d0 + 0.5D, d1 + 0.5D, d2 + 0.5D), IEntitySelector.c);
 
             if (list.size() > 0) {
