From 38a94e0fb8c9b3cb12db933ab4edabe51c118df2 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 6 Jun 2022 17:46:36 +0200
Subject: [PATCH] Placeholder of used command in unknown command message


diff --git a/src/main/java/net/titanium/config/SpigotConfig.java b/src/main/java/net/titanium/config/SpigotConfig.java
index 22d1073d4..63f4492b2 100644
--- a/src/main/java/net/titanium/config/SpigotConfig.java
+++ b/src/main/java/net/titanium/config/SpigotConfig.java
@@ -135,9 +135,12 @@ public class SpigotConfig extends TitaniumConfigSection {
         public String restart = "&cServer is restarting!";
         public String whitelist = "&cYou are not on the whitelist!";
         public String serverFull = "&cThe server is full!";
+        @Comment("You can use {0} to insert the server version.")
         public String outdatedClient = "&7Outdated client! Please use &c{0}";
+        @Comment("You can use {0} to insert the server version.")
         public String outdatedServer = "&7Outdated server! Server is on &c{0}";
 
+        @Comment("You can use {0} to insert the command that was tried to use (without slash).")
         public String unknownCommand = "&7Unknown command. Type &c/help &7for help.";
         public String internalErrorCommand = "&cAn internal error occurred while attempting to perform this command";
         public String internalErrorTabComplete = "&cAn internal error occurred while attempting to tab-complete this command";
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0a7d717a5..ab5cef0be 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -734,7 +734,10 @@ public final class CraftServer implements Server {
             return true;
         }
 
-        sender.sendMessage(TitaniumConfig.get().spigot.messages.unknownCommand);
+        int spaceIndex = commandLine.indexOf(' ');
+        String commandSubstring = commandLine.substring(0, spaceIndex != -1 ? spaceIndex : commandLine.length());
+
+        sender.sendMessage(java.text.MessageFormat.format(TitaniumConfig.get().spigot.messages.unknownCommand, commandSubstring));
 
         return false;
     }
-- 
2.36.0.windows.1

