From 16b650d9539e998063b46095573bb0ebde4cc1ee Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 31 Mar 2016 19:17:58 -0400
Subject: [PATCH] Implement [Paper-0101] Do not load chunks for light check by
 Aikar


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 2baed2815..7da9e9281 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -685,6 +685,7 @@ public abstract class World implements IBlockAccess {
             if (blockposition.getY() >= 256) {
                 blockposition = new BlockPosition(blockposition.getX(), 255, blockposition.getZ());
             }
+            if (!this.isLoaded(blockposition)) return 0; // Paper
 
             return this.getChunkAtWorldCoords(blockposition).a(blockposition, 0);
         }
-- 
2.38.1.windows.1

