From dd7841c3e3c10629e496dfff6d8b8c9034853306 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Thu, 6 Sep 2018 01:25:56 -0700
Subject: [PATCH] Implement [SportPaper-0162] Fix PlayerInteractEvent not
 cancelling properly by cswhite2000


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 83a4a923d..d80fef6e9 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -732,7 +732,24 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                 }
             }
 
-            if (!cancelled) {
+            if (cancelled) {
+                this.player.getBukkitEntity().updateInventory(); // SPIGOT-2524
+                // SportPaper start - Fix client desync
+                if (itemstack.getItem() == Item.getItemOf(Blocks.WATERLILY)) {
+                    MovingObjectPosition movingObjectPosition1 = this.player.world.rayTrace(vec3d, vec3d1, true, false, false);
+                    if (movingObjectPosition1 != null) {
+                        BlockPosition blockPosition = movingObjectPosition1.a().up();
+                        org.bukkit.craftbukkit.block.CraftBlockState.getBlockState(worldserver, blockPosition.getX(), blockPosition.getY(), blockPosition.getZ()).update(true, false);
+                    }
+                } else if (itemstack.getItem() == Items.BUCKET) {
+                    MovingObjectPosition movingObjectPosition1 = this.player.world.rayTrace(vec3d, vec3d1, true, false, false);
+                    if (movingObjectPosition1 != null) {
+                        BlockPosition blockPosition = movingObjectPosition1.a();
+                        org.bukkit.craftbukkit.block.CraftBlockState.getBlockState(worldserver, blockPosition.getX(), blockPosition.getY(), blockPosition.getZ()).update(true, false);
+                    }
+                }
+                // SportPaper end
+            } else {
                 this.player.playerInteractManager.useItem(this.player, this.player.world, itemstack);
             }
             }
-- 
2.36.0.windows.1

