From f645b6cf5d00b16145d2734f5a9ca537a5c6d58a Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 10 Jun 2022 23:46:51 +0200
Subject: [PATCH] Implement [Purpur-0150] Tick fluids config


diff --git a/src/main/java/net/minecraft/server/BlockFlowing.java b/src/main/java/net/minecraft/server/BlockFlowing.java
index 46516cd84..fc454d3d4 100644
--- a/src/main/java/net/minecraft/server/BlockFlowing.java
+++ b/src/main/java/net/minecraft/server/BlockFlowing.java
@@ -7,6 +7,7 @@ import java.util.Set;
 
 // CraftBukkit start
 import net.titanium.config.SpigotWorldConfig;
+import net.titanium.config.TitaniumConfig;
 import org.bukkit.block.BlockFace;
 import org.bukkit.event.block.BlockFromToEvent;
 // CraftBukkit end
@@ -159,6 +160,11 @@ public class BlockFlowing extends BlockFluids {
     }
 
     private void flow(World world, BlockPosition blockposition, IBlockData iblockdata, int i) {
+        // Titanium start
+        if (!TitaniumConfig.get().gameMechanics.tickFluids) {
+            return;
+        }
+        // Titanium end
         if (world.isLoaded(blockposition) && this.h(world, blockposition, iblockdata)) { // CraftBukkit - add isLoaded check
             if (iblockdata.getBlock() != Blocks.AIR) {
                 if (this.material == Material.LAVA) {
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index b7f7f6b63..c0af6c48d 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -85,6 +85,15 @@ public class TitaniumConfig extends TitaniumConfigSection {
     @Comment("Making this value even bigger could be good idea and have good impact on performance.")
     public int playerTimeStatisticsInterval = 20;
 
+    public GameMechanics gameMechanics = new GameMechanics();
+
+    public static class GameMechanics extends TitaniumConfigSection {
+
+
+        public boolean tickFluids = true;
+
+    }
+
     @Comment("The max NBT size in bytes. The default value is 2097152 which is ~2Mb - it's insane big.")
     @Comment("We are setting this to 50000 which is only ~0.05MB. That's enough in most cases.")
     @Comment("Smaller values (Our 50000 is probably enough, but smaller value would be even better) prevents most NBT related overflow exploits (for eg. Books)")
-- 
2.36.0.windows.1

