From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Wed, 26 Feb 2020 21:39:58 +0100
Subject: [PATCH] Implement [Yatopia-0030] Don't save Fireworks and Arrows by
 tr7zw


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index ff9386da3abc0c8fb0c1653c6c6ddccf8026a95c..bb7c8680979fd42990dbb069058ebfacac886f69 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -1262,6 +1262,11 @@ public abstract class Entity implements ICommandListener {
     public void b(Entity entity, int i) {}
 
     public boolean c(NBTTagCompound nbttagcompound) {
+        // YAPFA start - Don't save Fireworks and Arrows // Titanium - configuration for this
+        if(TitaniumConfig.get().disableFireworksAndArrowsSaving && (this instanceof EntityFireworks || this instanceof EntityArrow)) {
+            return false;
+        }
+        // YAPFA end
         String s = this.ag();
 
         if (!this.dead && s != null) {
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index c63e9fae1e3b84e9722b557cb9a8aa58bc98f673..927def851182a702dee714d85cab843759893c53 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -98,6 +98,9 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
     }
 
+    @Comment("Whether saving of fireworks and arrows should be disabled.")
+    public boolean disableFireworksAndArrowsSaving = false;
+
     @Comment("The max NBT size in bytes. The default value is 2097152 which is ~2Mb - it's insane big.")
     @Comment("We are setting this to 50000 which is only ~0.05MB. That's enough in most cases.")
     @Comment("Smaller values (Our 50000 is probably enough, but smaller value would be even better) prevents most NBT related overflow exploits (for eg. Books)")
