From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: frash23 <jacob@bytesizedpacket.com>
Date: Thu, 24 Mar 2016 04:07:04 +0100
Subject: [PATCH] Implement [Taco-0017] Add option to stop redstone firing
 BlockPhysicsEvent by frash23


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 3c994004c33458ac43891aaa01dfe32289e8849a..2c7564ed4d4f6d10404398551d0a124475ea3569 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -615,7 +615,7 @@ public abstract class World implements IBlockAccess {
             try {
                 // CraftBukkit start
                 CraftWorld world = ((WorldServer) this).getWorld();
-                if (world != null && ((WorldServer) this).hasPhysicsEvent) { // Paper
+                if (world != null && ((WorldServer) this).hasPhysicsEvent && (this.tacoConfigTitanium.redstoneFireBlockPhysicsEvent || !(block instanceof BlockRedstoneWire || block instanceof BlockRedstoneTorch || block instanceof BlockRepeater)))  { // Paper // TacoSpigot - Add config to disable redstone firing BlockPhysicsEvent
                     BlockPhysicsEvent event = new BlockPhysicsEvent(world.getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()), CraftMagicNumbers.getId(block));
                     this.getServer().getPluginManager().callEvent(event);
 
diff --git a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
index d91fa559e70ef79c9dd13f4c9ad80acc6fa972bc..51ba98e578456c89b6abcc3e94958059cd2c9de8 100644
--- a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
@@ -26,6 +26,11 @@ public class TacoSpigotWorldConfig extends TitaniumConfigSection {
 
     }
 
+    @Comment("Whether to fire BlockPhysicsEvent for redstone wires, torches and repeaters.")
+    @Comment("Disabling this could increase performance, but also break some plugins (for eg. cuboids, region protection, etc.).")
+    @CustomKey("redstone-fire-BlockPhysicsEvent")
+    public boolean redstoneFireBlockPhysicsEvent = true;
+
     public Hopper hopper = new Hopper();
 
     public static class Hopper extends TitaniumConfigSection {
@@ -35,7 +40,7 @@ public class TacoSpigotWorldConfig extends TitaniumConfigSection {
         public boolean pushBased = false;
 
         @Comment("Whether to fire InventoryMoveEvent for hoppers.")
-        @Comment("Disabling this could increase performance, but also breaks some plugins (for eg. cuboids, region protection, etc.).")
+        @Comment("Disabling this could increase performance, but also break some plugins (for eg. cuboids, region protection, etc.).")
         @CustomKey("fire-InventoryMoveItemEvent")
         public boolean fireInventoryMoveItemEvent = true;
 
