From ee596a6e2e94b406e2801438c24353ed79b60618 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 24 Jun 2022 22:55:48 +0200
Subject: [PATCH] Add UnknownCommandEvent


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f63b248ce..5e268ca82 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -82,6 +82,7 @@ import net.titanium.config.LegacyConfigHelper;
 import net.titanium.config.TitaniumConfig;
 import net.titanium.config.knockback.KnockbackConfig;
 import net.titanium.config.message.MessageConfig;
+import net.titanium.event.UnknownCommandEvent;
 import net.titanium.knockback.KnockbackProfile;
 import net.titanium.knockback.TitaniumKnockbackProfile;
 import org.apache.commons.lang.Validate;
@@ -740,11 +741,13 @@ public final class CraftServer implements Server {
             return true;
         }
 
-        int spaceIndex = commandLine.indexOf(' ');
-        String commandSubstring = commandLine.substring(0, spaceIndex != -1 ? spaceIndex : commandLine.length());
-
-        sender.sendMessage(java.text.MessageFormat.format(MessageConfig.get().unknownCommand, commandSubstring));
-
+        // Titanium start - UnknownCommandEvent
+        UnknownCommandEvent event = new UnknownCommandEvent(sender, commandLine);
+        this.pluginManager.callEvent(event);
+        if (!event.isCancelled()) {
+            sender.sendMessage(java.text.MessageFormat.format(MessageConfig.get().unknownCommand, event.getCommand())); // Titanium - placeholder of what command was tried to use
+        }
+        // Titanium end
         return false;
     }
 
-- 
2.36.0.windows.1

