From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Sun, 26 Jun 2022 01:10:05 -0500
Subject: [PATCH] Implement [Purpur-0290] Remove Timings by BillyGalbreath


diff --git a/src/main/java/co/aikar/timings/SpigotTimings.java b/src/main/java/co/aikar/timings/SpigotTimings.java
index b3cafafcbe92d1896dfd6925ee611d2ca6120f0a..c2f9c4240c9bbdf00ee8794472b10f76fe9be8d3 100644
--- a/src/main/java/co/aikar/timings/SpigotTimings.java
+++ b/src/main/java/co/aikar/timings/SpigotTimings.java
@@ -6,6 +6,7 @@ import org.bukkit.scheduler.BukkitTask;
 
 import org.bukkit.craftbukkit.scheduler.CraftTask;
 
+@Deprecated // Titanium (Purpur)
 public final class SpigotTimings {
 
     public static final Timing playerListTimer = Timings.ofSafe("Player List");
diff --git a/src/main/java/co/aikar/timings/WorldTimingsHandler.java b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
index 9ebc710fe7cebf71c130c84da2805c4314719b92..68da4638423baa981af677090fb5efd18816a63f 100644
--- a/src/main/java/co/aikar/timings/WorldTimingsHandler.java
+++ b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
@@ -5,6 +5,7 @@ import net.minecraft.server.World;
 /**
  * Set of timers per world, to track world specific timings.
  */
+@Deprecated // Titanium (Purpur)
 public class WorldTimingsHandler {
     public final Timing mobSpawn;
     public final Timing doChunkUnload;
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 1de17cafd0b7696d0024d40962a0e8c0f0aacf1d..58d8c9d78af5fdda9a41fe57ea27f87d3b3d146b 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -1073,7 +1073,7 @@ public class Chunk {
     }
 
     public void loadNearby(IChunkProvider ichunkprovider, IChunkProvider ichunkprovider1, int i, int j) {
-        world.timings.syncChunkLoadPostTimer.startTiming(); // Spigot
+        //world.timings.syncChunkLoadPostTimer.startTiming(); // Spigot // Purpur
         boolean flag = ichunkprovider.isChunkLoaded(i, j - 1);
         boolean flag1 = ichunkprovider.isChunkLoaded(i + 1, j);
         boolean flag2 = ichunkprovider.isChunkLoaded(i, j + 1);
@@ -1120,7 +1120,7 @@ public class Chunk {
             }
         }
 
-        world.timings.syncChunkLoadPostTimer.stopTiming(); // Spigot
+        //world.timings.syncChunkLoadPostTimer.stopTiming(); // Spigot // Purpur
     }
 
     public BlockPosition h(BlockPosition blockposition) {
diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index f56bed369ba2eb0bb8aa4d10ccd28250bdcce33a..a349100994125bb1e3e1f5ca64c329f62588f56f 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -162,7 +162,7 @@ public class ChunkProviderServer implements IChunkProvider {
         // CraftBukkit end
 
         if (chunk == null) {
-            world.timings.syncChunkLoadTimer.startTiming(); // Spigot
+            //world.timings.syncChunkLoadTimer.startTiming(); // Spigot // Purpur
             chunk = this.loadChunk(i, j);
             if (chunk == null) {
                 if (this.chunkProvider == null) {
@@ -214,7 +214,7 @@ public class ChunkProviderServer implements IChunkProvider {
             }
             // CraftBukkit end
             chunk.loadNearby(this, this, i, j);
-            world.timings.syncChunkLoadTimer.stopTiming(); // Spigot
+            //world.timings.syncChunkLoadTimer.stopTiming(); // Spigot // Purpur
         }
 
         unloadQueue.remove(key); // SportPaper
@@ -253,9 +253,9 @@ public class ChunkProviderServer implements IChunkProvider {
                 if (chunk != null) {
                     chunk.setLastSaved(this.world.getTime());
                     if (this.chunkProvider != null) {
-                        world.timings.syncChunkLoadStructuresTimer.startTiming(); // Spigot
+                        //world.timings.syncChunkLoadStructuresTimer.startTiming(); // Spigot // Purpur
                         this.chunkProvider.recreateStructures(chunk, i, j);
-                        world.timings.syncChunkLoadStructuresTimer.stopTiming(); // Spigot
+                        //world.timings.syncChunkLoadStructuresTimer.stopTiming(); // Spigot // Purpur
                     }
                 }
 
diff --git a/src/main/java/net/minecraft/server/ChunkRegionLoader.java b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
index 7889b91e705fcd1b221d0156e08b1dafe28e5eab..32b584440e348e9474d63aa5ad1b4aab41fbb131 100644
--- a/src/main/java/net/minecraft/server/ChunkRegionLoader.java
+++ b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
@@ -44,9 +44,9 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
 
     // CraftBukkit start - Add async variant, provide compatibility
     public Chunk a(World world, int i, int j) throws IOException {
-        world.timings.syncChunkLoadDataTimer.startTiming(); // Spigot
+        //world.timings.syncChunkLoadDataTimer.startTiming(); // Spigot // Purpur
         Object[] data = loadChunk(world, i, j);
-        world.timings.syncChunkLoadDataTimer.stopTiming(); // Spigot
+        //world.timings.syncChunkLoadDataTimer.stopTiming(); // Spigot // Purpur
         if (data != null) {
             Chunk chunk = (Chunk) data[0];
             NBTTagCompound nbttagcompound = (NBTTagCompound) data[1];
@@ -396,7 +396,7 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
 
     public void loadEntities(Chunk chunk, NBTTagCompound nbttagcompound, World world) {
         // CraftBukkit end
-        world.timings.syncChunkLoadEntitiesTimer.startTiming(); // Spigot
+        //world.timings.syncChunkLoadEntitiesTimer.startTiming(); // Spigot // Purpur
         NBTTagList nbttaglist1 = nbttagcompound.getList("Entities", 10);
 
         if (nbttaglist1 != null) {
@@ -422,8 +422,8 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
                 }
             }
         }
-        world.timings.syncChunkLoadEntitiesTimer.stopTiming(); // Spigot
-        world.timings.syncChunkLoadTileEntitiesTimer.startTiming(); // Spigot
+        //world.timings.syncChunkLoadEntitiesTimer.stopTiming(); // Spigot // Purpur
+        //world.timings.syncChunkLoadTileEntitiesTimer.startTiming(); // Spigot // Purpur
         NBTTagList nbttaglist2 = nbttagcompound.getList("TileEntities", 10);
 
         if (nbttaglist2 != null) {
@@ -436,8 +436,8 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
                 }
             }
         }
-        world.timings.syncChunkLoadTileEntitiesTimer.stopTiming(); // Spigot
-        world.timings.syncChunkLoadTileTicksTimer.startTiming(); // Spigot
+        //world.timings.syncChunkLoadTileEntitiesTimer.stopTiming(); // Spigot // Purpur
+        //world.timings.syncChunkLoadTileTicksTimer.startTiming(); // Spigot // Purpur
 
         if (nbttagcompound.hasKeyOfType("TileTicks", 9)) {
             NBTTagList nbttaglist3 = nbttagcompound.getList("TileTicks", 10);
@@ -457,7 +457,7 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
                 }
             }
         }
-        world.timings.syncChunkLoadTileTicksTimer.stopTiming(); // Spigot
+        //world.timings.syncChunkLoadTileTicksTimer.stopTiming(); // Spigot // Purpur
 
         // return chunk; // CraftBukkit
     }
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 7a44b757c34bb337812aeb7e9f6dd378d1a77486..45afb38bb7cd72014e6d50510feaf041be69e5a8 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -437,7 +437,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
     }
 
     public void aO() {
-        SpigotTimings.serverCommandTimer.startTiming(); // Spigot
+        //SpigotTimings.serverCommandTimer.startTiming(); // Spigot // Purpur
         // [Paper-0350] start - use a Queue for Queueing Commands
         ServerCommand servercommand;
         while ((servercommand = this.l.poll()) != null) {
@@ -454,7 +454,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
             // CraftBukkit end
         }
 
-        SpigotTimings.serverCommandTimer.stopTiming(); // Spigot
+        //SpigotTimings.serverCommandTimer.stopTiming(); // Spigot // Purpur
     }
 
     public boolean ae() {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 55b1cb38e3b670f01c984b95756d792bc3cd1c48..a814310c38aca1b6502582fe27e419534c1d4333 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -832,7 +832,7 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
 
     // WindSpigot - backport modern tick loop
     protected void A(BooleanSupplier shouldKeepTicking) throws ExceptionWorldConflict { // CraftBukkit - added throws
-        co.aikar.timings.TimingsManager.FULL_SERVER_TICK.startTiming(); // Spigot
+        //co.aikar.timings.TimingsManager.FULL_SERVER_TICK.startTiming(); // Spigot // Purpur
 
         // WindSpigot start - backport modern tick loop
         long i = getNanos();
@@ -870,7 +870,7 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
         }
 
         if (autosavePeriod > 0 && this.ticks % autosavePeriod == 0) { // CraftBukkit
-            SpigotTimings.worldSaveTimer.startTiming(); // Spigot
+            // SpigotTimings.worldSaveTimer.startTiming(); // Spigot // Purpur
             this.methodProfiler.a("save");
             this.v.savePlayers();
             // Spigot Start
@@ -885,7 +885,7 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
             // this.saveChunks(true);
             // Spigot End
             this.methodProfiler.b();
-            SpigotTimings.worldSaveTimer.stopTiming(); // Spigot
+            // SpigotTimings.worldSaveTimer.stopTiming(); // Spigot // Purpur
         }
 
         // WindSpigot start - backport modern tick loop
@@ -912,11 +912,11 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
         this.methodProfiler.b();
         this.methodProfiler.b();
         org.spigotmc.WatchdogThread.tick(); // Spigot
-        co.aikar.timings.TimingsManager.FULL_SERVER_TICK.stopTiming(); // Spigot
+        //co.aikar.timings.TimingsManager.FULL_SERVER_TICK.stopTiming(); // Spigot // Purpur
     }
 
     public void B() {
-        SpigotTimings.minecraftSchedulerTimer.startTiming(); // Spigot
+        //SpigotTimings.minecraftSchedulerTimer.startTiming(); // Spigot // Purpur
         this.methodProfiler.a("jobs");
         Queue queue = this.j;
 
@@ -927,27 +927,27 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
             SystemUtils.a(entry, MinecraftServer.LOGGER);
          }
         // Spigot end
-        SpigotTimings.minecraftSchedulerTimer.stopTiming(); // Spigot
+        //SpigotTimings.minecraftSchedulerTimer.stopTiming(); // Spigot // Purpur
 
         this.methodProfiler.c("levels");
 
-        SpigotTimings.bukkitSchedulerTimer.startTiming(); // Spigot
+        //SpigotTimings.bukkitSchedulerTimer.startTiming(); // Spigot // Purpur
         // CraftBukkit start
         this.server.getScheduler().mainThreadHeartbeat(this.ticks);
-        SpigotTimings.bukkitSchedulerTimer.stopTiming(); // Spigot
+        //SpigotTimings.bukkitSchedulerTimer.stopTiming(); // Spigot // Purpur
 
         // Run tasks that are waiting on processing
-        SpigotTimings.processQueueTimer.startTiming(); // Spigot
+        //SpigotTimings.processQueueTimer.startTiming(); // Spigot // Purpur
         while (!processQueue.isEmpty()) {
             processQueue.remove().run();
         }
-        SpigotTimings.processQueueTimer.stopTiming(); // Spigot
+        //SpigotTimings.processQueueTimer.stopTiming(); // Spigot // Purpur
 
-        SpigotTimings.chunkIOTickTimer.startTiming(); // Spigot
+        //SpigotTimings.chunkIOTickTimer.startTiming(); // Spigot // Purpur
         org.bukkit.craftbukkit.chunkio.ChunkIOExecutor.tick();
-        SpigotTimings.chunkIOTickTimer.stopTiming(); // Spigot
+        //SpigotTimings.chunkIOTickTimer.stopTiming(); // Spigot // Purpur
 
-        SpigotTimings.timeUpdateTimer.startTiming(); // Spigot
+        //SpigotTimings.timeUpdateTimer.startTiming(); // Spigot // Purpur
         // Send time updates to everyone, it will get the right time from the world the player is in.
         // Paper start - optimize time updates
         int i;
@@ -976,7 +976,7 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
                 }
             }
         }
-        SpigotTimings.timeUpdateTimer.stopTiming(); // Spigot
+        //SpigotTimings.timeUpdateTimer.stopTiming(); // Spigot // Purpur
 
         for (i = 0; i < this.worlds.size(); ++i) {
             long j = System.nanoTime();
@@ -999,9 +999,9 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
                 CrashReport crashreport;
 
                 try {
-                    worldserver.timings.doTick.startTiming(); // Spigot
+                    //worldserver.timings.doTick.startTiming(); // Spigot // Purpur
                     worldserver.doTick();
-                    worldserver.timings.doTick.stopTiming(); // Spigot
+                    //worldserver.timings.doTick.stopTiming(); // Spigot // Purpur
                 } catch (Throwable throwable) {
                     // Spigot Start
                     try {
@@ -1015,9 +1015,9 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
                 }
 
                 try {
-                    worldserver.timings.tickEntities.startTiming(); // Spigot
+                    //worldserver.timings.tickEntities.startTiming(); // Spigot // Purpur
                     worldserver.tickEntities();
-                    worldserver.timings.tickEntities.stopTiming(); // Spigot
+                    //worldserver.timings.tickEntities.stopTiming(); // Spigot // Purpur
                 } catch (Throwable throwable1) {
                     // Spigot Start
                     try {
@@ -1032,7 +1032,7 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
 
                 this.methodProfiler.b();
                 this.methodProfiler.a("tracker");
-                worldserver.timings.tracker.startTiming(); // Spigot
+                //worldserver.timings.tracker.startTiming(); // Spigot // Purpur
                 // Tuinity start - controlled flush for entity tracker packets
                 List<NetworkManager> disabledFlushes = new ArrayList<>(worldserver.players.size());
                 for (EntityHuman player : worldserver.players) {
@@ -1059,7 +1059,7 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
                     }
                 }
                 // Tuinity end
-                worldserver.timings.tracker.stopTiming(); // Spigot
+                //worldserver.timings.tracker.stopTiming(); // Spigot // Purpur
                 this.methodProfiler.b();
                 this.methodProfiler.b();
                 worldserver.explosionDensityCache.clear(); // PaperSpigot - Optimize explosions
@@ -1076,20 +1076,20 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
         // WindSpigot end
 
         this.methodProfiler.c("connection");
-        SpigotTimings.connectionTimer.startTiming(); // Spigot
+        //SpigotTimings.connectionTimer.startTiming(); // Spigot // Purpur
         this.aq().c();
-        SpigotTimings.connectionTimer.stopTiming(); // Spigot
+        //SpigotTimings.connectionTimer.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("players");
-        SpigotTimings.playerListTimer.startTiming(); // Spigot
+        //SpigotTimings.playerListTimer.startTiming(); // Spigot // Purpur
         this.v.tick();
-        SpigotTimings.playerListTimer.stopTiming(); // Spigot
+        //SpigotTimings.playerListTimer.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("tickables");
 
-        SpigotTimings.tickablesTimer.startTiming(); // Spigot
+        //SpigotTimings.tickablesTimer.startTiming(); // Spigot // Purpur
         for (i = 0; i < this.p.size(); ++i) {
             ((IUpdatePlayerListBox) this.p.get(i)).c();
         }
-        SpigotTimings.tickablesTimer.stopTiming(); // Spigot
+        //SpigotTimings.tickablesTimer.stopTiming(); // Spigot // Purpur
 
         this.methodProfiler.b();
     }
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 74e0fd86e0c707f73e54c7e2bcfa5fa2eb7a7cf7..a5717b5be54bbc2381f5493d50b64f9234645aa6 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1189,7 +1189,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
     // CraftBukkit end
 
    private void handleCommand(String s) {
-        SpigotTimings.playerCommandTimer.startTiming(); // Spigot
+        //SpigotTimings.playerCommandTimer.startTiming(); // Spigot // Purpur
        // CraftBukkit start - whole method
         if ( TitaniumConfig.get().spigot.commands.log ) // Spigot
         this.c.info(this.player.getName() + " issued server command: " + s);
@@ -1200,22 +1200,22 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
         this.server.getPluginManager().callEvent(event);
 
         if (event.isCancelled()) {
-            SpigotTimings.playerCommandTimer.stopTiming(); // Spigot
+            //SpigotTimings.playerCommandTimer.stopTiming(); // Spigot // Purpur
             return;
         }
 
         try {
             if (this.server.dispatchCommand(event.getPlayer(), event.getMessage().substring(1))) {
-                SpigotTimings.playerCommandTimer.stopTiming(); // Spigot
+                //SpigotTimings.playerCommandTimer.stopTiming(); // Spigot // Purpur
                 return;
             }
         } catch (org.bukkit.command.CommandException ex) {
             player.sendMessage(server.messageConfig.internalErrorCommand);
             java.util.logging.Logger.getLogger(PlayerConnection.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
-            SpigotTimings.playerCommandTimer.stopTiming(); // Spigot
+            //SpigotTimings.playerCommandTimer.stopTiming(); // Spigot // Purpur
             return;
         }
-        SpigotTimings.playerCommandTimer.stopTiming(); // Spigot
+        //SpigotTimings.playerCommandTimer.stopTiming(); // Spigot // Purpur
         // this.minecraftServer.getCommandHandler().a(this.player, s);
         // CraftBukkit end
     }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 2c7564ed4d4f6d10404398551d0a124475ea3569..4180406ed35307230417db04a853724e7d8d67e4 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -184,6 +184,7 @@ public abstract class World implements IBlockAccess {
     public final IonWorldConfig ionConfigTitanium;
     // Titanium end
 
+    @Deprecated
     public final co.aikar.timings.WorldTimingsHandler timings; // Spigot
 
     public CraftWorld getWorld() {
@@ -1502,7 +1503,7 @@ public abstract class World implements IBlockAccess {
         }
 
         this.methodProfiler.c("remove");
-        timings.entityRemoval.startTiming(); // Spigot
+        //timings.entityRemoval.startTiming(); // Spigot // Purpur
         this.entityList.removeAll(this.g);
 
         int j;
@@ -1523,14 +1524,14 @@ public abstract class World implements IBlockAccess {
         // Paper end
 
         this.g.clear();
-        timings.entityRemoval.stopTiming(); // Spigot
+        //timings.entityRemoval.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("regular");
 
         org.spigotmc.ActivationRange.activateEntities(this); // Spigot
-        timings.entityTick.startTiming(); // Spigot
+        //timings.entityTick.startTiming(); // Spigot // Purpur
         guardEntityList = true; // Spigot
         // CraftBukkit start - Use field for loop variable
-        co.aikar.timings.TimingHistory.entityTicks += this.entityList.size(); // Spigot
+        //co.aikar.timings.TimingHistory.entityTicks += this.entityList.size(); // Spigot // Purpur
         int entitiesThisCycle = 0;
         // PaperSpigot start - Disable tick limiters
         //if (tickPosition < 0) tickPosition = 0;
@@ -1551,12 +1552,12 @@ public abstract class World implements IBlockAccess {
             this.methodProfiler.a("tick");
             if (!entity.dead) {
                 try {
-                    entity.tickTimer.startTiming(); // Spigot
+                    //entity.tickTimer.startTiming(); // Spigot // Purpur
                     this.g(entity);
-                    entity.tickTimer.stopTiming(); // Spigot
+                    //entity.tickTimer.stopTiming(); // Spigot // Purpur
                 } catch (Throwable throwable1) {
                     // PaperSpigot start - Prevent tile entity and entity crashes
-                    entity.tickTimer.stopTiming();
+                    //entity.tickTimer.stopTiming(); // Purpur
                     String msg = "Entity threw exception at " + entity.world.getWorld().getName() + ":" + entity.locX + "," + entity.locY + "," + entity.locZ;
                     System.err.println(msg);
                     throwable1.printStackTrace();
@@ -1586,9 +1587,9 @@ public abstract class World implements IBlockAccess {
         }
         guardEntityList = false; // Spigot
 
-        timings.entityTick.stopTiming(); // Spigot
+        //timings.entityTick.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("blockEntities");
-        timings.tileEntityTick.startTiming(); // Spigot
+        //timings.tileEntityTick.startTiming(); // Spigot // Purpur
         this.M = true;
         // CraftBukkit start - From below, clean up tile entities before ticking them
         if (!this.c.isEmpty()) {
@@ -1617,11 +1618,11 @@ public abstract class World implements IBlockAccess {
 
                 if (this.isLoaded(blockposition) && this.N.a(blockposition)) {
                     try {
-                        tileentity.tickTimer.startTiming(); // Spigot
+                        //tileentity.tickTimer.startTiming(); // Spigot // Purpur
                         ((IUpdatePlayerListBox) tileentity).c();
                     } catch (Throwable throwable2) {
                         // PaperSpigot start - Prevent tile entity and entity crashes
-                        tileentity.tickTimer.stopTiming();
+                        //tileentity.tickTimer.stopTiming(); // Purpur
                         String msg = "TileEntity threw exception at " + tileentity.world.getWorld().getName() + ":" + tileentity.position.getX() + "," + tileentity.position.getY() + "," + tileentity.position.getZ();
                         System.err.println(msg);
                         throwable2.printStackTrace();
@@ -1633,7 +1634,7 @@ public abstract class World implements IBlockAccess {
                     }
                     // Spigot start
                     finally {
-                        tileentity.tickTimer.stopTiming();
+                        //tileentity.tickTimer.stopTiming(); // Purpur
                     }
                     // Spigot end
                 }
@@ -1649,8 +1650,8 @@ public abstract class World implements IBlockAccess {
             }
         }
 
-        timings.tileEntityTick.stopTiming(); // Spigot
-        timings.tileEntityPending.startTiming(); // Spigot
+        //timings.tileEntityTick.stopTiming(); // Spigot // Purpur
+        //timings.tileEntityPending.startTiming(); // Spigot // Purpur
         this.M = false;
         /* CraftBukkit start - Moved up
         if (!this.c.isEmpty()) {
@@ -1683,11 +1684,11 @@ public abstract class World implements IBlockAccess {
             this.b.clear();
         }
 
-        timings.tileEntityPending.stopTiming(); // Spigot
-        co.aikar.timings.TimingHistory.tileEntityTicks += this.tileEntityList.size(); // Spigot
+        //timings.tileEntityPending.stopTiming(); // Spigot // Purpur
+        //co.aikar.timings.TimingHistory.tileEntityTicks += this.tileEntityList.size(); // Spigot // Purpur
 
-        this.methodProfiler.b();
-        this.methodProfiler.b();
+        //this.methodProfiler.b(); // Purpur
+        //this.methodProfiler.b(); // Purpur
     }
 
     public boolean a(TileEntity tileentity) {
@@ -1748,7 +1749,7 @@ public abstract class World implements IBlockAccess {
             entity.lastPitch = entity.pitch;
             if (flag && entity.ad) {
                 ++entity.ticksLived;
-                ++co.aikar.timings.TimingHistory.activatedEntityTicks; // Spigot
+                //++co.aikar.timings.TimingHistory.activatedEntityTicks; // Spigot // Purpur
                 if (entity.vehicle != null) {
                     entity.ak();
                 } else {
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index b504a01f9ada3104604e79a661b6d56247485892..758ca7265ffafe5727091206305d51724de452cc 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -238,13 +238,13 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         // CraftBukkit start - Only call spawner if we have players online and the world allows for mobs or animals
         long time = this.worldData.getTime();
         if (this.getGameRules().getBoolean("doMobSpawning") && this.worldData.getType() != WorldType.DEBUG_ALL_BLOCK_STATES && (this.allowMonsters || this.allowAnimals) && (this instanceof WorldServer && this.players.size() > 0)) {
-            timings.mobSpawn.startTiming(); // Spigot
+            //timings.mobSpawn.startTiming(); // Spigot // Purpur
             this.R.a(this, this.allowMonsters && (this.ticksPerMonsterSpawns != 0 && time % this.ticksPerMonsterSpawns == 0L), this.allowAnimals && (this.ticksPerAnimalSpawns != 0 && time % this.ticksPerAnimalSpawns == 0L), this.worldData.getTime() % 400L == 0L);
-            timings.mobSpawn.stopTiming(); // Spigot
+            //timings.mobSpawn.stopTiming(); // Spigot // Purpur
             // CraftBukkit end
         }
         // CraftBukkit end
-        timings.doChunkUnload.startTiming(); // Spigot
+        //timings.doChunkUnload.startTiming(); // Spigot // Purpur
         this.methodProfiler.c("chunkSource");
 
         // Only unload if chunkProvider isn't null
@@ -263,35 +263,35 @@ public class WorldServer extends World implements IAsyncTaskHandler {
             this.worldData.setDayTime(this.worldData.getDayTime() + 1L);
         }
 
-        timings.doChunkUnload.stopTiming(); // Spigot
+        //timings.doChunkUnload.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("tickPending");
-        timings.scheduledBlocks.startTiming(); // Spigot
+        //timings.scheduledBlocks.startTiming(); // Spigot // Purpur
         this.a(false);
-        timings.scheduledBlocks.stopTiming(); // Spigot
+        //timings.scheduledBlocks.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("tickBlocks");
-        timings.chunkTicks.startTiming(); // Spigot
+        //timings.chunkTicks.startTiming(); // Spigot // Purpur
         this.h();
-        timings.chunkTicks.stopTiming(); // Spigot
+        //timings.chunkTicks.stopTiming(); // Spigot // Purpur
         SpigotWorldConfig.getAntiXray().flushUpdates(this); // PaperSpigot
         this.methodProfiler.c("chunkMap");
-        timings.doChunkMap.startTiming(); // Spigot
+        //timings.doChunkMap.startTiming(); // Spigot // Purpur
         this.manager.flush();
-        timings.doChunkMap.stopTiming(); // Spigot
+        //timings.doChunkMap.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("village");
-        timings.doVillages.startTiming(); // Spigot
+        //timings.doVillages.startTiming(); // Spigot // Purpur
         this.villages.tick();
         this.siegeManager.a();
-        timings.doVillages.stopTiming(); // Spigot
+        //timings.doVillages.stopTiming(); // Spigot // Purpur
         this.methodProfiler.c("portalForcer");
-        timings.doPortalForcer.startTiming(); // Spigot
+        //timings.doPortalForcer.startTiming(); // Spigot // Purpur
         this.Q.a(this.getTime());
-        timings.doPortalForcer.stopTiming(); // Spigot
+        //timings.doPortalForcer.stopTiming(); // Spigot // Purpur
         this.methodProfiler.b();
-        timings.doSounds.startTiming(); // Spigot
+        //timings.doSounds.startTiming(); // Spigot // Purpur
         this.ak();
 
         this.getWorld().processChunkGC(); // CraftBukkit
-        timings.doChunkGC.stopTiming(); // Spigot
+        //timings.doChunkGC.stopTiming(); // Spigot // Purpur
     }
 
     public BiomeBase.BiomeMeta a(EnumCreatureType enumcreaturetype, BlockPosition blockposition) {
@@ -491,7 +491,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 }
 
                 this.methodProfiler.c("tickBlocks");
-                timings.chunkTicksBlocks.startTiming(); // Spigot
+                //timings.chunkTicksBlocks.startTiming(); // Spigot // Purpur
                 i1 = titaniumConfig.disabledGamerules.randomTickSpeed ? 3 : this.getGameRules().c("randomTickSpeed");
                 if (i1 > 0) {
                     ChunkSection[] achunksection = chunk.getSections();
@@ -520,7 +520,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                         }
                     }
                 }
-                timings.chunkTicksBlocks.stopTiming(); // Spigot
+                //timings.chunkTicksBlocks.stopTiming(); // Spigot // Purpur
             }
 
         }
@@ -652,7 +652,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
                 this.methodProfiler.a("cleaning");
 
-                timings.scheduledBlocksCleanup.startTiming(); // Spigot
+                //timings.scheduledBlocksCleanup.startTiming(); // Spigot // Purpur
                 NextTickListEntry nextticklistentry;
 
                 for (int j = 0; j < i; ++j) {
@@ -665,7 +665,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                     this.M.remove(nextticklistentry);
                     this.V.add(nextticklistentry);
                 }
-                timings.scheduledBlocksCleanup.stopTiming(); // Spigot
+                //timings.scheduledBlocksCleanup.stopTiming(); // Spigot // Purpur
 
                 // PaperSpigot start - Allow redstone ticks to bypass the tickNextTickListCap
                 if (paperConfigTitanium.tickNextTickListCapIgnoresRedstone) {
@@ -686,7 +686,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
                 this.methodProfiler.b();
                 this.methodProfiler.a("ticking");
-                timings.scheduledBlocksTicking.startTiming(); // Spigot
+                //timings.scheduledBlocksTicking.startTiming(); // Spigot // Purpur
                 Iterator iterator = this.V.iterator();
 
                 while (iterator.hasNext()) {
@@ -697,7 +697,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                     if (this.areChunksLoadedBetween(nextticklistentry.a.a(-b0, -b0, -b0), nextticklistentry.a.a(b0, b0, b0))) {
                         IBlockData iblockdata = this.getType(nextticklistentry.a);
                         co.aikar.timings.Timing timing = iblockdata.getBlock().getTiming(); // Spigot
-                        timing.startTiming(); // Spigot
+                        //timing.startTiming(); // Spigot // Purpur
 
                         if (iblockdata.getBlock().getMaterial() != Material.AIR && Block.a(iblockdata.getBlock(), nextticklistentry.a())) {
                             try {
@@ -710,12 +710,12 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                                 throw new ReportedException(crashreport);
                             }
                         }
-                        timing.stopTiming(); // Spigot
+                        //timing.stopTiming(); // Spigot // Purpur
                     } else {
                         this.a(nextticklistentry.a, nextticklistentry.a(), 0);
                     }
                 }
-                timings.scheduledBlocksTicking.stopTiming(); // Spigot
+                //timings.scheduledBlocksTicking.stopTiming(); // Spigot // Purpur
 
                 this.methodProfiler.b();
                 this.V.clear();
diff --git a/src/main/java/net/titanium/config/SpigotConfig.java b/src/main/java/net/titanium/config/SpigotConfig.java
index dd47b50c567b8dd12938623b8fb9d7192ee26dc3..a9dddef6070939893b6a4dad5e3e2abdc4f65e41 100644
--- a/src/main/java/net/titanium/config/SpigotConfig.java
+++ b/src/main/java/net/titanium/config/SpigotConfig.java
@@ -161,49 +161,6 @@ public class SpigotConfig extends TitaniumConfigSection {
 
     }
 
-    @Comment("Various settings related to server timings and debugging.")
-    public Timings timings = new Timings();
-
-    public static class Timings extends TitaniumConfigSection {
-
-        public boolean enabled = true;
-        public boolean verbose = true;
-        public boolean serverNamePrivacy = false;
-        public List<String> hiddenConfigEntries = Arrays.asList("database", "bungeecord-addresses");
-        public int historyInterval = 300;
-        public int historyLength = 3600;
-
-        @Override
-        public void loadProcessedProperties() {
-            TimingsManager.privacy = serverNamePrivacy;
-            TimingsManager.hiddenConfigs = hiddenConfigEntries;
-
-            co.aikar.timings.Timings.setTimingsEnabled(enabled);
-            co.aikar.timings.Timings.setVerboseTimingsEnabled(verbose);
-            co.aikar.timings.Timings.setHistoryInterval(historyInterval * 20);
-            co.aikar.timings.Timings.setHistoryLength(historyLength * 20);
-
-            log(Level.INFO, "Spigot Timings: " + enabled +
-                    " - Verbose: " + verbose +
-                    " - Interval: " + timeSummary(co.aikar.timings.Timings.getHistoryInterval() / 20) +
-                    " - Length: " + timeSummary(co.aikar.timings.Timings.getHistoryLength() / 20));
-        }
-
-        private static String timeSummary(int seconds) {
-            String time = "";
-            if (seconds > 60 * 60) {
-                time += TimeUnit.SECONDS.toHours(seconds) + "h";
-                seconds /= 60;
-            }
-
-            if (seconds > 0) {
-                time += TimeUnit.SECONDS.toMinutes(seconds) + "m";
-            }
-            return time;
-        }
-
-    }
-
     public Stats stats = new Stats();
 
     public static class Stats extends TitaniumConfigSection {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 19cdb459e3f3474294bdc006b0370d58fa5df9a3..e050c4a7504bf61ec9e180bfcc82632f3202f6c1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -286,11 +286,11 @@ public class CraftWorld implements World {
         net.minecraft.server.Chunk chunk = world.chunkProviderServer.chunks.get(key);
 
         if (chunk == null) {
-            world.timings.syncChunkLoadTimer.startTiming(); // Spigot
+            //world.timings.syncChunkLoadTimer.startTiming(); // Spigot // Purpur
             chunk = world.chunkProviderServer.loadChunk(x, z);
 
             chunkLoadPostProcess(chunk, x, z);
-            world.timings.syncChunkLoadTimer.stopTiming(); // Spigot
+            //world.timings.syncChunkLoadTimer.stopTiming(); // Spigot // Purpur
         }
         world.chunkProviderServer.unloadQueue.remove(key);
         return chunk != null;
diff --git a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOProvider.java b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOProvider.java
index 1178ad7f7b7eda10917a3f92b400c788543d0312..c7e2b561f29fc38d10e5a3bd4447fb5d882bd685 100644
--- a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOProvider.java
+++ b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOProvider.java
@@ -49,9 +49,9 @@ class ChunkIOProvider implements AsynchronousExecutor.CallBackProvider<QueuedChu
         chunk.addEntities();
 
         if (queuedChunk.provider.chunkProvider != null) {
-            queuedChunk.provider.world.timings.syncChunkLoadStructuresTimer.startTiming(); // Spigot
+            //queuedChunk.provider.world.timings.syncChunkLoadStructuresTimer.startTiming(); // Spigot // Purpur
             queuedChunk.provider.chunkProvider.recreateStructures(chunk, queuedChunk.x, queuedChunk.z);
-            queuedChunk.provider.world.timings.syncChunkLoadStructuresTimer.stopTiming(); // Spigot
+            //queuedChunk.provider.world.timings.syncChunkLoadStructuresTimer.stopTiming(); // Spigot // Purpur
         }
 
         Server server = queuedChunk.provider.world.getServer();
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
index 44cbcdac82d11f23986676b3ca29c05e98ec80c3..8aec605a0979ffc26143874bcc2922e9ab552a17 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
@@ -60,9 +60,9 @@ public class CraftTask implements BukkitTask, Runnable { // Spigot
     }
 
     public void run() {
-        if (timings != null && isSync()) timings.startTiming(); // Spigot
+        //if (timings != null && isSync()) timings.startTiming(); // Spigot // Purpur
         task.run();
-        if (timings != null && isSync()) timings.stopTiming(); // Spigot
+        //if (timings != null && isSync()) timings.stopTiming(); // Spigot // Purpur
     }
 
     long getPeriod() {
diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index 3f10502c5c77de15777bafd5107b28c37a0da2a2..8b5b679d61b7105edb8e2cb4035fbc5b1fab2073 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -101,7 +101,7 @@ public class ActivationRange
      */
     public static void activateEntities(World world)
     {
-        SpigotTimings.entityActivationCheckTimer.startTiming();
+        //SpigotTimings.entityActivationCheckTimer.startTiming(); // Purpur
         final int miscActivationRange = world.spigotConfigTitanium.entityActivationRange.ambient;
         final int animalActivationRange = world.spigotConfigTitanium.entityActivationRange.animals;
         final int monsterActivationRange = world.spigotConfigTitanium.entityActivationRange.monsters;
@@ -135,7 +135,7 @@ public class ActivationRange
                 }
             }
         }
-        SpigotTimings.entityActivationCheckTimer.stopTiming();
+        //SpigotTimings.entityActivationCheckTimer.stopTiming(); // Purpur
     }
 
     /**
@@ -250,10 +250,10 @@ public class ActivationRange
      */
     public static boolean checkIfActive(Entity entity)
     {
-        SpigotTimings.checkIfActiveTimer.startTiming();
+        //SpigotTimings.checkIfActiveTimer.startTiming(); // Purpur
         // Never safe to skip fireworks or entities not yet added to chunk
         if ( !entity.isAddedToChunk() || entity instanceof EntityFireworks || entity.loadChunks ) { // PaperSpigot
-            SpigotTimings.checkIfActiveTimer.stopTiming();
+            //SpigotTimings.checkIfActiveTimer.stopTiming(); // Purpur
             return true;
         }
 
@@ -285,7 +285,7 @@ public class ActivationRange
         {
             isActive = false;
         }
-        SpigotTimings.checkIfActiveTimer.stopTiming();
+        //SpigotTimings.checkIfActiveTimer.stopTiming(); // Purpur
         return isActive;
     }
 }
diff --git a/src/main/java/org/spigotmc/AntiXray.java b/src/main/java/org/spigotmc/AntiXray.java
index da156330971fa2d86fcb3d4490b5300ba149dc7c..3d9021bc8083af5092a78000ad7dabc9e869abb0 100644
--- a/src/main/java/org/spigotmc/AntiXray.java
+++ b/src/main/java/org/spigotmc/AntiXray.java
@@ -107,9 +107,9 @@ public class AntiXray
                 return;
             }
             // PaperSpigot end
-            SpigotTimings.antiXrayUpdateTimer.startTiming();
+            //SpigotTimings.antiXrayUpdateTimer.startTiming(); // Purpur
             updateNearbyBlocks( world, position, 2, false ); // 2 is the radius, we shouldn't change it as that would make it exponentially slower
-            SpigotTimings.antiXrayUpdateTimer.stopTiming();
+            //SpigotTimings.antiXrayUpdateTimer.stopTiming(); // Purpur
         }
     }
 
@@ -121,9 +121,9 @@ public class AntiXray
     {
         if ( world.spigotConfigTitanium.antiXray.enabled )
         {
-            SpigotTimings.antiXrayObfuscateTimer.startTiming();
+            //SpigotTimings.antiXrayObfuscateTimer.startTiming(); // Purpur
             obfuscate( chunkX, chunkY, bitmask, buffer, world );
-            SpigotTimings.antiXrayObfuscateTimer.stopTiming();
+            //SpigotTimings.antiXrayObfuscateTimer.stopTiming(); // Purpur
         }
     }
 
