From 2826efc5c6daa27c815a0bbca0dac15132e6d945 Mon Sep 17 00:00:00 2001
From: "tsao-chi@the-lingo.org" <tsao-chi@the-lingo.org>
Date: Thu, 2 Apr 2020 11:16:18 +0800
Subject: [PATCH] Implement [Akarin-0010] Save Json list asynchronously


diff --git a/src/main/java/net/minecraft/server/JsonList.java b/src/main/java/net/minecraft/server/JsonList.java
index 8e4ecfd64..fd371f03a 100644
--- a/src/main/java/net/minecraft/server/JsonList.java
+++ b/src/main/java/net/minecraft/server/JsonList.java
@@ -13,6 +13,7 @@ import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
 import com.google.gson.JsonSerializationContext;
 import com.google.gson.JsonSerializer;
+import ga.windpvp.windspigot.async.AsyncUtil;
 import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.File;
@@ -150,17 +151,21 @@ public class JsonList<K, V extends JsonListEntry<K>> {
     }
 
     public void save() throws IOException {
-        Collection collection = this.d.values();
-        String s = this.b.toJson(collection);
-        BufferedWriter bufferedwriter = null;
-
-        try {
-            bufferedwriter = Files.newWriter(this.c, Charsets.UTF_8);
-            bufferedwriter.write(s);
-        } finally {
-            IOUtils.closeQuietly(bufferedwriter);
-        }
-
+        Runnable runnable = () -> { // Akarin - Save json list async
+            Collection collection = this.d.values();
+            String s = this.b.toJson(collection);
+            BufferedWriter bufferedwriter = null;
+
+            try {
+                bufferedwriter = Files.newWriter(this.c, Charsets.UTF_8);
+                bufferedwriter.write(s);
+            } catch (IOException ex) { // Akarin
+                JsonList.a.warn("Failed to save " + this.c, ex); // Akarin
+            } finally {
+                IOUtils.closeQuietly(bufferedwriter);
+            }
+        }; // Akarin
+        AsyncUtil.run(runnable); // Akarin
     }
 
     public void load() throws FileNotFoundException {
-- 
2.36.0.windows.1

