From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 4 Aug 2022 00:34:23 +0200
Subject: [PATCH] Implement [PaperBin-????] Optimise
 WorldServer#everyoneDeeplySleeping


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index c70eaac4070662b7b73bd95575d8fd0e183ec900..f9274a7b0da001f7f92f10d91120b8537213fcc5 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -291,6 +291,16 @@ public abstract class Entity implements ICommandListener {
         this.K();
     }
 
+    // Titanium start
+    public boolean isNether() {
+        return this.world.getWorld().getEnvironment() == org.bukkit.World.Environment.NETHER;
+    }
+
+    public boolean isEnd() {
+        return this.world.getWorld().getEnvironment() == org.bukkit.World.Environment.THE_END;
+    }
+    // Titanium end
+
     /**
      * PaperSpigot - Checks if the feature is enabled and the entity is above the nether world bedrock height
      */
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index bf80c4644bc9f6f22a35c7e9418251ac7f486808..93eb7e7287ea602dd8d5f34b946777972f34fe5c 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -356,7 +356,25 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     }
 
     public boolean everyoneDeeplySleeping() {
-        if (this.O && !this.isClientSide) {
+        // PaperBin start - WorldServer#everyoneDeeplySleeping optimization
+        if (this.players.isEmpty() || this.isClientSide || !this.O) {
+            return false;
+        }
+
+        for (EntityHuman human : this.players) {
+            if (!human.isSpectator() && !human.isDeeplySleeping() && !human.fauxSleeping) {
+                return false;
+            }
+
+            // Titanium start - ignore players who are not in overworld
+            if (!human.isNether() && !human.isEnd()) {
+                return false;
+            }
+            // Titanium end
+        }
+
+        return true;
+        /*if (this.O && !this.isClientSide) {
             Iterator iterator = this.players.iterator();
 
             // CraftBukkit - This allows us to assume that some people are in bed but not really, allowing time to pass in spite of AFKers
@@ -381,7 +399,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
             return false;
         } else {
             return false;
-        }
+        }*/
+        // PaperBin end
     }
 
     protected void h() {
