From e75f80503fffe99e68f2e6deec7fbde2ac70b2e1 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 4 Aug 2022 00:34:23 +0200
Subject: [PATCH] Implement [PaperBin-????] Optimise
 WorldServer#everyoneDeeplySleeping


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 111a25632..247c0ae3a 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -1781,6 +1781,16 @@ public abstract class EntityHuman extends EntityLiving {
         }
     }
 
+    // Titanium start
+    public boolean isNether() {
+        return this.world.getBiome(new BlockPosition(this.locX, this.locY, this.locZ)) == BiomeBase.HELL;
+    }
+
+    public boolean isEnd() {
+        return this.world.getBiome(new BlockPosition(this.locX, this.locY, this.locZ)) == BiomeBase.SKY;
+    }
+    // Titanium end
+
     static class SyntheticClass_1 {
 
         static final int[] a = new int[EnumDirection.values().length];
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 70d303a64..c97664137 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -356,7 +356,25 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     }
 
     public boolean everyoneDeeplySleeping() {
-        if (this.O && !this.isClientSide) {
+        // PaperBin start - WorldServer#everyoneDeeplySleeping optimization
+        if (this.players.isEmpty() || this.isClientSide || !this.O) {
+            return false;
+        }
+
+        for (EntityHuman human : this.players) {
+            if (!human.isSpectator() && !human.isDeeplySleeping() && !human.fauxSleeping) {
+                return false;
+            }
+
+            // Titanium start - ignore players who are not in overworld
+            if (!human.isNether() && !human.isEnd()) {
+                return false;
+            }
+            // Titanium end
+        }
+
+        return true;
+        /*if (this.O && !this.isClientSide) {
             Iterator iterator = this.players.iterator();
 
             // CraftBukkit - This allows us to assume that some people are in bed but not really, allowing time to pass in spite of AFKers
@@ -381,7 +399,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
             return false;
         } else {
             return false;
-        }
+        }*/
+        // PaperBin end
     }
 
     protected void h() {
-- 
2.36.0.windows.1

