From 155e8fc455fa79a905f9e0dea239ea19ca84f872 Mon Sep 17 00:00:00 2001
From: RoccoDev <roccodev.business@gmail.com>
Date: Tue, 15 Sep 2020 12:24:14 +0200
Subject: [PATCH] Implement [KigPaper-0112] Only trigger
 PlayerConnection#disconnect once by RoccoDev


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 5f655390e..5d8b685ea 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -91,6 +91,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
     private double q;
     private boolean checkMovement = true;
     private boolean processedDisconnect; // CraftBukkit - added
+    private boolean disconnecting; // KigPaper - added
 
     public PlayerConnection(MinecraftServer minecraftserver, NetworkManager networkmanager, EntityPlayer entityplayer) {
         this.minecraftServer = minecraftserver;
@@ -162,6 +163,9 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
     }
 
     public void disconnect(String s) {
+        // KigPaper - if already disconnecting, do nothing
+        if(disconnecting) return;
+
         // CraftBukkit start - fire PlayerKickEvent
         String leaveMessage = EnumChatFormat.YELLOW + this.player.getName() + " left the game.";
 
@@ -175,6 +179,9 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             // Do not kick the player
             return;
         }
+        // KigPaper - mark the player as disconnecting, so all other disconnect requests are ignored.
+        disconnecting = true;
+
         // Send the possibly modified leave message
         s = event.getReason();
         // CraftBukkit end
-- 
2.36.0.windows.1

