From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 28 Aug 2022 15:26:28 +0200
Subject: [PATCH] Move AsyncUtil to net.titanium.util.async


diff --git a/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java b/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java
index e2e5f8d6e50add4b980ab2989692be43c3474aab..52d0fa190d2ee4bff2d7334c39b831822f34660e 100644
--- a/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java
+++ b/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java
@@ -1,7 +1,7 @@
 package ga.windpvp.windspigot.async.entitytracker;
 
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
-import ga.windpvp.windspigot.async.AsyncUtil;
+import net.titanium.util.async.AsyncUtil;
 import ga.windpvp.windspigot.async.ResettableLatch;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
diff --git a/src/main/java/ga/windpvp/windspigot/async/pathsearch/SearchHandler.java b/src/main/java/ga/windpvp/windspigot/async/pathsearch/SearchHandler.java
index 79a7340e453708db607865555439977b6e752d6c..dd1276c28ca4168f5498401a59e6940f76080054 100644
--- a/src/main/java/ga/windpvp/windspigot/async/pathsearch/SearchHandler.java
+++ b/src/main/java/ga/windpvp/windspigot/async/pathsearch/SearchHandler.java
@@ -1,7 +1,7 @@
 package ga.windpvp.windspigot.async.pathsearch;
 
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
-import ga.windpvp.windspigot.async.AsyncUtil;
+import net.titanium.util.async.AsyncUtil;
 import ga.windpvp.windspigot.async.pathsearch.cache.SearchCacheEntryEntity;
 import ga.windpvp.windspigot.async.pathsearch.cache.SearchCacheEntryPosition;
 import java.util.concurrent.ExecutorService;
diff --git a/src/main/java/ga/windpvp/windspigot/commons/ConcurrentIntHashMap.java b/src/main/java/ga/windpvp/windspigot/commons/ConcurrentIntHashMap.java
index 4b5ec7c851f873a28ca465e346f274cec8e16b33..1c88d3789d6ab7c6db37e8c8516ce8fe0144cf8f 100644
--- a/src/main/java/ga/windpvp/windspigot/commons/ConcurrentIntHashMap.java
+++ b/src/main/java/ga/windpvp/windspigot/commons/ConcurrentIntHashMap.java
@@ -1,7 +1,7 @@
 package ga.windpvp.windspigot.commons;
 
 import net.minecraft.server.IntHashMap;
-import static ga.windpvp.windspigot.async.AsyncUtil.runSynchronized;
+import static net.titanium.util.async.AsyncUtil.runSynchronized;
 
 public class ConcurrentIntHashMap<V> extends IntHashMap<V> {
 
diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
index 9b267e871736c48e06b632b6e70fbe62db8adabc..560d15777064b28665eebd429f0beff526156d6b 100644
--- a/src/main/java/net/minecraft/server/Explosion.java
+++ b/src/main/java/net/minecraft/server/Explosion.java
@@ -3,7 +3,7 @@ package net.minecraft.server;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
-import ga.windpvp.windspigot.async.AsyncUtil;
+import net.titanium.util.async.AsyncUtil;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
diff --git a/src/main/java/ga/windpvp/windspigot/async/AsyncUtil.java b/src/main/java/net/titanium/util/async/AsyncUtil.java
similarity index 98%
rename from src/main/java/ga/windpvp/windspigot/async/AsyncUtil.java
rename to src/main/java/net/titanium/util/async/AsyncUtil.java
index 8bbd109e13469488fa21a92bfacbcbf1208c0013..b4dce733ca8674f458f078c1283f12bdf102ec0b 100644
--- a/src/main/java/ga/windpvp/windspigot/async/AsyncUtil.java
+++ b/src/main/java/net/titanium/util/async/AsyncUtil.java
@@ -1,4 +1,4 @@
-package ga.windpvp.windspigot.async;
+package net.titanium.util.async;
 
 import java.util.concurrent.Executor;
 import java.util.concurrent.ForkJoinPool;
