From ab1ffe9335a7de4be3916731c94e1d33fc991723 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 20 Jul 2022 11:10:02 +0200
Subject: [PATCH] Rebrand console to Titanium


diff --git a/src/main/java/net/titanium/console/TitaniumConsole.java b/src/main/java/net/titanium/console/TitaniumConsole.java
index 4825477ef..7f2101c44 100644
--- a/src/main/java/net/titanium/console/TitaniumConsole.java
+++ b/src/main/java/net/titanium/console/TitaniumConsole.java
@@ -32,7 +32,7 @@ public class TitaniumConsole extends SimpleTerminalConsole {
     @Override
     protected LineReader buildReader(LineReaderBuilder builder) {
         return super.buildReader(builder
-                .appName("PandaSpigot")
+                .appName("Titanium")
                 .variable(LineReader.HISTORY_FILE, Paths.get(".console_history"))
                 .completer(new TitaniumConsoleCompleter(this.server)));
     }
-- 
2.36.0.windows.1

