From d2149079622312da0efcf1202335c07a841b81de Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 8 Jul 2022 15:50:54 +0200
Subject: [PATCH] Rebrand AsyncEntityTracker thread name


diff --git a/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java b/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java
index 8f291fe84..d8f1f7136 100644
--- a/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java
+++ b/src/main/java/ga/windpvp/windspigot/async/entitytracker/AsyncEntityTracker.java
@@ -14,7 +14,7 @@ import net.titanium.config.TitaniumConfig;
 
 public class AsyncEntityTracker extends EntityTracker {
 
-    private static final ExecutorService trackingThreadExecutor = Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("WindSpigot Entity Tracker Thread").build());
+    private static final ExecutorService trackingThreadExecutor = Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("Titanium Entity Tracker Thread").build());
     private static final List<NetworkManager> disabledFlushes = Lists.newArrayList();
 
     private final ResettableLatch latch = new ResettableLatch(TitaniumConfig.get().entityTracker.async.threads);
-- 
2.36.0.windows.1

