From 47ac99702b7fb78e846c9399b517776feaec3a56 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Tue, 6 Oct 2015 19:10:21 -0700
Subject: [PATCH] Implement [Taco-0010] Optimize armor stands by Techcable


diff --git a/src/main/java/net/minecraft/server/EntityArmorStand.java b/src/main/java/net/minecraft/server/EntityArmorStand.java
index 2ea4a5fd3..c37f5fa68 100644
--- a/src/main/java/net/minecraft/server/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/server/EntityArmorStand.java
@@ -705,4 +705,13 @@ public class EntityArmorStand extends EntityLiving {
     public boolean ad() {
         return super.ad() && !this.s();
     }
+
+    // TacoSpigot start - add an option to make armor stands not move
+    @Override
+    public void move(double motX, double motY, double motZ) {
+        if (getWorld().tacoConfigTitanium.armorStand.optimizeMovement) return;
+        super.move(motX, motY, motZ);
+    }
+    // TacoSpigot end
+
 }
diff --git a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
index b249233f5..10693adef 100644
--- a/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/titanium/config/TacoSpigotWorldConfig.java
@@ -46,6 +46,16 @@ public class TacoSpigotWorldConfig extends TitaniumConfigSection {
 
     }
 
+    public ArmorStand armorStand = new ArmorStand();
+
+    public static class ArmorStand extends TitaniumConfigSection {
+
+        @Comment("Whether to disable movement of armor stands (gravity too).")
+        @Comment("Enabling this could increase performance, but it's not vanilla behavior.")
+        public boolean optimizeMovement = false;
+
+    }
+
     @Comment("Whether to allow non-players (animals, monsters etc.) to be on scoreboard.")
     public boolean allowNonPlayerEntitiesOnScoreboard = true;
 
-- 
2.36.0.windows.1

