From 9ef42effcb2495bd4f5d60c3a06dee038a67fc93 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 22 Jul 2022 13:48:07 +0200
Subject: [PATCH] Implement [WindSpigot-????] Concurrent map for tracked
 players


diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 1b70d8a44..cddc587a7 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -45,7 +46,7 @@ public class EntityTrackerEntry {
     // PaperSpigot start
     // Replace trackedPlayers Set with a Map. The value is true until the player receives
     // their first update (which is forced to have absolute coordinates), false afterward.
-    public java.util.Map<EntityPlayer, Boolean> trackedPlayerMap = new it.unimi.dsi.fastutil.objects.Reference2BooleanOpenHashMap<>(); // IonSpigot - HashMap -> Reference2BooleanMap
+    public java.util.Map<EntityPlayer, Boolean> trackedPlayerMap = Maps.newConcurrentMap(); // WindSpigot -> Reference2BooleanMap -> ConcurrentMap // IonSpigot - HashMap -> Reference2BooleanMap
     public Set<EntityPlayer> trackedPlayers = trackedPlayerMap.keySet();
     // PaperSpigot end
 
-- 
2.36.0.windows.1

