From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 22:12:12 +0200
Subject: [PATCH] Option to disable explosion particles


diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index e3f9894810d28137156a269a9814d78cf8268d7b..f99cd3a616da133b5da4c568ced5d9df379c5cf6 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import net.titanium.config.TitaniumConfig;
 import org.bukkit.event.entity.ExplosionPrimeEvent; // CraftBukkit
 
 public class EntityTNTPrimed extends Entity {
@@ -90,6 +91,7 @@ public class EntityTNTPrimed extends Entity {
             // CraftBukkit end
         } else {
             this.W();
+            if (TitaniumConfig.get().gameMechanics.disableExplosionParticles) return; // Titanium - Option to disable explosion particles
             this.world.addParticle(EnumParticle.SMOKE_NORMAL, this.locX, this.locY + 0.5D, this.locZ, 0.0D, 0.0D, 0.0D, new int[0]);
         }
 
diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
index d2290a02a88ffc661ea87686ef6d5d196b2d295b..876aea3ffcaa6c53eb3068b4e96c6b6cd947ed00 100644
--- a/src/main/java/net/minecraft/server/Explosion.java
+++ b/src/main/java/net/minecraft/server/Explosion.java
@@ -245,11 +245,13 @@ public class Explosion {
         float volume = source instanceof EntityTNTPrimed ? world.paperConfigTitanium.tntExplosionVolume : 4.0F;
         this.world.makeSound(this.posX, this.posY, this.posZ, "random.explode", volume, (1.0F + (this.world.random.nextFloat() - this.world.random.nextFloat()) * 0.2F) * 0.7F);
         // PaperSpigot end
+        if (!TitaniumConfig.get().gameMechanics.disableExplosionParticles) { // Titanium - Option to disable explosion particles
         if (this.size >= 2.0F && this.b) {
             this.world.addParticle(EnumParticle.EXPLOSION_HUGE, this.posX, this.posY, this.posZ, 1.0D, 0.0D, 0.0D, new int[0]);
         } else {
             this.world.addParticle(EnumParticle.EXPLOSION_LARGE, this.posX, this.posY, this.posZ, 1.0D, 0.0D, 0.0D, new int[0]);
         }
+        }
 
         Iterator iterator;
         BlockPosition blockposition;
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 955088cd5f71dfc085e71b491c40c051f75b192e..5118cd64ef76005f8fe4234a1d0fb4b5eaf3c42a 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1109,6 +1109,11 @@ public class WorldServer extends World implements IAsyncTaskHandler {
             explosion.clearBlocks();
         }
 
+        // Titanium start - Option to disable explosion particles
+        if (this.server.server.titaniumServer.getServerConfig().gameMechanics.disableExplosionParticles) {
+            return explosion;
+        }
+        // Titanium end
         Iterator iterator = this.players.iterator();
 
         while (iterator.hasNext()) {
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index f6a2b6911d9f59202f775d624ec0923435f8d0a6..ec161cd66bec9de81d341ecfa7b733cab5f4453b 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -109,6 +109,10 @@ public class TitaniumConfig extends TitaniumConfigSection {
         @Comment("Be careful with this option and use it only if you know what are you doing.")
         public boolean disableChunkTicking = false;
 
+        @Comment("Whether to disable explosion particles.")
+        @Comment("Mostly optimization for clients.")
+        public boolean disableExplosionParticles = false;
+
     }
 
     @Comment("Whether saving of fireworks and arrows should be disabled.")
