From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 23:17:20 +0200
Subject: [PATCH] Option to disable head rotation handling


diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 818f07dac2488ca882356e7f96983e8d8e84f71e..646732d2f5d613c38a1369d75a2338a9e88d9bcf 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -203,6 +203,8 @@ public class EntityTrackerEntry {
             int i;
             int j;
 
+            boolean handleHeadRotation = this.handleHeadRotation(); // Titanium - Option to disable head rotation handling
+
             if (this.tracker.vehicle == null) {
                 ++this.v;
                 i = MathHelper.floor(this.tracker.locX * 32.0D);
@@ -239,7 +241,7 @@ public class EntityTrackerEntry {
                         if ((!flag || !flag1) && !(this.tracker instanceof EntityArrow)) {
                             if (flag) {
                                 object = new PacketPlayOutEntity.PacketPlayOutRelEntityMove(this.tracker.getId(), (byte) j1, (byte) k1, (byte) l1, this.tracker.onGround);
-                            } else if (flag1) {
+                            } else if (flag1 && handleHeadRotation) { // Titanium - Option to disable head rotation handling
                                 object = new PacketPlayOutEntity.PacketPlayOutEntityLook(this.tracker.getId(), (byte) l, (byte) i1, this.tracker.onGround);
                             }
                         } else {
@@ -312,7 +314,7 @@ public class EntityTrackerEntry {
                 // CraftBukkit end */
 
                 this.x = false;
-            } else {
+            } else if (handleHeadRotation) { // Titanium - Option to disable head rotation handling){
                 i = MathHelper.d(this.tracker.yaw * 256.0F / 360.0F);
                 j = MathHelper.d(this.tracker.pitch * 256.0F / 360.0F);
                 boolean flag2 = Math.abs(i - this.yRot) >= 4 || Math.abs(j - this.xRot) >= 4;
@@ -331,7 +333,7 @@ public class EntityTrackerEntry {
             }
 
             i = MathHelper.d(this.tracker.getHeadRotation() * 256.0F / 360.0F);
-            if (Math.abs(i - this.i) >= 4) {
+            if (handleHeadRotation && Math.abs(i - this.i) >= 4) {
                 this.broadcast(new PacketPlayOutEntityHeadRotation(this.tracker, (byte) i));
                 this.i = i;
             }
@@ -510,7 +512,7 @@ public class EntityTrackerEntry {
                     }
 
                     // CraftBukkit start - Fix for nonsensical head yaw
-                    if (this.tracker instanceof EntityLiving) { // [SportPaper-0027] Fix head rotation packet spam
+                    if (this.tracker instanceof EntityLiving && this.handleHeadRotation()) { // [SportPaper-0027] Fix head rotation packet spam
                         this.i = MathHelper.d(this.tracker.getHeadRotation() * 256.0F / 360.0F);
                         // SportPaper start
                         /* This was originally introduced by CraftBukkit, though the implementation is wrong since it's broadcasting
@@ -691,4 +693,10 @@ public class EntityTrackerEntry {
         }
 
     }
+
+    // Titanium start - Option to disable entity head rotation handling
+    private boolean handleHeadRotation() {
+        return this.tracker instanceof EntityHuman || this.tracker instanceof EntityVillager || !TitaniumConfig.get().gameMechanics.disableHeadRotationHandling;
+    }
+    // Titanium end
 }
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 0cc3a807b323e4f26033ed61f3d34c449b2eb42f..de985e47e13214f268f9de8633ea9bd44a142193 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -115,6 +115,10 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
         @Comment("Whether to disable sync of position and velocity of TNT in water to client.")
         public boolean disableTntInWaterSync = false;
+        
+        @Comment("Whether to disable head rotation handling (like animals).")
+        @Comment("Only head rotation for Players and Villagers would be still handled")
+        public boolean disableHeadRotationHandling = false;
 
         public LightDetector lightDetector = new LightDetector();
 
