From 9c15a4768304d1f8f24066376001cf0b9f33a2f9 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sat, 27 Aug 2022 19:51:26 +0200
Subject: [PATCH] Configurable player data saving


diff --git a/src/main/java/net/minecraft/server/WorldNBTStorage.java b/src/main/java/net/minecraft/server/WorldNBTStorage.java
index d508977f..43416807 100644
--- a/src/main/java/net/minecraft/server/WorldNBTStorage.java
+++ b/src/main/java/net/minecraft/server/WorldNBTStorage.java
@@ -8,6 +8,7 @@ import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import net.titanium.config.TitaniumConfig;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -181,6 +182,11 @@ public class WorldNBTStorage implements IDataManager, IPlayerFileData {
     }
 
     public void save(EntityHuman entityhuman) {
+        // Titanium (Paper) start - configurable player data saving
+        if (!TitaniumConfig.get().paper.settings.savePlayerData) {
+            return;
+        }
+        // Titanium (Paper) end
         try {
             NBTTagCompound nbttagcompound = new NBTTagCompound();
 
@@ -201,6 +207,11 @@ public class WorldNBTStorage implements IDataManager, IPlayerFileData {
     }
 
     public NBTTagCompound load(EntityHuman entityhuman) {
+        // Titanium (Paper) start - configurable player data saving
+        if (!TitaniumConfig.get().paper.settings.savePlayerData) {
+            return null;
+        }
+        // Titanium (Paper) end
         NBTTagCompound nbttagcompound = null;
 
         try {
-- 
2.36.0.windows.1

