From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 17 Oct 2022 22:32:04 +0200
Subject: [PATCH] Make tab complete spam limiter configurable

Backported from newer paper versions

diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 57e63aa64231dc05375af2592ad64d3f766eb893..b1b1b547f17e8ec1298d1998a92f9dc4602da72d 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1975,7 +1975,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
     public void a(PacketPlayInTabComplete packetplayintabcomplete) {
         //PlayerConnectionUtils.ensureMainThread(packetplayintabcomplete, this, this.player.u()); // Paper - run this async
         // CraftBukkit start
-        if (chatSpamField.addAndGet(this, 10) > 500 && !this.minecraftServer.getPlayerList().isOp(this.player.getProfile())) {
+        if (chatSpamField.addAndGet(this, TitaniumConfig.get().paper.settings.spamLimiter.tabSpamIncrement) > TitaniumConfig.get().paper.settings.spamLimiter.tabSpamLimit && !this.minecraftServer.getPlayerList().isOp(this.player.getProfile())) { // Titanium - backport spam limiter configuration
             minecraftServer.postToMainThread(() -> this.disconnect("disconnect.spam")); // Paper
             return;
         }
diff --git a/src/main/java/net/titanium/config/PaperConfig.java b/src/main/java/net/titanium/config/PaperConfig.java
index 64208b6fcd2b3328fd06ef88bfaab51f0dde88a4..739d38d38fad32d74bcd9500be26fbc99b083bae 100644
--- a/src/main/java/net/titanium/config/PaperConfig.java
+++ b/src/main/java/net/titanium/config/PaperConfig.java
@@ -27,6 +27,18 @@ public class PaperConfig extends TitaniumConfigSection {
         @Comment("Whether interaction packets should be limited from clients.")
         public boolean limitPlayerInteractions = true;
 
+        public SpamLimiter spamLimiter = new SpamLimiter();
+
+        public static class SpamLimiter extends TitaniumConfigSection {
+
+            @Comment("The value which spam limiter will increase by when a player will tab complete.")
+            public int tabSpamIncrement = 10;
+
+            @Comment("The spam limiter threshold at which the player will be kicked.")
+            public int tabSpamLimit = 500;
+
+        }
+
         @Comment("Speed at which baby zombies can move.")
         public double babyZombieMovementSpeed = 0.5;
 
