From f5d6156eaef16af3024b5e420f0d1696dcfc080e Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 21:33:18 +0200
Subject: [PATCH] Small EntityTracker optimization


diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 6a94bbb9d..7bb549f61 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -232,10 +232,13 @@ public class EntityTrackerEntry {
                 int k1 = j - this.yLoc;
                 int l1 = k - this.zLoc;
                 Object object = null;
-                boolean flag = Math.abs(j1) >= 4 || Math.abs(k1) >= 4 || Math.abs(l1) >= 4 || this.m % 60 == 0;
-                boolean flag1 = Math.abs(l - this.yRot) >= 4 || Math.abs(i1 - this.xRot) >= 4;
+                // Titanium start - moved down
+                /*boolean flag = Math.abs(j1) >= 4 || Math.abs(k1) >= 4 || Math.abs(l1) >= 4 || this.m % 60 == 0;
+                boolean flag1 = Math.abs(l - this.yRot) >= 4 || Math.abs(i1 - this.xRot) >= 4;*/
+                // Titanium end
 
                 if (this.m > 0 || this.tracker instanceof EntityArrow) { // PaperSpigot - Moved up
+                    boolean flag = Math.abs(j1) >= 4 || Math.abs(k1) >= 4 || Math.abs(l1) >= 4 || this.m % 60 == 0;
                     // CraftBukkit start - Code moved from below
                     if (flag) {
                         this.xLoc = i;
@@ -243,6 +246,7 @@ public class EntityTrackerEntry {
                         this.zLoc = k;
                     }
 
+                    boolean flag1 = Math.abs(l - this.yRot) >= 4 || Math.abs(i1 - this.xRot) >= 4;
                     if (flag1) {
                         this.yRot = l;
                         this.xRot = i1;
-- 
2.36.0.windows.1

