From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RoccoDev <hey@rocco.dev>
Date: Mon, 26 Apr 2021 09:50:16 +0200
Subject: [PATCH] Implement [KigPaper-0170] Kick after receiving invalid
 creative action packet by RoccoDev


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index edf70ec9f24331ede7a6dce4e4b31378c8b019a9..7ff8ae0e4575103e86f96fd5398bd079415526ee 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1881,6 +1881,8 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                     entityitem.j();
                 }
             }
+        } else {
+            disconnect("Not in creative mode"); // KigPaper
         }
 
     }
