From 4380cf3475c3d948a38fd94c3d349b245390ba26 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 23:32:55 +0200
Subject: [PATCH] Option to optimize armor stands


diff --git a/src/main/java/net/minecraft/server/EntityArmorStand.java b/src/main/java/net/minecraft/server/EntityArmorStand.java
index c37f5fa68..9f2ad7df1 100644
--- a/src/main/java/net/minecraft/server/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/server/EntityArmorStand.java
@@ -3,6 +3,7 @@ package net.minecraft.server;
 import java.util.List;
 
 // CraftBukkit start
+import net.titanium.config.TitaniumConfig;
 import org.bukkit.inventory.EquipmentSlot;
 import org.bukkit.craftbukkit.CraftEquipmentSlot;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
@@ -51,12 +52,12 @@ public class EntityArmorStand extends EntityLiving {
     }
 
     public boolean bM() {
-        return super.bM() && !this.hasGravity();
+        return super.bM() && (TitaniumConfig.get().gameMechanics.optimizeArmorStands || !this.hasGravity()); // Titanium - optimize armor stands
     }
 
     protected void h() {
         super.h();
-        this.datawatcher.a(10, Byte.valueOf((byte) 0));
+        this.datawatcher.a(10, (byte) 0);
         this.datawatcher.a(11, EntityArmorStand.a);
         this.datawatcher.a(12, EntityArmorStand.b);
         this.datawatcher.a(13, EntityArmorStand.c);
@@ -245,6 +246,11 @@ public class EntityArmorStand extends EntityLiving {
     protected void s(Entity entity) {}
 
     protected void bL() {
+        // Titanium start - Optimize armor stands
+        if (TitaniumConfig.get().gameMechanics.optimizeArmorStands) {
+            return;
+        }
+        // Titanium end
         List list = this.world.getEntities(this, this.getBoundingBox());
 
         if (list != null && !list.isEmpty()) {
@@ -489,6 +495,11 @@ public class EntityArmorStand extends EntityLiving {
     }
 
     public void g(float f, float f1) {
+        // Titanium start - optimize armor stands
+        if (TitaniumConfig.get().gameMechanics.optimizeArmorStands) {
+            return;
+        }
+        // Titanium end
         if (!this.hasGravity()) {
             super.g(f, f1);
         }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index e13b3c992..7668c02c0 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1337,7 +1337,8 @@ public abstract class World implements IBlockAccess {
         }
         // Spigot end
 
-        if (TitaniumConfig.get().gameMechanics.disableEntitiesCollisions && (entity instanceof EntityHuman || entity instanceof EntityInsentient || entity instanceof EntityArmorStand)) return arraylist; // Titanium - Option to disable entities collisions
+        if (TitaniumConfig.get().gameMechanics.disableEntitiesCollisions && (entity instanceof EntityHuman || entity instanceof EntityInsentient)) return arraylist; // Titanium - Option to disable entities collisions
+        if (TitaniumConfig.get().gameMechanics.optimizeArmorStands && entity instanceof EntityArmorStand) return arraylist; // Titanium - Option to disable armor stands collisions
 
         if (entity instanceof EntityItem) return arraylist; // PaperSpigot - Optimize item movement
 
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 716648ad6..222ef239e 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -127,6 +127,9 @@ public class TitaniumConfig extends TitaniumConfigSection {
         @Comment("Whether to disable entities collisions.")
         public boolean disableEntitiesCollisions = false;
 
+        @Comment("Whether to optimize armor stands (ignore collisions and disable gravity checks).")
+        public boolean optimizeArmorStands = false;
+
         public LightDetector lightDetector = new LightDetector();
 
         public static class LightDetector extends TitaniumConfigSection {
-- 
2.36.0.windows.1

