From 5f02d10a461619381dd089d5bf2d1d75befa802c Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 21:49:49 +0200
Subject: [PATCH] Option to disable statistics and achievements


diff --git a/src/main/java/net/minecraft/server/StatisticManager.java b/src/main/java/net/minecraft/server/StatisticManager.java
index 5d6c3deac..e80e5d932 100644
--- a/src/main/java/net/minecraft/server/StatisticManager.java
+++ b/src/main/java/net/minecraft/server/StatisticManager.java
@@ -2,6 +2,7 @@ package net.minecraft.server;
 
 import com.google.common.collect.Maps;
 import java.util.Map;
+import net.titanium.config.TitaniumConfig;
 
 public class StatisticManager {
 
@@ -18,6 +19,11 @@ public class StatisticManager {
     }
 
     public void b(EntityHuman entityhuman, Statistic statistic, int i) {
+        // Titanium start - option to disable statistics
+        if (TitaniumConfig.get().gameMechanics.disableStatistics) {
+            return;
+        }
+        // Titanium end
         if (!statistic.d() || this.b((Achievement) statistic)) {
             // CraftBukkit start - fire Statistic events
             org.bukkit.event.Cancellable cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.handleStatisticsIncrease(entityhuman, statistic, this.getStatisticValue(statistic), i);
@@ -30,6 +36,11 @@ public class StatisticManager {
     }
 
     public void setStatistic(EntityHuman entityhuman, Statistic statistic, int i) {
+        // Titanium start - option to disable statistics
+        if (TitaniumConfig.get().gameMechanics.disableStatistics) {
+            return;
+        }
+        // Titanium end
         StatisticWrapper statisticwrapper = (StatisticWrapper) this.a.get(statistic);
 
         if (statisticwrapper == null) {
@@ -41,6 +52,11 @@ public class StatisticManager {
     }
 
     public int getStatisticValue(Statistic statistic) {
+        // Titanium start - option to disable statistics
+        if (TitaniumConfig.get().gameMechanics.disableStatistics) {
+            return 0;
+        }
+        // Titanium end
         StatisticWrapper statisticwrapper = (StatisticWrapper) this.a.get(statistic);
 
         return statisticwrapper == null ? 0 : statisticwrapper.a();
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index 4fedffb61..f343e4e7a 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -102,6 +102,11 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
         public boolean tickFluids = true;
 
+        @Comment("Whether to disable statistics (ESC -> Statistics) and achievements.")
+        @Comment("This option could improve performance a little especially on large servers.")
+        @Comment("Of course it's not vanilla mechanic.")
+        public boolean disableStatistics = false;
+
     }
 
     @Comment("The max NBT size in bytes. The default value is 2097152 which is ~2Mb - it's insane big.")
-- 
2.36.0.windows.1

