From 86c6b953a55475ff156ad688e9ace43ddd56ccd9 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 21 Sep 2022 01:18:03 +0200
Subject: [PATCH] Backport EntityPickupItemEvent


diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
index f9129b71..6ef9663f 100644
--- a/src/main/java/net/minecraft/server/EntityInsentient.java
+++ b/src/main/java/net/minecraft/server/EntityInsentient.java
@@ -400,7 +400,11 @@ public abstract class EntityInsentient extends EntityLiving {
                 }
             }
 
-            if (flag && this.a(itemstack)) {
+            // Titanium start - Backport EntityPickupItemEvent
+            boolean canPickup = flag && this.a(itemstack);
+            canPickup = !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(this, entityitem, 0, !canPickup).isCancelled();
+            if (canPickup) {
+                // Titanium end
                 if (itemstack1 != null && this.random.nextFloat() - 0.1F < this.dropChances[i]) {
                     this.a(itemstack1, 0.0F);
                 }
diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 8e1692bf..4366f79d 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -3,6 +3,7 @@ package net.minecraft.server;
 import java.util.Iterator;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.event.entity.EntityPickupItemEvent;
 import org.bukkit.event.player.PlayerPickupItemEvent; // CraftBukkit
 import net.techcable.tacospigot.HopperPusher; // TacoSpigot
 
@@ -321,15 +322,35 @@ public class EntityItem extends Entity implements HopperPusher {
 
             if (this.pickupDelay <= 0 && canHold > 0) {
                 itemstack.count = canHold;
-                PlayerPickupItemEvent event = new PlayerPickupItemEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                // Titanium start - backport EntityPickupItemEvent
+                PlayerPickupItemEvent playerEvent = new PlayerPickupItemEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
                 // event.setCancelled(!entityhuman.canPickUpLoot); TODO
-                this.world.getServer().getPluginManager().callEvent(event);
-                itemstack.count = canHold + remaining;
+                this.world.getServer().getPluginManager().callEvent(playerEvent);
+                //itemstack.count = canHold + remaining; // Titanium - moved down
 
-                if (event.isCancelled()) {
+                if (playerEvent.isCancelled()) {
+                    itemstack.count = i; // SPIGOT-5294 - restore count
                     return;
                 }
 
+                // Call newer event afterwards
+                EntityPickupItemEvent entityEvent = new EntityPickupItemEvent(entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                //entityEvent.setCancelled(!entityhuman.canPickUpLoot); TODO
+                this.world.getServer().getPluginManager().callEvent(entityEvent);
+                if (entityEvent.isCancelled()) {
+                    itemstack.count = i; // SPIGOT-5294 - restore count
+                    return;
+                }
+
+                // Update the ItemStack if it was changed in the event
+                ItemStack current = this.getItemStack();
+                if (!itemstack.equals(current)) {
+                    itemstack = current;
+                } else {
+                    itemstack.count = canHold + remaining; // Titanium - moved down
+                }
+                // Titanium end
+
                 // Possibly < 0; fix here so we do not have to modify code below
                 this.pickupDelay = 0;
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 72d21f36..0e220ae4 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -146,7 +146,8 @@ public class CraftEventFactory {
         return event;
     }
 
-    // Titanium start - Backport EntityPlaceEvent
+    // Titanium start - Backport event
+    // EntityPlaceEvent
     public static EntityPlaceEvent callEntityPlaceEvent(World world, BlockPosition clickPosition, EnumDirection clickedFace, EntityHuman human, Entity entity) {
         Player who = (human == null) ? null : (Player) human.getBukkitEntity();
 
@@ -163,6 +164,14 @@ public class CraftEventFactory {
 
         return event;
     }
+
+    // EntityPickupItemEvent
+    public static EntityPickupItemEvent callEntityPickupItemEvent(Entity who, EntityItem item, int remaining, boolean cancelled) {
+        EntityPickupItemEvent event = new EntityPickupItemEvent((LivingEntity) who.getBukkitEntity(), (org.bukkit.entity.Item) item.getBukkitEntity(), remaining);
+        event.setCancelled(cancelled);
+        Bukkit.getPluginManager().callEvent(event);
+        return event;
+    }
     // Titanium end
 
     /**
-- 
2.36.0.windows.1

