From 75330f3d6e8eb9b2224f37fed7c1e8fc0f70d9f1 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 14 Sep 2022 21:37:47 +0200
Subject: [PATCH] Remove cast to short in Experience Orb saving


diff --git a/src/main/java/net/minecraft/server/EntityExperienceOrb.java b/src/main/java/net/minecraft/server/EntityExperienceOrb.java
index d4a5c2563..58bb432f5 100644
--- a/src/main/java/net/minecraft/server/EntityExperienceOrb.java
+++ b/src/main/java/net/minecraft/server/EntityExperienceOrb.java
@@ -145,7 +145,7 @@ public class EntityExperienceOrb extends Entity {
     public void b(NBTTagCompound nbttagcompound) {
         nbttagcompound.setShort("Health", (short) ((byte) this.d));
         nbttagcompound.setShort("Age", (short) this.b);
-        nbttagcompound.setInt("Value", (short) this.value);
+        nbttagcompound.setInt("Value", this.value);
     }
 
     public void a(NBTTagCompound nbttagcompound) {
-- 
2.36.0.windows.1

