From af4281eb7f4851a9ff89ff135609d7f1324bd935 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 21:17:20 +0200
Subject: [PATCH] Allow render distance from 1, not 3


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 6b5a6fb7..506cb272 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -278,7 +278,7 @@ public class PlayerChunkMap {
     }
 
     public void a(int i) {
-        i = MathHelper.clamp(i, 3, 32);
+        i = MathHelper.clamp(i, 1, 32); // Titanium -> Allow distance from 1, not 3
         if (i != this.g) {
             int j = i - this.g;
             ArrayList arraylist = Lists.newArrayList(this.managedPlayers);
@@ -318,7 +318,7 @@ public class PlayerChunkMap {
 
     // PaperSpigot start - Player view distance API
     public void updateViewDistance(EntityPlayer player, int viewDistance) {
-        viewDistance = MathHelper.clamp(viewDistance, 3, 32);
+        viewDistance = MathHelper.clamp(viewDistance, 1, 32); // Titanium -> Allow distance from 1, not 3
         if (viewDistance != player.viewDistance) {
             int cx = (int) player.locX >> 4;
             int cz = (int) player.locZ >> 4;
-- 
2.36.0.windows.1

