From dd5a68b96aa8fdebe4723940ec9a1899d0a256ca Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 22 Sep 2022 20:57:03 +0200
Subject: [PATCH] Add com.flowpowered.flow-nbt library (fix SWM patch)


diff --git a/build.gradle b/build.gradle
index fb69e8155..bd42eafb7 100644
--- a/build.gradle
+++ b/build.gradle
@@ -31,7 +31,10 @@ dependencies {
     implementation 'com.eatthepath:fast-uuid:0.2.0' // Nacho - Use jchambers' FAST UUID method
     implementation 'com.github.matejtymes:javafixes:1.3.7.1' // WindSpigot
     implementation 'com.github.ben-manes.caffeine:caffeine:2.9.3' // Titanium - use caffeine instead of guava cache
-    compileOnly 'com.mojang:datafixerupper:1.0.20' // Titanium - Native SWM support
+    // Titanium start - Native SWM support
+    implementation 'com.flowpowered:flow-nbt:2.0.0'
+    compileOnly 'com.mojang:datafixerupper:1.0.20'
+    // Titanium end
     // Titanium libraries end
 
     testImplementation 'junit:junit:4.13.2' // Titanium - Updated junit to 4.13.2
-- 
2.36.0.windows.1

