From ff959c952419627c52e8ebe5db6a44f5401ce59a Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 24 Oct 2022 11:24:32 +0200
Subject: [PATCH] Improve versioning and add source & branch to Metrics


diff --git a/build.gradle b/build.gradle
index fcd0b9eb4..6f4b4a67b 100644
--- a/build.gradle
+++ b/build.gradle
@@ -1,7 +1,5 @@
 import com.github.jengelman.gradle.plugins.shadow.transformers.Log4j2PluginsCacheFileTransformer
 
-apply plugin: 'com.palantir.git-version'
-
 String minecraftVersion = "1_8_R3"
 
 dependencies {
@@ -70,10 +68,15 @@ test {
 
 jar {
     manifest {
-        def details = versionDetails()
-        def gitBranch = details.branchName
-        def gitHash = details.gitHash
-        def implementationVersion = System.getenv("GITHUB_RUN_NUMBER") ?: gitHash
+        String implementationVersion = "CUSTOM-${gitBranch}-${gitHash}";
+        String ghBuildNumber = System.getenv("GITHUB_RUN_NUMBER");
+        String jenkinsBuildNumber = System.getenv("BUILD_NUMBER");
+        if (ghBuildNumber != null) {
+            implementationVersion = "GH-${gitBranch}-${ghBuildNumber}";
+        } else if (jenkinsBuildNumber != null) {
+            implementationVersion = "JENKINS-${gitBranch}-${jenkinsBuildNumber}";
+        }
+
         attributes(
                 "Main-Class": "org.bukkit.craftbukkit.Main",
                 "Implementation-Title": "CraftBukkit",
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 17adad63f..e60f87391 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -21,6 +21,7 @@ import org.apache.logging.log4j.Logger;
 import java.io.PrintStream;
 import org.apache.logging.log4j.Level;
 
+import org.bukkit.Bukkit;
 import org.bukkit.craftbukkit.LoggerOutputStream;
 import co.aikar.timings.SpigotTimings; // Spigot
 import org.bukkit.event.server.ServerCommandEvent;
@@ -179,7 +180,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
             // Spigot start
             this.a((PlayerList) (new DedicatedPlayerList(this)));
             // Spigot end
-            new net.titanium.Metrics(); // PandaSpigot
+            prepareMetrics(); // Titanium - moved to method
             java.net.SocketAddress bindAddress;
             if (this.getServerIp().startsWith("unix:")) {
                 if (!io.netty.channel.epoll.Epoll.isAvailable()) {
@@ -352,6 +353,34 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
         }
     }
 
+    // Titanium start - moved to method
+    private static void prepareMetrics() {
+        net.titanium.Metrics metrics = new net.titanium.Metrics(); // PandaSpigot
+        // Titanium start - Add source & branch pie
+        String bukkitVersion = Bukkit.getVersion();
+        int idx = bukkitVersion.indexOf(" (MC: ");
+        if (idx != -1) {
+            bukkitVersion = bukkitVersion.substring(0, idx);
+        }
+        String source = "Unknown";
+        String branch = "Unknown";
+        java.util.regex.Matcher versionMatcher = java.util.regex.Pattern.compile("(git-Titanium-)([a-zA-Z0-9]+)-([a-zA-Z0-9]+)-([a-zA-Z0-9]+)").matcher(bukkitVersion);
+        if (versionMatcher.find()) {
+            source = versionMatcher.group(2);
+            if (source.equalsIgnoreCase("GH")) {
+                source = "GitHub";
+            } else {
+                source = versionMatcher.group(2).toLowerCase();
+                source = source.substring(0, 1).toUpperCase() + source.substring(1);
+            }
+            branch = versionMatcher.group(3);
+        }
+        metrics.addCustomChart(new net.titanium.Metrics.SimplePie("source", source));
+        metrics.addCustomChart(new net.titanium.Metrics.SimplePie("branch", branch));
+        // Titanium end
+    }
+    // Titanium end
+
     // CraftBukkit start
     public PropertyManager getPropertyManager() {
         return this.propertyManager;
diff --git a/src/main/java/net/titanium/Metrics.java b/src/main/java/net/titanium/Metrics.java
index 07080ee82..dcb40013f 100644
--- a/src/main/java/net/titanium/Metrics.java
+++ b/src/main/java/net/titanium/Metrics.java
@@ -558,6 +558,13 @@ public class Metrics {
             this.callable = callable;
         }
 
+        // Titanium start
+        public SimplePie(String chartId, String value) {
+            super(chartId);
+            this.callable = () -> value;
+        }
+        // Titanium end
+
         @Override
         protected JsonObjectBuilder.JsonObject getChartData() throws Exception {
             String value = callable.call();
-- 
2.38.1.windows.1

