From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 13 Nov 2022 06:40:00 +0100
Subject: [PATCH] Option to disable fire ticking


diff --git a/src/main/java/net/minecraft/server/BlockFire.java b/src/main/java/net/minecraft/server/BlockFire.java
index d32552ff8904500cd27d8ca2d82e4460b631d98e..189ce852e70fa72d2c6048a8ba295e74066e0f5f 100644
--- a/src/main/java/net/minecraft/server/BlockFire.java
+++ b/src/main/java/net/minecraft/server/BlockFire.java
@@ -4,6 +4,7 @@ import com.google.common.collect.Maps;
 import java.util.Map;
 import java.util.Random;
 
+import net.titanium.config.TitaniumConfig;
 // CraftBukkit start
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.block.BlockBurnEvent;
@@ -113,6 +114,7 @@ public class BlockFire extends Block {
     }
 
     public void b(World world, BlockPosition blockposition, IBlockData iblockdata, Random random) {
+        if (TitaniumConfig.get().gameMechanics.disableFireTicking) return; // Titanium
         if (world.getGameRules().getBoolean("doFireTick")) {
             if (!this.canPlace(world, blockposition)) {
                 fireExtinguished(world, blockposition); // CraftBukkit - invalid place location
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index aaff30019bf74bd2ac801a99bd698105ce3b55b6..bf0390b8e9bd6c6189fe3bb9911f63ebc4326739 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -125,6 +125,10 @@ public class TitaniumConfig extends TitaniumConfigSection {
         @Comment("Whether to optimize armor stands (ignore collisions and disable gravity checks).")
         public boolean optimizeArmorStands = false;
 
+        @Comment("Whether to disable fire ticking (fire won't spread, blocks won't burn etc.).")
+        @Comment("You can enable this option if you don't need it.")
+        public boolean disableFireTicking = false;
+
         public LightDetector lightDetector = new LightDetector();
 
         public static class LightDetector extends TitaniumConfigSection {
