From 9f5fedc85216c8b18decc04763ad9061e17e7622 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 21:56:36 +0200
Subject: [PATCH] Option to disable chunk ticking


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 02c90ea85..566b46b56 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -405,6 +405,9 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     }
 
     protected void h() {
+        if (this.server.server.titaniumServer.getServerConfig().gameMechanics.disableChunkTicking) {
+            return;
+        }
         super.h();
         if (this.worldData.getType() == WorldType.DEBUG_ALL_BLOCK_STATES) {
             // Spigot start
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index f343e4e7a..9918f22d3 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -107,6 +107,11 @@ public class TitaniumConfig extends TitaniumConfigSection {
         @Comment("Of course it's not vanilla mechanic.")
         public boolean disableStatistics = false;
 
+        @Comment("Whether to disable chunk ticking.")
+        @Comment("It could be huge performance improvement, but also could cause a lot of issues.")
+        @Comment("Be careful with this option and use it only if you know what are you doing.")
+        public boolean disableChunkTicking = false;
+
     }
 
     @Comment("The max NBT size in bytes. The default value is 2097152 which is ~2Mb - it's insane big.")
-- 
2.36.0.windows.1

