From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: vemacs <d@nkmem.es>
Date: Mon, 27 Jun 2016 17:33:39 -0700
Subject: [PATCH] Implement [Taco-0022] Implement SpawnerPreSpawnEvent by
 vemacs


diff --git a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
index 699681e9b26877759da954d66031b51c161524fc..917746df5d94491dae2ca358d33236accd332b0b 100644
--- a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
+++ b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
@@ -103,6 +103,13 @@ public abstract class MobSpawnerAbstract {
                 boolean flag = false;
 
                 for (int i = 0; i < this.spawnCount; ++i) {
+                    // TacoSpigot start
+                    net.techcable.tacospigot.event.entity.SpawnerPreSpawnEvent event = new net.techcable.tacospigot.event.entity.SpawnerPreSpawnEvent(MCUtil.toLocation(this.a(), blockposition), mobType);
+                    if (!event.callEvent()) {
+                        flag = true;
+                        break;
+                    }
+                    // TacoSpigot end
                     Entity entity = EntityTypes.createEntityByName(this.getMobName(), this.a());
 
                     if (entity == null) {
