From 9b20b47a544e9d2d74393920a27fc2aad610dc14 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 31 Oct 2022 23:11:16 +0100
Subject: [PATCH] Add 'chunk.async-spawn-chunks' option to config


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index ae75466f4..bcf456f12 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -498,7 +498,7 @@ public abstract class MinecraftServer extends ReentrantIAsyncHandler<TasksPerTic
                     }
 
                     ++i;
-                    worldserver.chunkProviderServer.getChunkAt(blockposition.getX() + k >> 4, blockposition.getZ() + l >> 4, true, true, AsyncPriority.URGENT, null); // Titanium - use new methods
+                    worldserver.chunkProviderServer.getChunkAt(blockposition.getX() + k >> 4, blockposition.getZ() + l >> 4, true, true, AsyncPriority.URGENT, TitaniumConfig.get().chunk.asyncSpawnChunks ? c -> {} : null); // Titanium - use new methods
                 }
             }
         }
diff --git a/src/main/java/net/titanium/config/TitaniumConfig.java b/src/main/java/net/titanium/config/TitaniumConfig.java
index d2922e080..faa2aeeab 100644
--- a/src/main/java/net/titanium/config/TitaniumConfig.java
+++ b/src/main/java/net/titanium/config/TitaniumConfig.java
@@ -291,6 +291,10 @@ public class TitaniumConfig extends TitaniumConfigSection {
 
         }
 
+        @Comment("Whether to load (or generate if possible) spawn chunks asynchronously.")
+        @Comment("It can speed up server start by a lot, but also can cause some issues - for eg. player can appear under the map (and die in void), if joined before chunks were fully loaded/generated.")
+        public boolean asyncSpawnChunks = false;
+
     }
 
     public Tnt tnt = new Tnt();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 080ccb496..a7f8b5d68 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1145,7 +1145,7 @@ public final class CraftServer implements Server {
                     }
 
                     BlockPosition chunkcoordinates = internal.getSpawn();
-                    internal.chunkProviderServer.getChunkAt(chunkcoordinates.getX() + j >> 4, chunkcoordinates.getZ() + k >> 4, true, true, AsyncPriority.URGENT, null); // Titanium - use new methods
+                    internal.chunkProviderServer.getChunkAt(chunkcoordinates.getX() + j >> 4, chunkcoordinates.getZ() + k >> 4, true, true, AsyncPriority.URGENT, TitaniumConfig.get().chunk.asyncSpawnChunks ? c -> {} : null); // Titanium - use new methods
                 }
             }
         }
-- 
2.38.1.windows.1

