From d1206e2ae47eb501cbc6ac651f22fe54860ebc98 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 10 Sep 2018 23:36:16 -0400
Subject: [PATCH] Implement [Paper-0348] Prevent chunk loading from Fluid
 Flowing by Aikar


diff --git a/src/main/java/net/minecraft/server/BlockFluids.java b/src/main/java/net/minecraft/server/BlockFluids.java
index db73f5d9c..000849f2b 100644
--- a/src/main/java/net/minecraft/server/BlockFluids.java
+++ b/src/main/java/net/minecraft/server/BlockFluids.java
@@ -85,7 +85,13 @@ public abstract class BlockFluids extends Block {
             int k;
 
             if (j < 0) {
-                if (!iblockaccess.getType(blockposition1).getBlock().getMaterial().isSolid()) {
+                // Paper start
+                IBlockData iblockdata = iblockaccess.getTypeIfLoaded(blockposition1);
+                if (iblockdata == null) {
+                    continue;
+                }
+                if (!iblockdata.getBlock().getMaterial().isSolid()) {
+                    // Paper end
                     j = this.f(iblockaccess, blockposition1.down());
                     if (j >= 0) {
                         k = j - (i - 8);
@@ -139,7 +145,13 @@ public abstract class BlockFluids extends Block {
             for (int j = 0; j < i; ++j) {
                 EnumDirection enumdirection = aenumdirection[j];
 
-                if (enumdirection != EnumDirection.DOWN && world.getType(blockposition.shift(enumdirection)).getBlock().getMaterial() == Material.WATER) {
+                // Paper start
+                IBlockData iblockdata1 = world.getTypeIfLoaded(blockposition.shift(enumdirection));
+                if (iblockdata1 == null) {
+                    continue;
+                }
+                if (enumdirection != EnumDirection.DOWN && iblockdata1.getBlock().getMaterial() == Material.WATER) {
+                    // Paper end
                     flag = true;
                     break;
                 }
@@ -148,14 +160,14 @@ public abstract class BlockFluids extends Block {
             if (flag) {
                 Integer integer = (Integer) iblockdata.get(BlockFluids.LEVEL);
 
-                if (integer.intValue() == 0) {
-                    world.setTypeUpdate(blockposition, Blocks.OBSIDIAN.getBlockData());
+                if (integer == 0) {
+                    if (world.setTypeUpdateIfLoaded(blockposition, Blocks.OBSIDIAN.getBlockData())) // Paper
                     this.fizz(world, blockposition);
                     return true;
                 }
 
-                if (integer.intValue() > 0) { // PaperSpigot
-                    world.setTypeUpdate(blockposition, Blocks.COBBLESTONE.getBlockData());
+                if (integer > 0) { // PaperSpigot
+                    if (world.setTypeUpdateIfLoaded(blockposition, Blocks.COBBLESTONE.getBlockData())) // Paper
                     this.fizz(world, blockposition);
                     return true;
                 }
diff --git a/src/main/java/net/minecraft/server/IBlockAccess.java b/src/main/java/net/minecraft/server/IBlockAccess.java
index 304e82754..6e71b75ec 100644
--- a/src/main/java/net/minecraft/server/IBlockAccess.java
+++ b/src/main/java/net/minecraft/server/IBlockAccess.java
@@ -6,6 +6,12 @@ public interface IBlockAccess {
 
     IBlockData getType(BlockPosition blockposition);
 
+    // Titanium start
+    default IBlockData getTypeIfLoaded(BlockPosition blockposition) {
+        return this.getType(blockposition);
+    }
+    // Titanium end
+
     boolean isEmpty(BlockPosition blockposition);
 
     int getBlockPower(BlockPosition blockposition, EnumDirection enumdirection);
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 3d8f7044c..0d0bc88a5 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -598,6 +598,12 @@ public abstract class World implements IBlockAccess {
         return this.setTypeAndData(blockposition, iblockdata, 3);
     }
 
+    // Titanium start
+    public boolean setTypeUpdateIfLoaded(BlockPosition blockposition, IBlockData iblockdata) {
+        return this.setTypeAndDataIfLoaded(blockposition, iblockdata, 3);
+    }
+    // Titanium end
+
     public void notify(BlockPosition blockposition) {
         for (int i = 0; i < this.u.size(); ++i) {
             ((IWorldAccess) this.u.get(i)).a(blockposition);
@@ -919,6 +925,7 @@ public abstract class World implements IBlockAccess {
     private static final IBlockData AIR_DATA = Blocks.AIR.getBlockData(); // Titanium - Micro optimization
 
     // Paper start - reduces need to do isLoaded before getType
+    @Override
     public IBlockData getTypeIfLoaded(BlockPosition blockposition) {
         // CraftBukkit start - tree generation
         if (captureTreeGeneration) {
-- 
2.38.1.windows.1

