From 947ec2b83e0505ce6863e2e9d3e9787de02c80e9 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 16 Oct 2022 22:06:02 +0200
Subject: [PATCH] Fix typo in /ping placeholder


diff --git a/src/main/java/net/titanium/command/PingCommand.java b/src/main/java/net/titanium/command/PingCommand.java
index b3ee9856..74e5ba38 100644
--- a/src/main/java/net/titanium/command/PingCommand.java
+++ b/src/main/java/net/titanium/command/PingCommand.java
@@ -49,7 +49,7 @@ public class PingCommand extends Command {
         int pingValue = player.spigot().getPing();
 
         pingMessage = pingMessage
-                .replace("{FOMRATED-PING}", NumberRange.inRangeToString(pingValue, messages.ping.pingFormat, true, true))
+                .replace("{FORMATTED-PING}", NumberRange.inRangeToString(pingValue, messages.ping.pingFormat, true, true))
                 .replace("{PING}", Integer.toString(pingValue))
                 .replace("{PLAYER}", player.getName());
 
-- 
2.36.0.windows.1

