From 92d5388733442d3d6eff42e990043ec67227575c Mon Sep 17 00:00:00 2001
From: Heath Logan Campbell <github@heathlogancampbell.com>
Date: Fri, 31 Jul 2020 12:43:00 +0200
Subject: [PATCH] Implement [Nacho-0027] Netty IP_TOS 0x18 by Health Logan
 Campbell


diff --git a/src/main/java/net/minecraft/server/ServerConnection.java b/src/main/java/net/minecraft/server/ServerConnection.java
index 4a2a1b565..d21a9293d 100644
--- a/src/main/java/net/minecraft/server/ServerConnection.java
+++ b/src/main/java/net/minecraft/server/ServerConnection.java
@@ -117,6 +117,7 @@ public class ServerConnection {
                 protected void initChannel(Channel channel) throws Exception {
                     try {
                         channel.config().setOption(ChannelOption.TCP_NODELAY, Boolean.valueOf(true));
+                        channel.config().setOption(ChannelOption.IP_TOS, 0x18); // Nacho - Optimize networking by using IP_TOS
                     } catch (ChannelException channelexception) {
                         ;
                     }
-- 
2.38.1.windows.1

