From fcd5897aab875fd36f3386d96b8386cefee9b9ff Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 24 Oct 2022 01:07:33 +0200
Subject: [PATCH] Move AsyncPacketThread to net.titanium.util.async


diff --git a/src/main/java/ga/windpvp/windspigot/async/thread/CombatThread.java b/src/main/java/ga/windpvp/windspigot/async/thread/CombatThread.java
index f31d4218e..a9f4ab47a 100644
--- a/src/main/java/ga/windpvp/windspigot/async/thread/CombatThread.java
+++ b/src/main/java/ga/windpvp/windspigot/async/thread/CombatThread.java
@@ -1,6 +1,7 @@
 package ga.windpvp.windspigot.async.thread;
 
 import net.titanium.config.TitaniumConfig;
+import net.titanium.util.async.AsyncPacketThread;
 
 public class CombatThread extends AsyncPacketThread {
 
diff --git a/src/main/java/ga/windpvp/windspigot/async/thread/AsyncPacketThread.java b/src/main/java/net/titanium/util/async/AsyncPacketThread.java
similarity index 99%
rename from src/main/java/ga/windpvp/windspigot/async/thread/AsyncPacketThread.java
rename to src/main/java/net/titanium/util/async/AsyncPacketThread.java
index a1a6ec54e..0cf4d95a0 100644
--- a/src/main/java/ga/windpvp/windspigot/async/thread/AsyncPacketThread.java
+++ b/src/main/java/net/titanium/util/async/AsyncPacketThread.java
@@ -1,4 +1,4 @@
-package ga.windpvp.windspigot.async.thread;
+package net.titanium.util.async;
 
 import ga.windpvp.windspigot.async.netty.Spigot404Write;
 import io.netty.util.concurrent.Future;
diff --git a/src/main/java/org/spigotmc/TicksPerSecondCommand.java b/src/main/java/org/spigotmc/TicksPerSecondCommand.java
index 7087735af..1950fa752 100644
--- a/src/main/java/org/spigotmc/TicksPerSecondCommand.java
+++ b/src/main/java/org/spigotmc/TicksPerSecondCommand.java
@@ -1,13 +1,10 @@
 package org.spigotmc;
 
-import ga.windpvp.windspigot.async.thread.AsyncPacketThread;
 import ga.windpvp.windspigot.async.thread.CombatThread;
 import java.util.Arrays;
 import java.util.List;
-import java.util.stream.Collectors;
 import net.titanium.Titanium;
 import net.titanium.config.message.MessageConfig;
-import org.apache.commons.lang3.time.DurationFormatUtils;
 import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
 import org.bukkit.command.Command;
-- 
2.36.0.windows.1

