From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lew <skycubesv@gmail.com>
Date: Sun, 4 Dec 2022 13:07:04 +0100
Subject: [PATCH] Implement [Nacho-????] Use fastutil in HandshakeListener


diff --git a/src/main/java/net/minecraft/server/HandshakeListener.java b/src/main/java/net/minecraft/server/HandshakeListener.java
index 3d8743a29e35720b0564d4f2b2fc777e3caf5752..4f2e3ed1718bdd71f3ed41d17c6a6b8a1669a02c 100644
--- a/src/main/java/net/minecraft/server/HandshakeListener.java
+++ b/src/main/java/net/minecraft/server/HandshakeListener.java
@@ -11,7 +11,7 @@ public class HandshakeListener implements PacketHandshakingInListener {
 
     private static final com.google.gson.Gson gson = new com.google.gson.Gson(); // Spigot
     // CraftBukkit start - add fields
-    private static final HashMap<InetAddress, Long> throttleTracker = new HashMap<InetAddress, Long>();
+    private static final java.util.Map<InetAddress, Long> throttleTracker = new it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<>(); // Nacho
     private static int throttleCounter = 0;
     // CraftBukkit end
 
