From b5399dcf72a2b63ad9e183b4ae8a6367eea80b0d Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 26 Oct 2022 21:17:20 +0200
Subject: [PATCH] Add LivingEntity#removePotionEffects


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index cb0a5b90f..2f09cbef0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -323,6 +323,13 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().removeEffect(type.getId());
     }
 
+    // Titanium start
+    @Override
+    public void removePotionEffects() {
+        getHandle().removeAllEffects();
+    }
+    // Titanium end
+
     public Collection<PotionEffect> getActivePotionEffects() {
         List<PotionEffect> effects = new ArrayList<PotionEffect>();
         for (Object raw : getHandle().effects.values()) {
-- 
2.36.0.windows.1

