From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aleksander Jagiello <themolkapl@gmail.com>
Date: Sun, 24 Jan 2021 22:17:29 +0100
Subject: [PATCH] Implement [Paper-0567] Add getMainThreadExecutor to
 BukkitScheduler by TheMolkaPL


diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
index 5748f634570588645d2b8f36fe8e7c0063e87bbb..e62651f0a4722bddac4f2c3caf82878fef872829 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -569,4 +569,15 @@ public class CraftScheduler implements BukkitScheduler {
     public BukkitTask runTaskTimerAsynchronously(Plugin plugin, BukkitRunnable task, long delay, long period) throws IllegalArgumentException {
         return runTaskTimerAsynchronously(plugin, (Runnable) task, delay, period);
     }
+
+    // Paper start - add getMainThreadExecutor
+    @Override
+    public Executor getMainThreadExecutor(Plugin plugin) {
+        Validate.notNull(plugin, "Plugin cannot be null");
+        return command -> {
+            Validate.notNull(command, "Command cannot be null");
+            this.runTask(plugin, command);
+        };
+    }
+    // Paper end
 }
