From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 4 Jan 2023 17:32:32 +0100
Subject: [PATCH] Backport some message sending API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 958b2ffa201fef7a479cc66830cd9b624fea9302..f15989be18d6bee7fe88ce2316fd81277d4fd3e5 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -261,11 +261,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     // Titanium start
     @Override
-    public void sendActionBar(BaseComponent... message) {
-        if(getHandle().playerConnection == null) return;
-        PacketPlayOutChat packet = new PacketPlayOutChat(null, (byte) 2);
-        packet.components = message;
-        getHandle().playerConnection.sendPacket(packet);
+    public void sendActionBar(String message) {
+        if(getHandle().playerConnection == null || message == null || message.isEmpty()) return;
+        getHandle().playerConnection.sendPacket(new PacketPlayOutChat(IChatBaseComponent.ChatSerializer.a("{\"text\":\"" + message + "\"}"), (byte) 2));
     }
     // Titanium end
 
@@ -1642,6 +1640,24 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             getHandle().playerConnection.sendPacket(packet);
         }
 
+        @Override
+        public void sendMessage(net.md_5.bungee.api.ChatMessageType position, BaseComponent component) {
+            sendMessage( position, new BaseComponent[] { component } );
+        }
+
+        @Override
+        public void sendMessage(net.md_5.bungee.api.ChatMessageType position, BaseComponent... components) {
+            if ( getHandle().playerConnection == null ) return;
+
+            PacketPlayOutChat packet = new PacketPlayOutChat(null, (byte) position.ordinal());
+            // Action bar doesn't render colours, replace colours with legacy section symbols
+            if (position == net.md_5.bungee.api.ChatMessageType.ACTION_BAR) {
+                components = new BaseComponent[]{new net.md_5.bungee.api.chat.TextComponent(BaseComponent.toLegacyText(components))};
+            }
+            packet.components = components;
+            getHandle().playerConnection.sendPacket(packet);
+        }
+
         // PaperSpigot start - Implement affects spawning API
         @Override
         public boolean getAffectsSpawning() {
