From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Tue, 24 Jan 2023 20:07:33 +0100
Subject: [PATCH] Backport PlayerArmorChangeEvent


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 9f9da19e8fa510af0769b951f7bf013a70ee680a..cf33363b7b4e3e80ec13e34e649a0ec3d686860c 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1583,6 +1583,11 @@ public abstract class EntityLiving extends Entity {
                 ItemStack itemstack1 = this.getEquipment(j);
 
                 if (!ItemStack.matches(itemstack1, itemstack)) {
+                    // Titanium start - Backport PlayerArmorChangeEvent
+                    if (j > 0 && this instanceof EntityPlayer && com.destroystokyo.paper.event.player.PlayerArmorChangeEvent.getHandlerList().getRegisteredListeners().length > 0) {
+                        CraftEventFactory.callPlayerArmorChangeEvent((EntityPlayer) this, j, itemstack, itemstack1);
+                    }
+                    // Titanium end
                     ((WorldServer) this.world).getTracker().a((Entity) this, (Packet) (new PacketPlayOutEntityEquipment(this.getId(), j, itemstack1)));
                     if (itemstack != null) {
                         this.c.a(itemstack.B());
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index f72e54935e3cde77e8104651bc27acf00560a367..a854a5683f662e6cb68e2d69e4f3bd75cf436be1 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -197,6 +197,16 @@ public class CraftEventFactory {
 
         return event;
     }
+
+    // PlayerArmorChangeEvent
+    public static com.destroystokyo.paper.event.player.PlayerArmorChangeEvent callPlayerArmorChangeEvent(EntityPlayer player, int slot, ItemStack oldNms, ItemStack newNms) {
+        org.bukkit.inventory.ItemStack oldItem = CraftItemStack.asBukkitCopy(oldNms);
+        org.bukkit.inventory.ItemStack newItem = CraftItemStack.asBukkitCopy(newNms);
+        com.destroystokyo.paper.event.player.PlayerArmorChangeEvent.SlotType slotType = com.destroystokyo.paper.event.player.PlayerArmorChangeEvent.SlotType.getByOrdinal(slot);
+        com.destroystokyo.paper.event.player.PlayerArmorChangeEvent event = new com.destroystokyo.paper.event.player.PlayerArmorChangeEvent(player.getBukkitEntity(), slotType, oldItem, newItem);
+        Bukkit.getPluginManager().callEvent(event);
+        return event;
+    }
     // Titanium end
 
     /**
