From d811860d0d5465b1a65412eb91f4c19d0d43c775 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 24 Oct 2022 01:02:32 +0200
Subject: [PATCH] Always add lines with TPS of Hit Detection/Knockback Threads


diff --git a/src/main/java/net/titanium/config/message/MessageConfig.java b/src/main/java/net/titanium/config/message/MessageConfig.java
index 7f9a8174..b77b36be 100644
--- a/src/main/java/net/titanium/config/message/MessageConfig.java
+++ b/src/main/java/net/titanium/config/message/MessageConfig.java
@@ -69,8 +69,8 @@ public class MessageConfig extends TitaniumConfigSection implements Server.Titan
     @Comment("Supported placeholders: {TPS_AVG}, {TPS_1M}, {TPS_5M}, {TPS_15M}, {TPS_HITS_AVG}, {TPS_HITS_1M}, {TPS_HITS_5M}, {TPS_HITS_15M}, {TPS_KNOCKBACK_AVG}, {TPS_KNOCKBACK_1M}, {TPS_KNOCKBACK_5M}, {TPS_KNOCKBACK_15M}, {MEMORY_USED}, {FREE_MEMORY}, {MAX_MEMORY}, {UPTIME}")
     public List<String> ticksPerSeconds = Arrays.asList(
             "&6TPS from last 1m, 5m, 15m: &f{TPS_AVG}",
-            TitaniumConfig.get().combat.async.hitDetection ? "&6TPS (Hit Detection) from last 1m, 5m, 15m: &f{TPS_HITS_AVG}" : "",
-            TitaniumConfig.get().combat.async.knockback ? "&6TPS (Knockback) from last 1m, 5m, 15m: &f{TPS_KNOCKBACK_AVG}" : "",
+            "&6TPS (Hit Detection) from last 1m, 5m, 15m: &f{TPS_HITS_AVG}",
+            "&6TPS (Knockback) from last 1m, 5m, 15m: &f{TPS_KNOCKBACK_AVG}",
             " ",
             "&6Memory usage: &a{MEMORY_USED}&a/{MEMORY_MAX} MB",
             "&6Uptime: &a{UPTIME}"
-- 
2.36.0.windows.1

