From 524a455fb2955460a7499d47d494c63cd67a36ca Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 24 Oct 2022 11:24:32 +0200
Subject: [PATCH] Add Jenkins build number


diff --git a/build.gradle b/build.gradle
index bd42eafb..f6846890 100644
--- a/build.gradle
+++ b/build.gradle
@@ -74,7 +74,16 @@ jar {
         def details = versionDetails()
         def gitBranch = details.branchName
         def gitHash = details.gitHash
-        def implementationVersion = System.getenv("GITHUB_RUN_NUMBER") ?: gitHash
+
+        String implementationVersion = gitHash;
+        String ghBuildNumber = System.getenv("GITHUB_RUN_NUMBER");
+        String jenkinsBuildNumber = System.getenv("BUILD_NUMBER");
+        if (ghBuildNumber != null) {
+            implementationVersion = "${ghBuildNumber}-GH";
+        } else if (jenkinsBuildNumber != null) {
+            implementationVersion = "${jenkinsBuildNumber}-JENKINS";
+        }
+
         attributes(
                 "Main-Class": "org.bukkit.craftbukkit.Main",
                 "Implementation-Title": "CraftBukkit",
@@ -85,7 +94,8 @@ jar {
                 "Specification-Vendor": "Bukkit Team",
                 "Multi-Release": "true", // PandaSpigot
                 "Git-Branch": gitBranch,
-                "Git-Commit": gitHash
+                "Git-Commit": gitHash,
+                "Build-Number": implementationVersion,
         )
     }
 }
\ No newline at end of file
-- 
2.36.0.windows.1

