From e69387926ea1dda26f2c224356f9e589b61fae8c Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Tue, 25 Oct 2022 07:56:27 +0200
Subject: [PATCH] Temporary nLogin fix


diff --git a/src/main/java/net/titanium/config/SpigotConfig.java b/src/main/java/net/titanium/config/SpigotConfig.java
index c426e882..f1e2d061 100644
--- a/src/main/java/net/titanium/config/SpigotConfig.java
+++ b/src/main/java/net/titanium/config/SpigotConfig.java
@@ -114,6 +114,8 @@ public class SpigotConfig extends TitaniumConfigSection {
                 log("Debug logging is disabled");
             }
 
+            org.spigotmc.SpigotConfig.bungee = this.bungeecord; //TODO: Find better way to do this
+
             WatchdogThread.doStart(timeoutTime, restartOnCrash);
 
             log("Server Ping Player Sample Count: " + sampleCount);
-- 
2.36.0.windows.1

